{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "62e01815-131b-408d-a078-9a0126656d25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Searching for papers...\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=0\n",
      "getting the next page\n",
      "0\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=1\n",
      "getting the next page\n",
      "1\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=2\n",
      "getting the next page\n",
      "2\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=3\n",
      "getting the next page\n",
      "3\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=4\n",
      "getting the next page\n",
      "4\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=5\n",
      "getting the next page\n",
      "5\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=6\n",
      "getting the next page\n",
      "6\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=7\n",
      "getting the next page\n",
      "7\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=8\n",
      "getting the next page\n",
      "8\n",
      "https://www.biorxiv.org/search/%2522graph genome%2522%20numresults%3A75%20sort%3Apublication-date%20direction%3Adescending?page=9\n",
      "done finding papers\n",
      "found 744\n",
      "['https://www.biorxiv.org/content/10.1101/2024.07.30.605897v1', 'https://www.biorxiv.org/content/10.1101/2023.09.11.557209v2', 'https://www.biorxiv.org/content/10.1101/2024.07.26.605365v1', 'https://www.biorxiv.org/content/10.1101/2024.07.22.604635v1', 'https://www.biorxiv.org/content/10.1101/2024.06.03.597150v2', 'https://www.biorxiv.org/content/10.1101/2024.02.19.579948v3', 'https://www.biorxiv.org/content/10.1101/2024.07.22.603478v1', 'https://www.biorxiv.org/content/10.1101/2024.07.10.602853v2', 'https://www.biorxiv.org/content/10.1101/2024.07.02.601807v2', 'https://www.biorxiv.org/content/10.1101/2024.07.09.602727v1', 'https://www.biorxiv.org/content/10.1101/2024.07.09.602638v1', 'https://www.biorxiv.org/content/10.1101/2023.10.20.563230v3', 'https://www.biorxiv.org/content/10.1101/2024.07.08.602537v1', 'https://www.biorxiv.org/content/10.1101/2024.07.05.602100v1', 'https://www.biorxiv.org/content/10.1101/2024.07.02.600549v1', 'https://www.biorxiv.org/content/10.1101/2023.08.21.553699v3', 'https://www.biorxiv.org/content/10.1101/2024.07.01.601500v1', 'https://www.biorxiv.org/content/10.1101/2024.06.14.599082v2', 'https://www.biorxiv.org/content/10.1101/2024.03.06.583483v2', 'https://www.biorxiv.org/content/10.1101/2024.06.11.598241v1', 'https://www.biorxiv.org/content/10.1101/2024.06.10.598220v1', 'https://www.biorxiv.org/content/10.1101/2024.05.23.595509v2', 'https://www.biorxiv.org/content/10.1101/2020.10.01.322164v4', 'https://www.biorxiv.org/content/10.1101/2024.04.22.590609v2', 'https://www.biorxiv.org/content/10.1101/2024.05.29.596541v1', 'https://www.biorxiv.org/content/10.1101/2023.11.21.568151v5', 'https://www.biorxiv.org/content/10.1101/2024.05.24.595829v1', 'https://www.biorxiv.org/content/10.1101/2024.05.24.595489v1', 'https://www.biorxiv.org/content/10.1101/2024.05.20.595060v1', 'https://www.biorxiv.org/content/10.1101/2024.05.17.594728v1', 'https://www.biorxiv.org/content/10.1101/2024.05.17.594646v3', 'https://www.biorxiv.org/content/10.1101/2024.05.15.594184v1', 'https://www.biorxiv.org/content/10.1101/2023.11.30.568362v2', 'https://www.biorxiv.org/content/10.1101/2024.05.13.593871v1', 'https://www.biorxiv.org/content/10.1101/2023.07.26.550757v2', 'https://www.biorxiv.org/content/10.1101/2023.12.01.569581v2', 'https://www.biorxiv.org/content/10.1101/2024.05.13.591068v1', 'https://www.biorxiv.org/content/10.1101/2023.04.15.536996v3', 'https://www.biorxiv.org/content/10.1101/2024.03.21.586149v2', 'https://www.biorxiv.org/content/10.1101/2024.04.23.590803v1', 'https://www.biorxiv.org/content/10.1101/2023.11.03.565466v2', 'https://www.biorxiv.org/content/10.1101/2024.04.18.590093v1', 'https://www.biorxiv.org/content/10.1101/2023.01.20.524991v5', 'https://www.biorxiv.org/content/10.1101/2024.04.05.588231v1', 'https://www.biorxiv.org/content/10.1101/2024.03.29.587318v1', 'https://www.biorxiv.org/content/10.1101/2024.03.25.586631v1', 'https://www.biorxiv.org/content/10.1101/2022.11.04.514718v2', 'https://www.biorxiv.org/content/10.1101/2024.03.25.586701v1', 'https://www.biorxiv.org/content/10.1101/2024.01.29.577700v2', 'https://www.biorxiv.org/content/10.1101/2024.03.18.585512v1', 'https://www.biorxiv.org/content/10.1101/2024.03.17.585367v1', 'https://www.biorxiv.org/content/10.1101/2024.03.17.585420v1', 'https://www.biorxiv.org/content/10.1101/2024.03.13.584790v1', 'https://www.biorxiv.org/content/10.1101/2024.03.07.583855v3', 'https://www.biorxiv.org/content/10.1101/2024.03.11.584353v1', 'https://www.biorxiv.org/content/10.1101/2022.06.22.497172v2', 'https://www.biorxiv.org/content/10.1101/2024.03.06.582860v1', 'https://www.biorxiv.org/content/10.1101/2024.02.27.581927v1', 'https://www.biorxiv.org/content/10.1101/2024.02.25.581922v1', 'https://www.biorxiv.org/content/10.1101/2024.02.19.580943v1', 'https://www.biorxiv.org/content/10.1101/2024.02.14.580266v1', 'https://www.biorxiv.org/content/10.1101/2024.02.15.580441v1', 'https://www.biorxiv.org/content/10.1101/2024.02.18.580171v1', 'https://www.biorxiv.org/content/10.1101/2023.11.15.566493v2', 'https://www.biorxiv.org/content/10.1101/2023.09.13.557552v2', 'https://www.biorxiv.org/content/10.1101/2024.02.11.579857v1', 'https://www.biorxiv.org/content/10.1101/2022.12.13.520286v2', 'https://www.biorxiv.org/content/10.1101/2024.01.21.576391v2', 'https://www.biorxiv.org/content/10.1101/2024.02.02.578587v1', 'https://www.biorxiv.org/content/10.1101/2024.01.30.578060v1', 'https://www.biorxiv.org/content/10.1101/2023.05.15.540553v2', 'https://www.biorxiv.org/content/10.1101/2022.09.30.510284v2', 'https://www.biorxiv.org/content/10.1101/2024.01.25.577290v1', 'https://www.biorxiv.org/content/10.1101/2024.01.25.577285v1', 'https://www.biorxiv.org/content/10.1101/2022.12.02.518832v3', 'https://www.biorxiv.org/content/10.1101/2024.01.21.576581v1', 'https://www.biorxiv.org/content/10.1101/2024.01.22.576626v1', 'https://www.biorxiv.org/content/10.1101/2023.11.21.568164v2', 'https://www.biorxiv.org/content/10.1101/2024.01.02.573821v2', 'https://www.biorxiv.org/content/10.1101/2023.12.22.572963v1', 'https://www.biorxiv.org/content/10.1101/2023.12.17.572049v1', 'https://www.biorxiv.org/content/10.1101/2023.12.13.571553v2', 'https://www.biorxiv.org/content/10.1101/2023.12.13.571436v1', 'https://www.biorxiv.org/content/10.1101/2023.12.11.570761v1', 'https://www.biorxiv.org/content/10.1101/2023.06.02.543359v2', 'https://www.biorxiv.org/content/10.1101/2023.03.21.533599v4', 'https://www.biorxiv.org/content/10.1101/2023.12.08.570284v1', 'https://www.biorxiv.org/content/10.1101/2023.12.06.570386v1', 'https://www.biorxiv.org/content/10.1101/2023.11.28.568999v1', 'https://www.biorxiv.org/content/10.1101/2023.11.23.568413v1', 'https://www.biorxiv.org/content/10.1101/2023.11.20.567826v1', 'https://www.biorxiv.org/content/10.1101/2023.01.31.526521v2', 'https://www.biorxiv.org/content/10.1101/2023.11.12.566748v1', 'https://www.biorxiv.org/content/10.1101/2023.11.14.567069v1', 'https://www.biorxiv.org/content/10.1101/2023.11.01.565214v2', 'https://www.biorxiv.org/content/10.1101/2023.11.11.566695v1', 'https://www.biorxiv.org/content/10.1101/2023.10.31.564804v1', 'https://www.biorxiv.org/content/10.1101/2023.07.21.550101v2', 'https://www.biorxiv.org/content/10.1101/2023.10.29.564611v1', 'https://www.biorxiv.org/content/10.1101/2023.05.31.543043v4', 'https://www.biorxiv.org/content/10.1101/2023.10.26.564281v1', 'https://www.biorxiv.org/content/10.1101/2023.03.08.531760v3', 'https://www.biorxiv.org/content/10.1101/2023.10.18.563013v1', 'https://www.biorxiv.org/content/10.1101/2023.10.08.561343v1', 'https://www.biorxiv.org/content/10.1101/2023.10.06.561245v1', 'https://www.biorxiv.org/content/10.1101/2023.10.04.560829v1', 'https://www.biorxiv.org/content/10.1101/2023.09.18.558339v3', 'https://www.biorxiv.org/content/10.1101/2023.10.01.559299v1', 'https://www.biorxiv.org/content/10.1101/2023.09.04.556257v3', 'https://www.biorxiv.org/content/10.1101/2023.04.14.536701v2', 'https://www.biorxiv.org/content/10.1101/2023.09.11.557005v1', 'https://www.biorxiv.org/content/10.1101/2023.04.04.535632v2', 'https://www.biorxiv.org/content/10.1101/2023.09.07.556769v1', 'https://www.biorxiv.org/content/10.1101/2023.09.07.545205v1', 'https://www.biorxiv.org/content/10.1101/2023.06.19.545611v3', 'https://www.biorxiv.org/content/10.1101/2023.09.01.555813v1', 'https://www.biorxiv.org/content/10.1101/2023.06.07.543773v2', 'https://www.biorxiv.org/content/10.1101/2023.03.31.535143v3', 'https://www.biorxiv.org/content/10.1101/2023.08.20.551368v1', 'https://www.biorxiv.org/content/10.1101/2023.05.26.542525v2', 'https://www.biorxiv.org/content/10.1101/2023.08.04.551372v1', 'https://www.biorxiv.org/content/10.1101/2023.07.20.549842v1', 'https://www.biorxiv.org/content/10.1101/2023.07.20.549822v1', 'https://www.biorxiv.org/content/10.1101/2023.07.17.549334v2', 'https://www.biorxiv.org/content/10.1101/2023.07.19.549631v1', 'https://www.biorxiv.org/content/10.1101/2023.07.09.548288v1', 'https://www.biorxiv.org/content/10.1101/2023.07.07.548136v1', 'https://www.biorxiv.org/content/10.1101/2023.05.12.540616v3', 'https://www.biorxiv.org/content/10.1101/2023.06.25.546295v1', 'https://www.biorxiv.org/content/10.1101/2023.06.21.545875v1', 'https://www.biorxiv.org/content/10.1101/2023.06.21.545871v1', 'https://www.biorxiv.org/content/10.1101/2023.06.20.545441v1', 'https://www.biorxiv.org/content/10.1101/2023.04.11.536385v2', 'https://www.biorxiv.org/content/10.1101/2023.04.29.538821v2', 'https://www.biorxiv.org/content/10.1101/2023.06.07.544145v1', 'https://www.biorxiv.org/content/10.1101/2023.06.08.544213v1', 'https://www.biorxiv.org/content/10.1101/2023.06.02.542089v1', 'https://www.biorxiv.org/content/10.1101/2023.05.28.542320v2', 'https://www.biorxiv.org/content/10.1101/2023.05.09.539895v2', 'https://www.biorxiv.org/content/10.1101/2023.05.17.540956v1', 'https://www.biorxiv.org/content/10.1101/2023.05.08.539909v1', 'https://www.biorxiv.org/content/10.1101/2023.05.05.539552v1', 'https://www.biorxiv.org/content/10.1101/2023.05.04.539481v2', 'https://www.biorxiv.org/content/10.1101/2023.05.02.539078v1', 'https://www.biorxiv.org/content/10.1101/2023.04.21.537793v1', 'https://www.biorxiv.org/content/10.1101/2023.04.20.537700v1', 'https://www.biorxiv.org/content/10.1101/2023.04.17.537254v1', 'https://www.biorxiv.org/content/10.1101/2020.12.08.416024v3', 'https://www.biorxiv.org/content/10.1101/2022.10.06.511217v3', 'https://www.biorxiv.org/content/10.1101/2023.04.06.535797v1', 'https://www.biorxiv.org/content/10.1101/2022.10.29.514266v2', 'https://www.biorxiv.org/content/10.1101/2023.03.30.534931v1', 'https://www.biorxiv.org/content/10.1101/2023.03.28.534531v1', 'https://www.biorxiv.org/content/10.1101/2023.02.24.529942v3', 'https://www.biorxiv.org/content/10.1101/2023.03.14.532524v1', 'https://www.biorxiv.org/content/10.1101/2023.02.08.527665v1', 'https://www.biorxiv.org/content/10.1101/2023.02.27.530215v2', 'https://www.biorxiv.org/content/10.1101/2023.03.05.531147v1', 'https://www.biorxiv.org/content/10.1101/2023.02.15.528703v1', 'https://www.biorxiv.org/content/10.1101/2023.02.05.527102v1', 'https://www.biorxiv.org/content/10.1101/2023.02.01.526717v1', 'https://www.biorxiv.org/content/10.1101/2023.01.24.524926v1', 'https://www.biorxiv.org/content/10.1101/2023.01.20.524849v1', 'https://www.biorxiv.org/content/10.1101/2023.01.19.524778v1', 'https://www.biorxiv.org/content/10.1101/2023.01.11.523611v1', 'https://www.biorxiv.org/content/10.1101/2023.01.10.523267v1', 'https://www.biorxiv.org/content/10.1101/2023.01.11.523546v1', 'https://www.biorxiv.org/content/10.1101/2023.01.04.522651v1', 'https://www.biorxiv.org/content/10.1101/2022.12.31.522367v1', 'https://www.biorxiv.org/content/10.1101/2022.12.18.520013v1', 'https://www.biorxiv.org/content/10.1101/2022.12.16.520696v1', 'https://www.biorxiv.org/content/10.1101/2022.12.08.518931v3', 'https://www.biorxiv.org/content/10.1101/2022.12.08.519610v1', 'https://www.biorxiv.org/content/10.1101/2022.12.06.519410v1', 'https://www.biorxiv.org/content/10.1101/2022.11.29.518374v1', 'https://www.biorxiv.org/content/10.1101/2022.12.02.518787v1', 'https://www.biorxiv.org/content/10.1101/2022.12.01.518742v1', 'https://www.biorxiv.org/content/10.1101/2022.11.26.518030v1', 'https://www.biorxiv.org/content/10.1101/2022.11.25.517947v1', 'https://www.biorxiv.org/content/10.1101/2022.11.05.515129v1', 'https://www.biorxiv.org/content/10.1101/2022.11.04.515226v1', 'https://www.biorxiv.org/content/10.1101/2022.11.03.515020v1', 'https://www.biorxiv.org/content/10.1101/2022.11.04.515094v1', 'https://www.biorxiv.org/content/10.1101/2022.11.03.515100v1', 'https://www.biorxiv.org/content/10.1101/2022.09.23.509203v2', 'https://www.biorxiv.org/content/10.1101/2022.10.27.513881v2', 'https://www.biorxiv.org/content/10.1101/2022.10.27.513962v1', 'https://www.biorxiv.org/content/10.1101/2022.10.26.513890v2', 'https://www.biorxiv.org/content/10.1101/2022.10.24.513174v2', 'https://www.biorxiv.org/content/10.1101/2022.10.21.513181v3', 'https://www.biorxiv.org/content/10.1101/2022.10.14.512303v2', 'https://www.biorxiv.org/content/10.1101/2022.09.26.509451v2', 'https://www.biorxiv.org/content/10.1101/2022.09.22.508453v1', 'https://www.biorxiv.org/content/10.1101/2022.09.17.508368v2', 'https://www.biorxiv.org/content/10.1101/2022.09.06.505536v1', 'https://www.biorxiv.org/content/10.1101/2022.09.05.506692v1', 'https://www.biorxiv.org/content/10.1101/2022.08.30.505939v2', 'https://www.biorxiv.org/content/10.1101/2022.08.11.503648v1', 'https://www.biorxiv.org/content/10.1101/2022.08.02.502545v1', 'https://www.biorxiv.org/content/10.1101/2022.08.04.502770v1', 'https://www.biorxiv.org/content/10.1101/2022.08.01.502266v1', 'https://www.biorxiv.org/content/10.1101/2022.07.09.499321v1', 'https://www.biorxiv.org/content/10.1101/2022.07.06.499059v1', 'https://www.biorxiv.org/content/10.1101/2022.07.05.498785v1', 'https://www.biorxiv.org/content/10.1101/2022.06.30.498290v1', 'https://www.biorxiv.org/content/10.1101/2022.06.27.497795v1', 'https://www.biorxiv.org/content/10.1101/2022.06.20.496811v1', 'https://www.biorxiv.org/content/10.1101/2022.06.19.496725v1', 'https://www.biorxiv.org/content/10.1101/2022.06.08.495395v2', 'https://www.biorxiv.org/content/10.1101/2022.06.06.494894v2', 'https://www.biorxiv.org/content/10.1101/2022.05.27.493681v1', 'https://www.biorxiv.org/content/10.1101/2022.05.23.493179v1', 'https://www.biorxiv.org/content/10.1101/2022.05.24.493244v1', 'https://www.biorxiv.org/content/10.1101/2022.05.19.492613v2', 'https://www.biorxiv.org/content/10.1101/2022.05.19.492566v2', 'https://www.biorxiv.org/content/10.1101/2022.05.17.492399v3', 'https://www.biorxiv.org/content/10.1101/2022.05.11.491472v1', 'https://www.biorxiv.org/content/10.1101/2022.05.10.491293v1', 'https://www.biorxiv.org/content/10.1101/2022.05.06.490930v1', 'https://www.biorxiv.org/content/10.1101/2022.05.05.490773v2', 'https://www.biorxiv.org/content/10.1101/2022.04.25.489326v1', 'https://www.biorxiv.org/content/10.1101/2022.04.23.489213v2', 'https://www.biorxiv.org/content/10.1101/2022.03.29.486305v1', 'https://www.biorxiv.org/content/10.1101/2022.03.28.486092v1', 'https://www.biorxiv.org/content/10.1101/2022.03.25.485477v1', 'https://www.biorxiv.org/content/10.1101/2022.03.24.485682v1', 'https://www.biorxiv.org/content/10.1101/2022.03.23.485539v1', 'https://www.biorxiv.org/content/10.1101/2022.03.14.484211v2', 'https://www.biorxiv.org/content/10.1101/2022.02.23.481560v1', 'https://www.biorxiv.org/content/10.1101/2022.02.18.481102v1', 'https://www.biorxiv.org/content/10.1101/2022.02.20.480002v1', 'https://www.biorxiv.org/content/10.1101/2022.02.15.480535v1', 'https://www.biorxiv.org/content/10.1101/2022.02.11.480026v2', 'https://www.biorxiv.org/content/10.1101/2022.01.24.474473v2', 'https://www.biorxiv.org/content/10.1101/2022.01.18.476738v1', 'https://www.biorxiv.org/content/10.1101/2022.01.15.476199v2', 'https://www.biorxiv.org/content/10.1101/2022.01.13.476259v1', 'https://www.biorxiv.org/content/10.1101/2022.01.11.475790v1', 'https://www.biorxiv.org/content/10.1101/2022.01.11.475882v2', 'https://www.biorxiv.org/content/10.1101/2021.03.26.437240v3', 'https://www.biorxiv.org/content/10.1101/2021.12.22.472709v2', 'https://www.biorxiv.org/content/10.1101/2021.12.20.473562v1', 'https://www.biorxiv.org/content/10.1101/2021.12.16.473094v1', 'https://www.biorxiv.org/content/10.1101/2021.12.16.472765v1', 'https://www.biorxiv.org/content/10.1101/2021.12.15.472871v4', 'https://www.biorxiv.org/content/10.1101/2021.12.14.472718v2', 'https://www.biorxiv.org/content/10.1101/2021.03.22.436437v3', 'https://www.biorxiv.org/content/10.1101/2021.10.25.465678v2', 'https://www.biorxiv.org/content/10.1101/2021.12.03.471074v3', 'https://www.biorxiv.org/content/10.1101/2021.12.03.471027v4', 'https://www.biorxiv.org/content/10.1101/2021.11.24.469861v1', 'https://www.biorxiv.org/content/10.1101/2021.11.24.469859v2', 'https://www.biorxiv.org/content/10.1101/2021.10.08.463637v3', 'https://www.biorxiv.org/content/10.1101/2021.09.09.459700v2', 'https://www.biorxiv.org/content/10.1101/2021.11.10.467921v3', 'https://www.biorxiv.org/content/10.1101/2021.11.09.467907v2', 'https://www.biorxiv.org/content/10.1101/2021.10.29.466422v2', 'https://www.biorxiv.org/content/10.1101/2021.11.02.466900v2', 'https://www.biorxiv.org/content/10.1101/2020.06.08.139964v3', 'https://www.biorxiv.org/content/10.1101/2020.12.10.420448v2', 'https://www.biorxiv.org/content/10.1101/2021.09.29.462206v2', 'https://www.biorxiv.org/content/10.1101/2021.09.28.462151v1', 'https://www.biorxiv.org/content/10.1101/2021.09.28.462251v1', 'https://www.biorxiv.org/content/10.1101/2021.09.15.460475v2', 'https://www.biorxiv.org/content/10.1101/2021.09.14.460272v1', 'https://www.biorxiv.org/content/10.1101/2021.09.10.459728v1', 'https://www.biorxiv.org/content/10.1101/2021.09.07.459217v1', 'https://www.biorxiv.org/content/10.1101/2021.09.03.458862v1', 'https://www.biorxiv.org/content/10.1101/2021.09.02.458674v1', 'https://www.biorxiv.org/content/10.1101/2021.09.01.458518v1', 'https://www.biorxiv.org/content/10.1101/2021.08.30.458191v1', 'https://www.biorxiv.org/content/10.1101/2021.08.24.457596v1', 'https://www.biorxiv.org/content/10.1101/2021.08.26.457874v1', 'https://www.biorxiv.org/content/10.1101/2021.03.19.436173v2', 'https://www.biorxiv.org/content/10.1101/2021.08.11.456034v1', 'https://www.biorxiv.org/content/10.1101/2021.05.20.444927v2', 'https://www.biorxiv.org/content/10.1101/2021.08.05.455277v1', 'https://www.biorxiv.org/content/10.1101/2021.08.03.454900v1', 'https://www.biorxiv.org/content/10.1101/2021.03.23.436560v3', 'https://www.biorxiv.org/content/10.1101/2021.07.30.454505v2', 'https://www.biorxiv.org/content/10.1101/2021.07.23.453484v2', 'https://www.biorxiv.org/content/10.1101/2020.07.28.224584v3', 'https://www.biorxiv.org/content/10.1101/2021.07.18.452853v2', 'https://www.biorxiv.org/content/10.1101/2021.07.12.452063v1', 'https://www.biorxiv.org/content/10.1101/2021.02.04.429752v3', 'https://www.biorxiv.org/content/10.1101/2021.02.18.431773v3', 'https://www.biorxiv.org/content/10.1101/2021.07.02.450893v1', 'https://www.biorxiv.org/content/10.1101/2021.07.02.450883v1', 'https://www.biorxiv.org/content/10.1101/2021.03.16.435535v2', 'https://www.biorxiv.org/content/10.1101/2021.06.22.449514v1', 'https://www.biorxiv.org/content/10.1101/2021.06.23.449389v1', 'https://www.biorxiv.org/content/10.1101/2021.06.16.448630v1', 'https://www.biorxiv.org/content/10.1101/2021.06.14.448320v1', 'https://www.biorxiv.org/content/10.1101/2021.06.08.446897v1', 'https://www.biorxiv.org/content/10.1101/2021.06.05.447065v5', 'https://www.biorxiv.org/content/10.1101/2021.05.23.445375v1', 'https://www.biorxiv.org/content/10.1101/2021.05.17.444424v2', 'https://www.biorxiv.org/content/10.1101/2021.05.17.444412v2', 'https://www.biorxiv.org/content/10.1101/2021.05.17.444522v1', 'https://www.biorxiv.org/content/10.1101/2021.05.17.444419v1', 'https://www.biorxiv.org/content/10.1101/2021.05.12.443913v1', 'https://www.biorxiv.org/content/10.1101/2021.02.24.429166v2', 'https://www.biorxiv.org/content/10.1101/2021.05.04.442591v1', 'https://www.biorxiv.org/content/10.1101/2021.04.27.441597v2', 'https://www.biorxiv.org/content/10.1101/2021.04.21.440749v2', 'https://www.biorxiv.org/content/10.1101/2021.04.21.440844v1', 'https://www.biorxiv.org/content/10.1101/2021.04.19.440412v1', 'https://www.biorxiv.org/content/10.1101/2021.04.02.438271v3', 'https://www.biorxiv.org/content/10.1101/2020.03.29.014704v4', 'https://www.biorxiv.org/content/10.1101/2020.11.23.394700v3', 'https://www.biorxiv.org/content/10.1101/2020.12.22.423111v3', 'https://www.biorxiv.org/content/10.1101/2021.04.02.438176v1', 'https://www.biorxiv.org/content/10.1101/2020.12.20.423642v4', 'https://www.biorxiv.org/content/10.1101/2021.03.29.437575v1', 'https://www.biorxiv.org/content/10.1101/2021.03.26.437296v1', 'https://www.biorxiv.org/content/10.1101/2021.03.22.436504v1', 'https://www.biorxiv.org/content/10.1101/2020.08.07.241729v2', 'https://www.biorxiv.org/content/10.1101/2021.03.08.434372v1', 'https://www.biorxiv.org/content/10.1101/2020.07.24.219691v3', 'https://www.biorxiv.org/content/10.1101/2021.02.23.432530v2', 'https://www.biorxiv.org/content/10.1101/2021.02.26.433113v1', 'https://www.biorxiv.org/content/10.1101/2021.02.04.429837v2', 'https://www.biorxiv.org/content/10.1101/2021.02.22.432219v1', 'https://www.biorxiv.org/content/10.1101/2020.08.22.262931v2', 'https://www.biorxiv.org/content/10.1101/2020.06.16.155838v3', 'https://www.biorxiv.org/content/10.1101/2021.02.16.431561v1', 'https://www.biorxiv.org/content/10.1101/2021.02.16.431410v1', 'https://www.biorxiv.org/content/10.1101/2021.02.12.431026v1', 'https://www.biorxiv.org/content/10.1101/2021.02.11.430759v1', 'https://www.biorxiv.org/content/10.1101/2021.02.08.430279v1', 'https://www.biorxiv.org/content/10.1101/2021.02.05.429839v1', 'https://www.biorxiv.org/content/10.1101/2021.02.07.430053v1', 'https://www.biorxiv.org/content/10.1101/2021.02.04.429812v1', 'https://www.biorxiv.org/content/10.1101/2020.04.27.058388v2', 'https://www.biorxiv.org/content/10.1101/2021.02.03.429603v1', 'https://www.biorxiv.org/content/10.1101/2021.02.02.429378v1', 'https://www.biorxiv.org/content/10.1101/2021.02.02.429443v1', 'https://www.biorxiv.org/content/10.1101/2020.12.04.412486v2', 'https://www.biorxiv.org/content/10.1101/2020.10.21.349605v2', 'https://www.biorxiv.org/content/10.1101/2021.01.29.428792v1', 'https://www.biorxiv.org/content/10.1101/2021.01.25.428049v1', 'https://www.biorxiv.org/content/10.1101/2020.10.13.331058v2', 'https://www.biorxiv.org/content/10.1101/2021.01.22.427741v1', 'https://www.biorxiv.org/content/10.1101/2021.01.17.427048v1', 'https://www.biorxiv.org/content/10.1101/2021.01.12.426317v2', 'https://www.biorxiv.org/content/10.1101/2021.01.15.426843v1', 'https://www.biorxiv.org/content/10.1101/2021.01.15.426838v1', 'https://www.biorxiv.org/content/10.1101/2021.01.11.426108v1', 'https://www.biorxiv.org/content/10.1101/2021.01.08.425845v1', 'https://www.biorxiv.org/content/10.1101/2020.07.03.186692v2', 'https://www.biorxiv.org/content/10.1101/2020.12.15.422939v1', 'https://www.biorxiv.org/content/10.1101/848366v2', 'https://www.biorxiv.org/content/10.1101/2020.12.11.422246v1', 'https://www.biorxiv.org/content/10.1101/2020.11.26.400507v1', 'https://www.biorxiv.org/content/10.1101/2020.11.12.380378v3', 'https://www.biorxiv.org/content/10.1101/2020.11.17.386292v1', 'https://www.biorxiv.org/content/10.1101/2020.10.28.358903v1', 'https://www.biorxiv.org/content/10.1101/2020.10.26.356543v1', 'https://www.biorxiv.org/content/10.1101/2020.02.01.930370v3', 'https://www.biorxiv.org/content/10.1101/2020.10.19.345819v2', 'https://www.biorxiv.org/content/10.1101/2020.08.07.242032v3', 'https://www.biorxiv.org/content/10.1101/2020.10.09.333419v1', 'https://www.biorxiv.org/content/10.1101/264242v4', 'https://www.biorxiv.org/content/10.1101/548123v2', 'https://www.biorxiv.org/content/10.1101/2020.06.15.151837v2', 'https://www.biorxiv.org/content/10.1101/2020.09.24.311720v1', 'https://www.biorxiv.org/content/10.1101/2020.09.21.307363v1', 'https://www.biorxiv.org/content/10.1101/2020.09.21.293175v1', 'https://www.biorxiv.org/content/10.1101/672840v6', 'https://www.biorxiv.org/content/10.1101/2020.09.16.299495v1', 'https://www.biorxiv.org/content/10.1101/2020.05.07.078311v2', 'https://www.biorxiv.org/content/10.1101/2020.09.07.286369v2', 'https://www.biorxiv.org/content/10.1101/2020.04.27.064790v2', 'https://www.biorxiv.org/content/10.1101/2019.12.12.866210v2', 'https://www.biorxiv.org/content/10.1101/2020.08.04.234435v3', 'https://www.biorxiv.org/content/10.1101/2020.08.10.235309v3', 'https://www.biorxiv.org/content/10.1101/2020.08.07.242396v1', 'https://www.biorxiv.org/content/10.1101/2020.07.31.230938v1', 'https://www.biorxiv.org/content/10.1101/2020.07.23.216937v1', 'https://www.biorxiv.org/content/10.1101/2020.07.21.214783v1', 'https://www.biorxiv.org/content/10.1101/2020.07.21.213975v1', 'https://www.biorxiv.org/content/10.1101/2020.07.15.203513v1', 'https://www.biorxiv.org/content/10.1101/2020.07.08.194233v1', 'https://www.biorxiv.org/content/10.1101/2020.07.02.185421v1', 'https://www.biorxiv.org/content/10.1101/2020.01.12.903443v3', 'https://www.biorxiv.org/content/10.1101/2020.06.18.151886v1', 'https://www.biorxiv.org/content/10.1101/2020.02.21.959205v3', 'https://www.biorxiv.org/content/10.1101/2020.06.09.142919v1', 'https://www.biorxiv.org/content/10.1101/2020.05.21.108324v1', 'https://www.biorxiv.org/content/10.1101/2020.05.14.095265v1', 'https://www.biorxiv.org/content/10.1101/2020.05.07.082107v1', 'https://www.biorxiv.org/content/10.1101/2019.12.24.888297v2', 'https://www.biorxiv.org/content/10.1101/2020.05.05.079368v1', 'https://www.biorxiv.org/content/10.1101/2020.04.23.056317v1', 'https://www.biorxiv.org/content/10.1101/2020.04.21.050617v1', 'https://www.biorxiv.org/content/10.1101/2020.04.14.040592v1', 'https://www.biorxiv.org/content/10.1101/2020.04.11.036871v1', 'https://www.biorxiv.org/content/10.1101/2020.04.07.029389v2', 'https://www.biorxiv.org/content/10.1101/644740v2', 'https://www.biorxiv.org/content/10.1101/808378v2', 'https://www.biorxiv.org/content/10.1101/2019.12.20.884676v2', 'https://www.biorxiv.org/content/10.1101/2020.03.03.975219v4', 'https://www.biorxiv.org/content/10.1101/2020.03.14.992248v3', 'https://www.biorxiv.org/content/10.1101/2020.03.11.988014v1', 'https://www.biorxiv.org/content/10.1101/2020.03.08.982629v1', 'https://www.biorxiv.org/content/10.1101/2020.03.06.981548v1', 'https://www.biorxiv.org/content/10.1101/2020.03.03.975821v1', 'https://www.biorxiv.org/content/10.1101/678698v2', 'https://www.biorxiv.org/content/10.1101/2020.02.28.969394v1', 'https://www.biorxiv.org/content/10.1101/2020.02.24.963009v1', 'https://www.biorxiv.org/content/10.1101/2020.02.18.949735v1', 'https://www.biorxiv.org/content/10.1101/2020.02.17.953539v1', 'https://www.biorxiv.org/content/10.1101/2020.02.11.942037v1', 'https://www.biorxiv.org/content/10.1101/645721v3', 'https://www.biorxiv.org/content/10.1101/511402v2', 'https://www.biorxiv.org/content/10.1101/2019.12.20.885269v2', 'https://www.biorxiv.org/content/10.1101/2020.01.27.921817v1', 'https://www.biorxiv.org/content/10.1101/2020.01.23.917021v1', 'https://www.biorxiv.org/content/10.1101/2020.01.21.914168v1', 'https://www.biorxiv.org/content/10.1101/462788v3', 'https://www.biorxiv.org/content/10.1101/2019.12.20.882423v3', 'https://www.biorxiv.org/content/10.1101/427161v4', 'https://www.biorxiv.org/content/10.1101/2019.12.19.882142v1', 'https://www.biorxiv.org/content/10.1101/2019.12.18.881292v1', 'https://www.biorxiv.org/content/10.1101/2019.12.16.879023v1', 'https://www.biorxiv.org/content/10.1101/2019.12.13.875021v1', 'https://www.biorxiv.org/content/10.1101/863316v1', 'https://www.biorxiv.org/content/10.1101/861005v1', 'https://www.biorxiv.org/content/10.1101/804674v4', 'https://www.biorxiv.org/content/10.1101/858142v1', 'https://www.biorxiv.org/content/10.1101/850313v1', 'https://www.biorxiv.org/content/10.1101/844746v1', 'https://www.biorxiv.org/content/10.1101/822510v2', 'https://www.biorxiv.org/content/10.1101/836296v2', 'https://www.biorxiv.org/content/10.1101/812917v2', 'https://www.biorxiv.org/content/10.1101/778753v2', 'https://www.biorxiv.org/content/10.1101/831057v1', 'https://www.biorxiv.org/content/10.1101/828350v1', 'https://www.biorxiv.org/content/10.1101/706531v2', 'https://www.biorxiv.org/content/10.1101/813634v1', 'https://www.biorxiv.org/content/10.1101/668921v3', 'https://www.biorxiv.org/content/10.1101/814731v1', 'https://www.biorxiv.org/content/10.1101/810812v1', 'https://www.biorxiv.org/content/10.1101/223404v3', 'https://www.biorxiv.org/content/10.1101/794461v1', 'https://www.biorxiv.org/content/10.1101/741173v3', 'https://www.biorxiv.org/content/10.1101/790725v1', 'https://www.biorxiv.org/content/10.1101/782755v2', 'https://www.biorxiv.org/content/10.1101/767251v1', 'https://www.biorxiv.org/content/10.1101/758144v1', 'https://www.biorxiv.org/content/10.1101/626457v2', 'https://www.biorxiv.org/content/10.1101/457101v2', 'https://www.biorxiv.org/content/10.1101/733980v1', 'https://www.biorxiv.org/content/10.1101/483693v4', 'https://www.biorxiv.org/content/10.1101/695338v2', 'https://www.biorxiv.org/content/10.1101/564278v3', 'https://www.biorxiv.org/content/10.1101/723890v1', 'https://www.biorxiv.org/content/10.1101/524090v2', 'https://www.biorxiv.org/content/10.1101/704288v1', 'https://www.biorxiv.org/content/10.1101/619122v2', 'https://www.biorxiv.org/content/10.1101/694554v1', 'https://www.biorxiv.org/content/10.1101/686584v1', 'https://www.biorxiv.org/content/10.1101/504050v2', 'https://www.biorxiv.org/content/10.1101/680827v2', 'https://www.biorxiv.org/content/10.1101/656306v2', 'https://www.biorxiv.org/content/10.1101/663864v1', 'https://www.biorxiv.org/content/10.1101/416230v3', 'https://www.biorxiv.org/content/10.1101/651539v1', 'https://www.biorxiv.org/content/10.1101/651638v1', 'https://www.biorxiv.org/content/10.1101/649178v1', 'https://www.biorxiv.org/content/10.1101/342352v3', 'https://www.biorxiv.org/content/10.1101/625244v2', 'https://www.biorxiv.org/content/10.1101/342865v6', 'https://www.biorxiv.org/content/10.1101/620682v1', 'https://www.biorxiv.org/content/10.1101/287177v4', 'https://www.biorxiv.org/content/10.1101/615385v1', 'https://www.biorxiv.org/content/10.1101/601716v1', 'https://www.biorxiv.org/content/10.1101/600403v1', 'https://www.biorxiv.org/content/10.1101/587717v1', 'https://www.biorxiv.org/content/10.1101/584029v1', 'https://www.biorxiv.org/content/10.1101/305268v3', 'https://www.biorxiv.org/content/10.1101/540120v2', 'https://www.biorxiv.org/content/10.1101/569095v1', 'https://www.biorxiv.org/content/10.1101/567370v1', 'https://www.biorxiv.org/content/10.1101/560755v1', 'https://www.biorxiv.org/content/10.1101/560367v1', 'https://www.biorxiv.org/content/10.1101/560805v1', 'https://www.biorxiv.org/content/10.1101/557439v1', 'https://www.biorxiv.org/content/10.1101/493742v2', 'https://www.biorxiv.org/content/10.1101/545640v1', 'https://www.biorxiv.org/content/10.1101/542811v1', 'https://www.biorxiv.org/content/10.1101/460345v2', 'https://www.biorxiv.org/content/10.1101/533166v1', 'https://www.biorxiv.org/content/10.1101/532853v1', 'https://www.biorxiv.org/content/10.1101/523068v1', 'https://www.biorxiv.org/content/10.1101/185892v4', 'https://www.biorxiv.org/content/10.1101/511014v1', 'https://www.biorxiv.org/content/10.1101/483701v1', 'https://www.biorxiv.org/content/10.1101/339820v3', 'https://www.biorxiv.org/content/10.1101/400374v3', 'https://www.biorxiv.org/content/10.1101/393488v2', 'https://www.biorxiv.org/content/10.1101/479105v1', 'https://www.biorxiv.org/content/10.1101/477216v1', 'https://www.biorxiv.org/content/10.1101/465831v1', 'https://www.biorxiv.org/content/10.1101/464222v2', 'https://www.biorxiv.org/content/10.1101/462036v1', 'https://www.biorxiv.org/content/10.1101/459693v1', 'https://www.biorxiv.org/content/10.1101/456905v1', 'https://www.biorxiv.org/content/10.1101/447870v2', 'https://www.biorxiv.org/content/10.1101/447037v1', 'https://www.biorxiv.org/content/10.1101/443028v1', 'https://www.biorxiv.org/content/10.1101/436691v1', 'https://www.biorxiv.org/content/10.1101/438655v1', 'https://www.biorxiv.org/content/10.1101/434001v1', 'https://www.biorxiv.org/content/10.1101/252668v2', 'https://www.biorxiv.org/content/10.1101/430538v1', 'https://www.biorxiv.org/content/10.1101/311720v2', 'https://www.biorxiv.org/content/10.1101/420208v1', 'https://www.biorxiv.org/content/10.1101/413138v1', 'https://www.biorxiv.org/content/10.1101/179705v3', 'https://www.biorxiv.org/content/10.1101/409573v1', 'https://www.biorxiv.org/content/10.1101/405290v1', 'https://www.biorxiv.org/content/10.1101/398446v1', 'https://www.biorxiv.org/content/10.1101/224733v3', 'https://www.biorxiv.org/content/10.1101/390955v1', 'https://www.biorxiv.org/content/10.1101/388454v2', 'https://www.biorxiv.org/content/10.1101/387597v1', 'https://www.biorxiv.org/content/10.1101/383992v1', 'https://www.biorxiv.org/content/10.1101/378356v1', 'https://www.biorxiv.org/content/10.1101/301903v3', 'https://www.biorxiv.org/content/10.1101/373233v1', 'https://www.biorxiv.org/content/10.1101/314880v2', 'https://www.biorxiv.org/content/10.1101/159152v2', 'https://www.biorxiv.org/content/10.1101/354456v1', 'https://www.biorxiv.org/content/10.1101/354241v1', 'https://www.biorxiv.org/content/10.1101/344093v1', 'https://www.biorxiv.org/content/10.1101/249482v2', 'https://www.biorxiv.org/content/10.1101/338194v1', 'https://www.biorxiv.org/content/10.1101/331215v1', 'https://www.biorxiv.org/content/10.1101/330472v1', 'https://www.biorxiv.org/content/10.1101/328856v1', 'https://www.biorxiv.org/content/10.1101/192112v3', 'https://www.biorxiv.org/content/10.1101/317552v1', 'https://www.biorxiv.org/content/10.1101/308833v1', 'https://www.biorxiv.org/content/10.1101/300699v1', 'https://www.biorxiv.org/content/10.1101/297754v1', 'https://www.biorxiv.org/content/10.1101/233148v2', 'https://www.biorxiv.org/content/10.1101/194530v4', 'https://www.biorxiv.org/content/10.1101/236711v2', 'https://www.biorxiv.org/content/10.1101/283101v2', 'https://www.biorxiv.org/content/10.1101/281360v1', 'https://www.biorxiv.org/content/10.1101/203398v2', 'https://www.biorxiv.org/content/10.1101/261586v1', 'https://www.biorxiv.org/content/10.1101/277400v2', 'https://www.biorxiv.org/content/10.1101/266197v1', 'https://www.biorxiv.org/content/10.1101/259044v1', 'https://www.biorxiv.org/content/10.1101/254987v1', 'https://www.biorxiv.org/content/10.1101/253492v1', 'https://www.biorxiv.org/content/10.1101/233460v1', 'https://www.biorxiv.org/content/10.1101/233080v1', 'https://www.biorxiv.org/content/10.1101/226720v1', 'https://www.biorxiv.org/content/10.1101/223701v1', 'https://www.biorxiv.org/content/10.1101/177253v2', 'https://www.biorxiv.org/content/10.1101/161513v3', 'https://www.biorxiv.org/content/10.1101/219352v1', 'https://www.biorxiv.org/content/10.1101/216077v1', 'https://www.biorxiv.org/content/10.1101/215483v1', 'https://www.biorxiv.org/content/10.1101/212506v1', 'https://www.biorxiv.org/content/10.1101/200477v2', 'https://www.biorxiv.org/content/10.1101/197020v1', 'https://www.biorxiv.org/content/10.1101/179960v2', 'https://www.biorxiv.org/content/10.1101/152983v2', 'https://www.biorxiv.org/content/10.1101/158204v2', 'https://www.biorxiv.org/content/10.1101/163808v1', 'https://www.biorxiv.org/content/10.1101/160614v1', 'https://www.biorxiv.org/content/10.1101/155903v1', 'https://www.biorxiv.org/content/10.1101/151670v1', 'https://www.biorxiv.org/content/10.1101/149948v3', 'https://www.biorxiv.org/content/10.1101/149468v2', 'https://www.biorxiv.org/content/10.1101/148932v1', 'https://www.biorxiv.org/content/10.1101/147777v2', 'https://www.biorxiv.org/content/10.1101/142711v1', 'https://www.biorxiv.org/content/10.1101/093096v2', 'https://www.biorxiv.org/content/10.1101/136556v3', 'https://www.biorxiv.org/content/10.1101/132084v1', 'https://www.biorxiv.org/content/10.1101/116681v2', 'https://www.biorxiv.org/content/10.1101/127829v2', 'https://www.biorxiv.org/content/10.1101/125542v1', 'https://www.biorxiv.org/content/10.1101/096719v2', 'https://www.biorxiv.org/content/10.1101/125534v1', 'https://www.biorxiv.org/content/10.1101/125229v1', 'https://www.biorxiv.org/content/10.1101/124941v1', 'https://www.biorxiv.org/content/10.1101/123612v1', 'https://www.biorxiv.org/content/10.1101/122077v1', 'https://www.biorxiv.org/content/10.1101/121160v5', 'https://www.biorxiv.org/content/10.1101/120154v2', 'https://www.biorxiv.org/content/10.1101/117838v1', 'https://www.biorxiv.org/content/10.1101/101816v1', 'https://www.biorxiv.org/content/10.1101/116293v1', 'https://www.biorxiv.org/content/10.1101/113563v1', 'https://www.biorxiv.org/content/10.1101/073445v2', 'https://www.biorxiv.org/content/10.1101/108100v2', 'https://www.biorxiv.org/content/10.1101/106989v1', 'https://www.biorxiv.org/content/10.1101/103549v1', 'https://www.biorxiv.org/content/10.1101/101576v2', 'https://www.biorxiv.org/content/10.1101/101493v1', 'https://www.biorxiv.org/content/10.1101/099069v1', 'https://www.biorxiv.org/content/10.1101/098335v1', 'https://www.biorxiv.org/content/10.1101/071282v2', 'https://www.biorxiv.org/content/10.1101/056317v2', 'https://www.biorxiv.org/content/10.1101/094235v1', 'https://www.biorxiv.org/content/10.1101/081786v2', 'https://www.biorxiv.org/content/10.1101/087759v1', 'https://www.biorxiv.org/content/10.1101/086504v3', 'https://www.biorxiv.org/content/10.1101/085126v1', 'https://www.biorxiv.org/content/10.1101/084707v1', 'https://www.biorxiv.org/content/10.1101/081885v1', 'https://www.biorxiv.org/content/10.1101/076067v1', 'https://www.biorxiv.org/content/10.1101/074807v1', 'https://www.biorxiv.org/content/10.1101/048983v5', 'https://www.biorxiv.org/content/10.1101/069922v1', 'https://www.biorxiv.org/content/10.1101/028134v3', 'https://www.biorxiv.org/content/10.1101/067215v1', 'https://www.biorxiv.org/content/10.1101/063206v1', 'https://www.biorxiv.org/content/10.1101/059410v1', 'https://www.biorxiv.org/content/10.1101/057448v2', 'https://www.biorxiv.org/content/10.1101/056846v1', 'https://www.biorxiv.org/content/10.1101/056085v1', 'https://www.biorxiv.org/content/10.1101/049213v1', 'https://www.biorxiv.org/content/10.1101/055020v1', 'https://www.biorxiv.org/content/10.1101/054783v1', 'https://www.biorxiv.org/content/10.1101/054635v1', 'https://www.biorxiv.org/content/10.1101/048595v2', 'https://www.biorxiv.org/content/10.1101/048413v1', 'https://www.biorxiv.org/content/10.1101/048033v1', 'https://www.biorxiv.org/content/10.1101/044354v1', 'https://www.biorxiv.org/content/10.1101/043042v2', 'https://www.biorxiv.org/content/10.1101/038463v3', 'https://www.biorxiv.org/content/10.1101/040261v1', 'https://www.biorxiv.org/content/10.1101/039495v1', 'https://www.biorxiv.org/content/10.1101/039123v1', 'https://www.biorxiv.org/content/10.1101/038562v1', 'https://www.biorxiv.org/content/10.1101/036954v1', 'https://www.biorxiv.org/content/10.1101/035238v1', 'https://www.biorxiv.org/content/10.1101/029660v2', 'https://www.biorxiv.org/content/10.1101/032151v1', 'https://www.biorxiv.org/content/10.1101/027318v2', 'https://www.biorxiv.org/content/10.1101/030270v2', 'https://www.biorxiv.org/content/10.1101/027904v2', 'https://www.biorxiv.org/content/10.1101/025254v2', 'https://www.biorxiv.org/content/10.1101/025148v1', 'https://www.biorxiv.org/content/10.1101/022897v1', 'https://www.biorxiv.org/content/10.1101/020024v1', 'https://www.biorxiv.org/content/10.1101/018978v1', 'https://www.biorxiv.org/content/10.1101/018333v1', 'https://www.biorxiv.org/content/10.1101/014118v1', 'https://www.biorxiv.org/content/10.1101/013425v1', 'https://www.biorxiv.org/content/10.1101/012112v2', 'https://www.biorxiv.org/content/10.1101/010769v1', 'https://www.biorxiv.org/content/10.1101/004507v2', 'https://www.biorxiv.org/content/10.1101/007476v2', 'https://www.biorxiv.org/content/10.1101/009225v1', 'https://www.biorxiv.org/content/10.1101/007757v1', 'https://www.biorxiv.org/content/10.1101/005579v2', 'https://www.biorxiv.org/content/10.1101/007666v1', 'https://www.biorxiv.org/content/10.1101/006627v2', 'https://www.biorxiv.org/content/10.1101/001834v6', 'https://www.biorxiv.org/content/10.1101/003962v2', 'https://www.biorxiv.org/content/10.1101/002469v1']\n",
      "Downloading papers...\n",
      "Downloaded: papers/2024.07.30.605897v1.pdf\n",
      "HTML file saved to html/papers/2024.07.30.605897v1.pdf\n",
      "Downloaded: papers/2023.09.11.557209v2.pdf\n",
      "HTML file saved to html/papers/2023.09.11.557209v2.pdf\n",
      "Downloaded: papers/2024.07.26.605365v1.pdf\n",
      "HTML file saved to html/papers/2024.07.26.605365v1.pdf\n",
      "Downloaded: papers/2024.07.22.604635v1.pdf\n",
      "HTML file saved to html/papers/2024.07.22.604635v1.pdf\n",
      "Downloaded: papers/2024.06.03.597150v2.pdf\n",
      "HTML file saved to html/papers/2024.06.03.597150v2.pdf\n",
      "Downloaded: papers/2024.02.19.579948v3.pdf\n",
      "HTML file saved to html/papers/2024.02.19.579948v3.pdf\n",
      "Downloaded: papers/2024.07.22.603478v1.pdf\n",
      "HTML file saved to html/papers/2024.07.22.603478v1.pdf\n",
      "Downloaded: papers/2024.07.10.602853v2.pdf\n",
      "HTML file saved to html/papers/2024.07.10.602853v2.pdf\n",
      "Downloaded: papers/2024.07.02.601807v2.pdf\n",
      "HTML file saved to html/papers/2024.07.02.601807v2.pdf\n",
      "Downloaded: papers/2024.07.09.602727v1.pdf\n",
      "HTML file saved to html/papers/2024.07.09.602727v1.pdf\n",
      "Downloaded: papers/2024.07.09.602638v1.pdf\n",
      "HTML file saved to html/papers/2024.07.09.602638v1.pdf\n",
      "Downloaded: papers/2023.10.20.563230v3.pdf\n",
      "HTML file saved to html/papers/2023.10.20.563230v3.pdf\n",
      "Downloaded: papers/2024.07.08.602537v1.pdf\n",
      "HTML file saved to html/papers/2024.07.08.602537v1.pdf\n",
      "Downloaded: papers/2024.07.05.602100v1.pdf\n",
      "HTML file saved to html/papers/2024.07.05.602100v1.pdf\n",
      "Downloaded: papers/2024.07.02.600549v1.pdf\n",
      "HTML file saved to html/papers/2024.07.02.600549v1.pdf\n",
      "Downloaded: papers/2023.08.21.553699v3.pdf\n",
      "HTML file saved to html/papers/2023.08.21.553699v3.pdf\n",
      "Downloaded: papers/2024.07.01.601500v1.pdf\n",
      "HTML file saved to html/papers/2024.07.01.601500v1.pdf\n",
      "Downloaded: papers/2024.06.14.599082v2.pdf\n",
      "HTML file saved to html/papers/2024.06.14.599082v2.pdf\n",
      "Downloaded: papers/2024.03.06.583483v2.pdf\n",
      "HTML file saved to html/papers/2024.03.06.583483v2.pdf\n",
      "Downloaded: papers/2024.06.11.598241v1.pdf\n",
      "HTML file saved to html/papers/2024.06.11.598241v1.pdf\n",
      "Downloaded: papers/2024.06.10.598220v1.pdf\n",
      "HTML file saved to html/papers/2024.06.10.598220v1.pdf\n",
      "Downloaded: papers/2024.05.23.595509v2.pdf\n",
      "HTML file saved to html/papers/2024.05.23.595509v2.pdf\n",
      "Downloaded: papers/2020.10.01.322164v4.pdf\n",
      "HTML file saved to html/papers/2020.10.01.322164v4.pdf\n",
      "Downloaded: papers/2024.04.22.590609v2.pdf\n",
      "HTML file saved to html/papers/2024.04.22.590609v2.pdf\n",
      "Downloaded: papers/2024.05.29.596541v1.pdf\n",
      "HTML file saved to html/papers/2024.05.29.596541v1.pdf\n",
      "Downloaded: papers/2023.11.21.568151v5.pdf\n",
      "HTML file saved to html/papers/2023.11.21.568151v5.pdf\n",
      "Downloaded: papers/2024.05.24.595829v1.pdf\n",
      "HTML file saved to html/papers/2024.05.24.595829v1.pdf\n",
      "Downloaded: papers/2024.05.24.595489v1.pdf\n",
      "HTML file saved to html/papers/2024.05.24.595489v1.pdf\n",
      "Downloaded: papers/2024.05.20.595060v1.pdf\n",
      "HTML file saved to html/papers/2024.05.20.595060v1.pdf\n",
      "Downloaded: papers/2024.05.17.594728v1.pdf\n",
      "HTML file saved to html/papers/2024.05.17.594728v1.pdf\n",
      "Downloaded: papers/2024.05.17.594646v3.pdf\n",
      "HTML file saved to html/papers/2024.05.17.594646v3.pdf\n",
      "Downloaded: papers/2024.05.15.594184v1.pdf\n",
      "HTML file saved to html/papers/2024.05.15.594184v1.pdf\n",
      "Downloaded: papers/2023.11.30.568362v2.pdf\n",
      "HTML file saved to html/papers/2023.11.30.568362v2.pdf\n",
      "Downloaded: papers/2024.05.13.593871v1.pdf\n",
      "HTML file saved to html/papers/2024.05.13.593871v1.pdf\n",
      "Downloaded: papers/2023.07.26.550757v2.pdf\n",
      "HTML file saved to html/papers/2023.07.26.550757v2.pdf\n",
      "Downloaded: papers/2023.12.01.569581v2.pdf\n",
      "HTML file saved to html/papers/2023.12.01.569581v2.pdf\n",
      "Downloaded: papers/2024.05.13.591068v1.pdf\n",
      "HTML file saved to html/papers/2024.05.13.591068v1.pdf\n",
      "Downloaded: papers/2023.04.15.536996v3.pdf\n",
      "HTML file saved to html/papers/2023.04.15.536996v3.pdf\n",
      "Downloaded: papers/2024.03.21.586149v2.pdf\n",
      "HTML file saved to html/papers/2024.03.21.586149v2.pdf\n",
      "Downloaded: papers/2024.04.23.590803v1.pdf\n",
      "HTML file saved to html/papers/2024.04.23.590803v1.pdf\n",
      "Downloaded: papers/2023.11.03.565466v2.pdf\n",
      "HTML file saved to html/papers/2023.11.03.565466v2.pdf\n",
      "Downloaded: papers/2024.04.18.590093v1.pdf\n",
      "HTML file saved to html/papers/2024.04.18.590093v1.pdf\n",
      "Downloaded: papers/2023.01.20.524991v5.pdf\n",
      "HTML file saved to html/papers/2023.01.20.524991v5.pdf\n",
      "Downloaded: papers/2024.04.05.588231v1.pdf\n",
      "HTML file saved to html/papers/2024.04.05.588231v1.pdf\n",
      "Downloaded: papers/2024.03.29.587318v1.pdf\n",
      "HTML file saved to html/papers/2024.03.29.587318v1.pdf\n",
      "Downloaded: papers/2024.03.25.586631v1.pdf\n",
      "HTML file saved to html/papers/2024.03.25.586631v1.pdf\n",
      "Downloaded: papers/2022.11.04.514718v2.pdf\n",
      "HTML file saved to html/papers/2022.11.04.514718v2.pdf\n",
      "Downloaded: papers/2024.03.25.586701v1.pdf\n",
      "HTML file saved to html/papers/2024.03.25.586701v1.pdf\n",
      "Downloaded: papers/2024.01.29.577700v2.pdf\n",
      "HTML file saved to html/papers/2024.01.29.577700v2.pdf\n",
      "Downloaded: papers/2024.03.18.585512v1.pdf\n",
      "HTML file saved to html/papers/2024.03.18.585512v1.pdf\n",
      "Downloaded: papers/2024.03.17.585367v1.pdf\n",
      "HTML file saved to html/papers/2024.03.17.585367v1.pdf\n",
      "Downloaded: papers/2024.03.17.585420v1.pdf\n",
      "HTML file saved to html/papers/2024.03.17.585420v1.pdf\n",
      "Downloaded: papers/2024.03.13.584790v1.pdf\n",
      "HTML file saved to html/papers/2024.03.13.584790v1.pdf\n",
      "Downloaded: papers/2024.03.07.583855v3.pdf\n",
      "HTML file saved to html/papers/2024.03.07.583855v3.pdf\n",
      "Downloaded: papers/2024.03.11.584353v1.pdf\n",
      "HTML file saved to html/papers/2024.03.11.584353v1.pdf\n",
      "Downloaded: papers/2022.06.22.497172v2.pdf\n",
      "HTML file saved to html/papers/2022.06.22.497172v2.pdf\n",
      "Downloaded: papers/2024.03.06.582860v1.pdf\n",
      "HTML file saved to html/papers/2024.03.06.582860v1.pdf\n",
      "Downloaded: papers/2024.02.27.581927v1.pdf\n",
      "HTML file saved to html/papers/2024.02.27.581927v1.pdf\n",
      "Downloaded: papers/2024.02.25.581922v1.pdf\n",
      "HTML file saved to html/papers/2024.02.25.581922v1.pdf\n",
      "Downloaded: papers/2024.02.19.580943v1.pdf\n",
      "HTML file saved to html/papers/2024.02.19.580943v1.pdf\n",
      "Downloaded: papers/2024.02.15.580441v1.pdf\n",
      "HTML file saved to html/papers/2024.02.15.580441v1.pdf\n",
      "Downloaded: papers/2024.02.18.580171v1.pdf\n",
      "HTML file saved to html/papers/2024.02.18.580171v1.pdf\n",
      "Downloaded: papers/2023.11.15.566493v2.pdf\n",
      "HTML file saved to html/papers/2023.11.15.566493v2.pdf\n",
      "Downloaded: papers/2023.09.13.557552v2.pdf\n",
      "HTML file saved to html/papers/2023.09.13.557552v2.pdf\n",
      "Downloaded: papers/2024.02.11.579857v1.pdf\n",
      "HTML file saved to html/papers/2024.02.11.579857v1.pdf\n",
      "Downloaded: papers/2022.12.13.520286v2.pdf\n",
      "HTML file saved to html/papers/2022.12.13.520286v2.pdf\n",
      "Downloaded: papers/2024.01.21.576391v2.pdf\n",
      "HTML file saved to html/papers/2024.01.21.576391v2.pdf\n",
      "Downloaded: papers/2024.02.02.578587v1.pdf\n",
      "HTML file saved to html/papers/2024.02.02.578587v1.pdf\n",
      "Downloaded: papers/2024.01.30.578060v1.pdf\n",
      "HTML file saved to html/papers/2024.01.30.578060v1.pdf\n",
      "Downloaded: papers/2023.05.15.540553v2.pdf\n",
      "HTML file saved to html/papers/2023.05.15.540553v2.pdf\n",
      "Downloaded: papers/2022.09.30.510284v2.pdf\n",
      "HTML file saved to html/papers/2022.09.30.510284v2.pdf\n",
      "Downloaded: papers/2024.01.25.577290v1.pdf\n",
      "HTML file saved to html/papers/2024.01.25.577290v1.pdf\n",
      "Downloaded: papers/2024.01.25.577285v1.pdf\n",
      "HTML file saved to html/papers/2024.01.25.577285v1.pdf\n",
      "Downloaded: papers/2022.12.02.518832v3.pdf\n",
      "HTML file saved to html/papers/2022.12.02.518832v3.pdf\n",
      "Downloaded: papers/2024.01.21.576581v1.pdf\n",
      "HTML file saved to html/papers/2024.01.21.576581v1.pdf\n",
      "Downloaded: papers/2024.01.22.576626v1.pdf\n",
      "HTML file saved to html/papers/2024.01.22.576626v1.pdf\n",
      "Downloaded: papers/2023.11.21.568164v2.pdf\n",
      "HTML file saved to html/papers/2023.11.21.568164v2.pdf\n",
      "Downloaded: papers/2024.01.02.573821v2.pdf\n",
      "HTML file saved to html/papers/2024.01.02.573821v2.pdf\n",
      "Downloaded: papers/2023.12.22.572963v1.pdf\n",
      "HTML file saved to html/papers/2023.12.22.572963v1.pdf\n",
      "Downloaded: papers/2023.12.17.572049v1.pdf\n",
      "HTML file saved to html/papers/2023.12.17.572049v1.pdf\n",
      "Downloaded: papers/2023.12.13.571553v2.pdf\n",
      "HTML file saved to html/papers/2023.12.13.571553v2.pdf\n",
      "Downloaded: papers/2023.12.13.571436v1.pdf\n",
      "HTML file saved to html/papers/2023.12.13.571436v1.pdf\n",
      "Downloaded: papers/2023.12.11.570761v1.pdf\n",
      "HTML file saved to html/papers/2023.12.11.570761v1.pdf\n",
      "Downloaded: papers/2023.06.02.543359v2.pdf\n",
      "HTML file saved to html/papers/2023.06.02.543359v2.pdf\n",
      "Downloaded: papers/2023.03.21.533599v4.pdf\n",
      "HTML file saved to html/papers/2023.03.21.533599v4.pdf\n",
      "Downloaded: papers/2023.12.08.570284v1.pdf\n",
      "HTML file saved to html/papers/2023.12.08.570284v1.pdf\n",
      "Downloaded: papers/2023.12.06.570386v1.pdf\n",
      "HTML file saved to html/papers/2023.12.06.570386v1.pdf\n",
      "Downloaded: papers/2023.11.28.568999v1.pdf\n",
      "HTML file saved to html/papers/2023.11.28.568999v1.pdf\n",
      "Downloaded: papers/2023.11.23.568413v1.pdf\n",
      "HTML file saved to html/papers/2023.11.23.568413v1.pdf\n",
      "Downloaded: papers/2023.11.20.567826v1.pdf\n",
      "HTML file saved to html/papers/2023.11.20.567826v1.pdf\n",
      "Downloaded: papers/2023.01.31.526521v2.pdf\n",
      "HTML file saved to html/papers/2023.01.31.526521v2.pdf\n",
      "Downloaded: papers/2023.11.12.566748v1.pdf\n",
      "HTML file saved to html/papers/2023.11.12.566748v1.pdf\n",
      "Downloaded: papers/2023.11.14.567069v1.pdf\n",
      "HTML file saved to html/papers/2023.11.14.567069v1.pdf\n",
      "Downloaded: papers/2023.11.01.565214v2.pdf\n",
      "HTML file saved to html/papers/2023.11.01.565214v2.pdf\n",
      "Downloaded: papers/2023.11.11.566695v1.pdf\n",
      "HTML file saved to html/papers/2023.11.11.566695v1.pdf\n",
      "Downloaded: papers/2023.10.31.564804v1.pdf\n",
      "HTML file saved to html/papers/2023.10.31.564804v1.pdf\n",
      "Downloaded: papers/2023.07.21.550101v2.pdf\n",
      "HTML file saved to html/papers/2023.07.21.550101v2.pdf\n",
      "Downloaded: papers/2023.10.29.564611v1.pdf\n",
      "HTML file saved to html/papers/2023.10.29.564611v1.pdf\n",
      "Downloaded: papers/2023.05.31.543043v4.pdf\n",
      "HTML file saved to html/papers/2023.05.31.543043v4.pdf\n",
      "Downloaded: papers/2023.10.26.564281v1.pdf\n",
      "HTML file saved to html/papers/2023.10.26.564281v1.pdf\n",
      "Downloaded: papers/2023.03.08.531760v3.pdf\n",
      "HTML file saved to html/papers/2023.03.08.531760v3.pdf\n",
      "Downloaded: papers/2023.10.18.563013v1.pdf\n",
      "HTML file saved to html/papers/2023.10.18.563013v1.pdf\n",
      "Downloaded: papers/2023.10.08.561343v1.pdf\n",
      "HTML file saved to html/papers/2023.10.08.561343v1.pdf\n",
      "Downloaded: papers/2023.10.06.561245v1.pdf\n",
      "HTML file saved to html/papers/2023.10.06.561245v1.pdf\n",
      "Downloaded: papers/2023.09.18.558339v3.pdf\n",
      "HTML file saved to html/papers/2023.09.18.558339v3.pdf\n",
      "Downloaded: papers/2023.10.01.559299v1.pdf\n",
      "HTML file saved to html/papers/2023.10.01.559299v1.pdf\n",
      "Downloaded: papers/2023.09.04.556257v3.pdf\n",
      "HTML file saved to html/papers/2023.09.04.556257v3.pdf\n",
      "Downloaded: papers/2023.04.14.536701v2.pdf\n",
      "HTML file saved to html/papers/2023.04.14.536701v2.pdf\n",
      "Downloaded: papers/2023.09.11.557005v1.pdf\n",
      "HTML file saved to html/papers/2023.09.11.557005v1.pdf\n",
      "Downloaded: papers/2023.04.04.535632v2.pdf\n",
      "HTML file saved to html/papers/2023.04.04.535632v2.pdf\n",
      "Downloaded: papers/2023.09.07.556769v1.pdf\n",
      "HTML file saved to html/papers/2023.09.07.556769v1.pdf\n",
      "Downloaded: papers/2023.09.07.545205v1.pdf\n",
      "HTML file saved to html/papers/2023.09.07.545205v1.pdf\n",
      "Downloaded: papers/2023.06.19.545611v3.pdf\n",
      "HTML file saved to html/papers/2023.06.19.545611v3.pdf\n",
      "Downloaded: papers/2023.09.01.555813v1.pdf\n",
      "HTML file saved to html/papers/2023.09.01.555813v1.pdf\n",
      "Downloaded: papers/2023.06.07.543773v2.pdf\n",
      "HTML file saved to html/papers/2023.06.07.543773v2.pdf\n",
      "Downloaded: papers/2023.03.31.535143v3.pdf\n",
      "HTML file saved to html/papers/2023.03.31.535143v3.pdf\n",
      "Downloaded: papers/2023.08.20.551368v1.pdf\n",
      "HTML file saved to html/papers/2023.08.20.551368v1.pdf\n",
      "Downloaded: papers/2023.05.26.542525v2.pdf\n",
      "HTML file saved to html/papers/2023.05.26.542525v2.pdf\n",
      "Downloaded: papers/2023.08.04.551372v1.pdf\n",
      "HTML file saved to html/papers/2023.08.04.551372v1.pdf\n",
      "Downloaded: papers/2023.07.20.549842v1.pdf\n",
      "HTML file saved to html/papers/2023.07.20.549842v1.pdf\n",
      "Downloaded: papers/2023.07.20.549822v1.pdf\n",
      "HTML file saved to html/papers/2023.07.20.549822v1.pdf\n",
      "Downloaded: papers/2023.07.17.549334v2.pdf\n",
      "HTML file saved to html/papers/2023.07.17.549334v2.pdf\n",
      "Downloaded: papers/2023.07.19.549631v1.pdf\n",
      "HTML file saved to html/papers/2023.07.19.549631v1.pdf\n",
      "Downloaded: papers/2023.07.09.548288v1.pdf\n",
      "HTML file saved to html/papers/2023.07.09.548288v1.pdf\n",
      "Downloaded: papers/2023.07.07.548136v1.pdf\n",
      "HTML file saved to html/papers/2023.07.07.548136v1.pdf\n",
      "Downloaded: papers/2023.05.12.540616v3.pdf\n",
      "HTML file saved to html/papers/2023.05.12.540616v3.pdf\n",
      "Downloaded: papers/2023.06.25.546295v1.pdf\n",
      "HTML file saved to html/papers/2023.06.25.546295v1.pdf\n",
      "Downloaded: papers/2023.06.21.545875v1.pdf\n",
      "HTML file saved to html/papers/2023.06.21.545875v1.pdf\n",
      "Downloaded: papers/2023.06.21.545871v1.pdf\n",
      "HTML file saved to html/papers/2023.06.21.545871v1.pdf\n",
      "Downloaded: papers/2023.06.20.545441v1.pdf\n",
      "HTML file saved to html/papers/2023.06.20.545441v1.pdf\n",
      "Downloaded: papers/2023.04.11.536385v2.pdf\n",
      "HTML file saved to html/papers/2023.04.11.536385v2.pdf\n",
      "Downloaded: papers/2023.04.29.538821v2.pdf\n",
      "HTML file saved to html/papers/2023.04.29.538821v2.pdf\n",
      "Downloaded: papers/2023.06.07.544145v1.pdf\n",
      "HTML file saved to html/papers/2023.06.07.544145v1.pdf\n",
      "Downloaded: papers/2023.06.08.544213v1.pdf\n",
      "HTML file saved to html/papers/2023.06.08.544213v1.pdf\n",
      "Downloaded: papers/2023.06.02.542089v1.pdf\n",
      "HTML file saved to html/papers/2023.06.02.542089v1.pdf\n",
      "Downloaded: papers/2023.05.28.542320v2.pdf\n",
      "HTML file saved to html/papers/2023.05.28.542320v2.pdf\n",
      "Downloaded: papers/2023.05.09.539895v2.pdf\n",
      "HTML file saved to html/papers/2023.05.09.539895v2.pdf\n",
      "Downloaded: papers/2023.05.17.540956v1.pdf\n",
      "HTML file saved to html/papers/2023.05.17.540956v1.pdf\n",
      "Downloaded: papers/2023.05.08.539909v1.pdf\n",
      "HTML file saved to html/papers/2023.05.08.539909v1.pdf\n",
      "Downloaded: papers/2023.05.05.539552v1.pdf\n",
      "HTML file saved to html/papers/2023.05.05.539552v1.pdf\n",
      "Downloaded: papers/2023.05.04.539481v2.pdf\n",
      "HTML file saved to html/papers/2023.05.04.539481v2.pdf\n",
      "Downloaded: papers/2023.05.02.539078v1.pdf\n",
      "HTML file saved to html/papers/2023.05.02.539078v1.pdf\n",
      "Downloaded: papers/2023.04.21.537793v1.pdf\n",
      "HTML file saved to html/papers/2023.04.21.537793v1.pdf\n",
      "Downloaded: papers/2023.04.20.537700v1.pdf\n",
      "HTML file saved to html/papers/2023.04.20.537700v1.pdf\n",
      "Downloaded: papers/2023.04.17.537254v1.pdf\n",
      "HTML file saved to html/papers/2023.04.17.537254v1.pdf\n",
      "Downloaded: papers/2020.12.08.416024v3.pdf\n",
      "HTML file saved to html/papers/2020.12.08.416024v3.pdf\n",
      "Downloaded: papers/2022.10.06.511217v3.pdf\n",
      "HTML file saved to html/papers/2022.10.06.511217v3.pdf\n",
      "Downloaded: papers/2023.04.06.535797v1.pdf\n",
      "HTML file saved to html/papers/2023.04.06.535797v1.pdf\n",
      "Downloaded: papers/2022.10.29.514266v2.pdf\n",
      "HTML file saved to html/papers/2022.10.29.514266v2.pdf\n",
      "Downloaded: papers/2023.03.30.534931v1.pdf\n",
      "HTML file saved to html/papers/2023.03.30.534931v1.pdf\n",
      "Downloaded: papers/2023.03.28.534531v1.pdf\n",
      "HTML file saved to html/papers/2023.03.28.534531v1.pdf\n",
      "Downloaded: papers/2023.02.24.529942v3.pdf\n",
      "HTML file saved to html/papers/2023.02.24.529942v3.pdf\n",
      "Downloaded: papers/2023.03.14.532524v1.pdf\n",
      "HTML file saved to html/papers/2023.03.14.532524v1.pdf\n",
      "Downloaded: papers/2023.02.08.527665v1.pdf\n",
      "HTML file saved to html/papers/2023.02.08.527665v1.pdf\n",
      "Downloaded: papers/2023.02.27.530215v2.pdf\n",
      "HTML file saved to html/papers/2023.02.27.530215v2.pdf\n",
      "Downloaded: papers/2023.03.05.531147v1.pdf\n",
      "HTML file saved to html/papers/2023.03.05.531147v1.pdf\n",
      "Downloaded: papers/2023.02.15.528703v1.pdf\n",
      "HTML file saved to html/papers/2023.02.15.528703v1.pdf\n",
      "Downloaded: papers/2023.02.05.527102v1.pdf\n",
      "HTML file saved to html/papers/2023.02.05.527102v1.pdf\n",
      "Downloaded: papers/2023.02.01.526717v1.pdf\n",
      "HTML file saved to html/papers/2023.02.01.526717v1.pdf\n",
      "Downloaded: papers/2023.01.24.524926v1.pdf\n",
      "HTML file saved to html/papers/2023.01.24.524926v1.pdf\n",
      "Downloaded: papers/2023.01.20.524849v1.pdf\n",
      "HTML file saved to html/papers/2023.01.20.524849v1.pdf\n",
      "Downloaded: papers/2023.01.19.524778v1.pdf\n",
      "HTML file saved to html/papers/2023.01.19.524778v1.pdf\n",
      "Downloaded: papers/2023.01.11.523611v1.pdf\n",
      "HTML file saved to html/papers/2023.01.11.523611v1.pdf\n",
      "Downloaded: papers/2023.01.10.523267v1.pdf\n",
      "HTML file saved to html/papers/2023.01.10.523267v1.pdf\n",
      "Downloaded: papers/2023.01.11.523546v1.pdf\n",
      "HTML file saved to html/papers/2023.01.11.523546v1.pdf\n",
      "Downloaded: papers/2023.01.04.522651v1.pdf\n",
      "HTML file saved to html/papers/2023.01.04.522651v1.pdf\n",
      "Downloaded: papers/2022.12.31.522367v1.pdf\n",
      "HTML file saved to html/papers/2022.12.31.522367v1.pdf\n",
      "Downloaded: papers/2022.12.18.520013v1.pdf\n",
      "HTML file saved to html/papers/2022.12.18.520013v1.pdf\n",
      "Downloaded: papers/2022.12.16.520696v1.pdf\n",
      "HTML file saved to html/papers/2022.12.16.520696v1.pdf\n",
      "Downloaded: papers/2022.12.08.518931v3.pdf\n",
      "HTML file saved to html/papers/2022.12.08.518931v3.pdf\n",
      "Downloaded: papers/2022.12.08.519610v1.pdf\n",
      "HTML file saved to html/papers/2022.12.08.519610v1.pdf\n",
      "Downloaded: papers/2022.12.06.519410v1.pdf\n",
      "HTML file saved to html/papers/2022.12.06.519410v1.pdf\n",
      "Downloaded: papers/2022.11.29.518374v1.pdf\n",
      "HTML file saved to html/papers/2022.11.29.518374v1.pdf\n",
      "Downloaded: papers/2022.12.02.518787v1.pdf\n",
      "HTML file saved to html/papers/2022.12.02.518787v1.pdf\n",
      "Downloaded: papers/2022.12.01.518742v1.pdf\n",
      "HTML file saved to html/papers/2022.12.01.518742v1.pdf\n",
      "Downloaded: papers/2022.11.26.518030v1.pdf\n",
      "HTML file saved to html/papers/2022.11.26.518030v1.pdf\n",
      "Downloaded: papers/2022.11.25.517947v1.pdf\n",
      "HTML file saved to html/papers/2022.11.25.517947v1.pdf\n",
      "Downloaded: papers/2022.11.05.515129v1.pdf\n",
      "HTML file saved to html/papers/2022.11.05.515129v1.pdf\n",
      "Downloaded: papers/2022.11.04.515226v1.pdf\n",
      "HTML file saved to html/papers/2022.11.04.515226v1.pdf\n",
      "Downloaded: papers/2022.11.03.515020v1.pdf\n",
      "HTML file saved to html/papers/2022.11.03.515020v1.pdf\n",
      "Downloaded: papers/2022.11.04.515094v1.pdf\n",
      "HTML file saved to html/papers/2022.11.04.515094v1.pdf\n",
      "Downloaded: papers/2022.11.03.515100v1.pdf\n",
      "HTML file saved to html/papers/2022.11.03.515100v1.pdf\n",
      "Downloaded: papers/2022.09.23.509203v2.pdf\n",
      "HTML file saved to html/papers/2022.09.23.509203v2.pdf\n",
      "Downloaded: papers/2022.10.27.513881v2.pdf\n",
      "HTML file saved to html/papers/2022.10.27.513881v2.pdf\n",
      "Downloaded: papers/2022.10.27.513962v1.pdf\n",
      "HTML file saved to html/papers/2022.10.27.513962v1.pdf\n",
      "Downloaded: papers/2022.10.26.513890v2.pdf\n",
      "HTML file saved to html/papers/2022.10.26.513890v2.pdf\n",
      "Downloaded: papers/2022.10.24.513174v2.pdf\n",
      "HTML file saved to html/papers/2022.10.24.513174v2.pdf\n",
      "Downloaded: papers/2022.10.21.513181v3.pdf\n",
      "HTML file saved to html/papers/2022.10.21.513181v3.pdf\n",
      "Downloaded: papers/2022.10.14.512303v2.pdf\n",
      "HTML file saved to html/papers/2022.10.14.512303v2.pdf\n",
      "Downloaded: papers/2022.09.26.509451v2.pdf\n",
      "HTML file saved to html/papers/2022.09.26.509451v2.pdf\n",
      "Downloaded: papers/2022.09.22.508453v1.pdf\n",
      "HTML file saved to html/papers/2022.09.22.508453v1.pdf\n",
      "Downloaded: papers/2022.09.17.508368v2.pdf\n",
      "HTML file saved to html/papers/2022.09.17.508368v2.pdf\n",
      "Downloaded: papers/2022.09.06.505536v1.pdf\n",
      "HTML file saved to html/papers/2022.09.06.505536v1.pdf\n",
      "Downloaded: papers/2022.09.05.506692v1.pdf\n",
      "HTML file saved to html/papers/2022.09.05.506692v1.pdf\n",
      "Downloaded: papers/2022.08.30.505939v2.pdf\n",
      "HTML file saved to html/papers/2022.08.30.505939v2.pdf\n",
      "Downloaded: papers/2022.08.11.503648v1.pdf\n",
      "HTML file saved to html/papers/2022.08.11.503648v1.pdf\n",
      "Downloaded: papers/2022.08.02.502545v1.pdf\n",
      "HTML file saved to html/papers/2022.08.02.502545v1.pdf\n",
      "Downloaded: papers/2022.08.04.502770v1.pdf\n",
      "HTML file saved to html/papers/2022.08.04.502770v1.pdf\n",
      "Downloaded: papers/2022.08.01.502266v1.pdf\n",
      "HTML file saved to html/papers/2022.08.01.502266v1.pdf\n",
      "Downloaded: papers/2022.07.09.499321v1.pdf\n",
      "HTML file saved to html/papers/2022.07.09.499321v1.pdf\n",
      "Downloaded: papers/2022.07.06.499059v1.pdf\n",
      "HTML file saved to html/papers/2022.07.06.499059v1.pdf\n",
      "Downloaded: papers/2022.07.05.498785v1.pdf\n",
      "HTML file saved to html/papers/2022.07.05.498785v1.pdf\n",
      "Downloaded: papers/2022.06.30.498290v1.pdf\n",
      "HTML file saved to html/papers/2022.06.30.498290v1.pdf\n",
      "Downloaded: papers/2022.06.27.497795v1.pdf\n",
      "HTML file saved to html/papers/2022.06.27.497795v1.pdf\n",
      "Downloaded: papers/2022.06.20.496811v1.pdf\n",
      "HTML file saved to html/papers/2022.06.20.496811v1.pdf\n",
      "Downloaded: papers/2022.06.19.496725v1.pdf\n",
      "HTML file saved to html/papers/2022.06.19.496725v1.pdf\n",
      "Downloaded: papers/2022.06.08.495395v2.pdf\n",
      "HTML file saved to html/papers/2022.06.08.495395v2.pdf\n",
      "Downloaded: papers/2022.06.06.494894v2.pdf\n",
      "HTML file saved to html/papers/2022.06.06.494894v2.pdf\n",
      "Downloaded: papers/2022.05.27.493681v1.pdf\n",
      "HTML file saved to html/papers/2022.05.27.493681v1.pdf\n",
      "Downloaded: papers/2022.05.23.493179v1.pdf\n",
      "HTML file saved to html/papers/2022.05.23.493179v1.pdf\n",
      "Downloaded: papers/2022.05.24.493244v1.pdf\n",
      "HTML file saved to html/papers/2022.05.24.493244v1.pdf\n",
      "Downloaded: papers/2022.05.19.492613v2.pdf\n",
      "HTML file saved to html/papers/2022.05.19.492613v2.pdf\n",
      "Downloaded: papers/2022.05.19.492566v2.pdf\n",
      "HTML file saved to html/papers/2022.05.19.492566v2.pdf\n",
      "Downloaded: papers/2022.05.17.492399v3.pdf\n",
      "HTML file saved to html/papers/2022.05.17.492399v3.pdf\n",
      "Downloaded: papers/2022.05.11.491472v1.pdf\n",
      "HTML file saved to html/papers/2022.05.11.491472v1.pdf\n",
      "Downloaded: papers/2022.05.10.491293v1.pdf\n",
      "HTML file saved to html/papers/2022.05.10.491293v1.pdf\n",
      "Downloaded: papers/2022.05.06.490930v1.pdf\n",
      "HTML file saved to html/papers/2022.05.06.490930v1.pdf\n",
      "Downloaded: papers/2022.05.05.490773v2.pdf\n",
      "HTML file saved to html/papers/2022.05.05.490773v2.pdf\n",
      "Downloaded: papers/2022.04.25.489326v1.pdf\n",
      "HTML file saved to html/papers/2022.04.25.489326v1.pdf\n",
      "Downloaded: papers/2022.04.23.489213v2.pdf\n",
      "HTML file saved to html/papers/2022.04.23.489213v2.pdf\n",
      "Downloaded: papers/2022.03.29.486305v1.pdf\n",
      "HTML file saved to html/papers/2022.03.29.486305v1.pdf\n",
      "Downloaded: papers/2022.03.28.486092v1.pdf\n",
      "HTML file saved to html/papers/2022.03.28.486092v1.pdf\n",
      "Downloaded: papers/2022.03.25.485477v1.pdf\n",
      "HTML file saved to html/papers/2022.03.25.485477v1.pdf\n",
      "Downloaded: papers/2022.03.24.485682v1.pdf\n",
      "HTML file saved to html/papers/2022.03.24.485682v1.pdf\n",
      "Downloaded: papers/2022.03.23.485539v1.pdf\n",
      "HTML file saved to html/papers/2022.03.23.485539v1.pdf\n",
      "Downloaded: papers/2022.03.14.484211v2.pdf\n",
      "HTML file saved to html/papers/2022.03.14.484211v2.pdf\n",
      "Downloaded: papers/2022.02.23.481560v1.pdf\n",
      "HTML file saved to html/papers/2022.02.23.481560v1.pdf\n",
      "Downloaded: papers/2022.02.18.481102v1.pdf\n",
      "HTML file saved to html/papers/2022.02.18.481102v1.pdf\n",
      "Downloaded: papers/2022.02.20.480002v1.pdf\n",
      "HTML file saved to html/papers/2022.02.20.480002v1.pdf\n",
      "Downloaded: papers/2022.02.15.480535v1.pdf\n",
      "HTML file saved to html/papers/2022.02.15.480535v1.pdf\n",
      "Downloaded: papers/2022.02.11.480026v2.pdf\n",
      "HTML file saved to html/papers/2022.02.11.480026v2.pdf\n",
      "Downloaded: papers/2022.01.24.474473v2.pdf\n",
      "HTML file saved to html/papers/2022.01.24.474473v2.pdf\n",
      "Downloaded: papers/2022.01.18.476738v1.pdf\n",
      "HTML file saved to html/papers/2022.01.18.476738v1.pdf\n",
      "Downloaded: papers/2022.01.15.476199v2.pdf\n",
      "HTML file saved to html/papers/2022.01.15.476199v2.pdf\n",
      "Downloaded: papers/2022.01.13.476259v1.pdf\n",
      "HTML file saved to html/papers/2022.01.13.476259v1.pdf\n",
      "Downloaded: papers/2022.01.11.475790v1.pdf\n",
      "HTML file saved to html/papers/2022.01.11.475790v1.pdf\n",
      "Downloaded: papers/2022.01.11.475882v2.pdf\n",
      "HTML file saved to html/papers/2022.01.11.475882v2.pdf\n",
      "Downloaded: papers/2021.03.26.437240v3.pdf\n",
      "HTML file saved to html/papers/2021.03.26.437240v3.pdf\n",
      "Downloaded: papers/2021.12.22.472709v2.pdf\n",
      "HTML file saved to html/papers/2021.12.22.472709v2.pdf\n",
      "Downloaded: papers/2021.12.20.473562v1.pdf\n",
      "HTML file saved to html/papers/2021.12.20.473562v1.pdf\n",
      "Downloaded: papers/2021.12.16.473094v1.pdf\n",
      "HTML file saved to html/papers/2021.12.16.473094v1.pdf\n",
      "Downloaded: papers/2021.12.16.472765v1.pdf\n",
      "HTML file saved to html/papers/2021.12.16.472765v1.pdf\n",
      "Downloaded: papers/2021.12.15.472871v4.pdf\n",
      "HTML file saved to html/papers/2021.12.15.472871v4.pdf\n",
      "Downloaded: papers/2021.12.14.472718v2.pdf\n",
      "HTML file saved to html/papers/2021.12.14.472718v2.pdf\n",
      "Downloaded: papers/2021.03.22.436437v3.pdf\n",
      "HTML file saved to html/papers/2021.03.22.436437v3.pdf\n",
      "Downloaded: papers/2021.10.25.465678v2.pdf\n",
      "HTML file saved to html/papers/2021.10.25.465678v2.pdf\n",
      "Downloaded: papers/2021.12.03.471074v3.pdf\n",
      "HTML file saved to html/papers/2021.12.03.471074v3.pdf\n",
      "Downloaded: papers/2021.12.03.471027v4.pdf\n",
      "HTML file saved to html/papers/2021.12.03.471027v4.pdf\n",
      "Downloaded: papers/2021.11.24.469861v1.pdf\n",
      "HTML file saved to html/papers/2021.11.24.469861v1.pdf\n",
      "Downloaded: papers/2021.11.24.469859v2.pdf\n",
      "HTML file saved to html/papers/2021.11.24.469859v2.pdf\n",
      "Downloaded: papers/2021.10.08.463637v3.pdf\n",
      "HTML file saved to html/papers/2021.10.08.463637v3.pdf\n",
      "Downloaded: papers/2021.09.09.459700v2.pdf\n",
      "HTML file saved to html/papers/2021.09.09.459700v2.pdf\n",
      "Downloaded: papers/2021.11.10.467921v3.pdf\n",
      "HTML file saved to html/papers/2021.11.10.467921v3.pdf\n",
      "Downloaded: papers/2021.11.09.467907v2.pdf\n",
      "HTML file saved to html/papers/2021.11.09.467907v2.pdf\n",
      "Downloaded: papers/2021.10.29.466422v2.pdf\n",
      "HTML file saved to html/papers/2021.10.29.466422v2.pdf\n",
      "Downloaded: papers/2021.11.02.466900v2.pdf\n",
      "HTML file saved to html/papers/2021.11.02.466900v2.pdf\n",
      "Downloaded: papers/2020.06.08.139964v3.pdf\n",
      "HTML file saved to html/papers/2020.06.08.139964v3.pdf\n",
      "Downloaded: papers/2020.12.10.420448v2.pdf\n",
      "HTML file saved to html/papers/2020.12.10.420448v2.pdf\n",
      "Downloaded: papers/2021.09.29.462206v2.pdf\n",
      "HTML file saved to html/papers/2021.09.29.462206v2.pdf\n",
      "Downloaded: papers/2021.09.28.462151v1.pdf\n",
      "HTML file saved to html/papers/2021.09.28.462151v1.pdf\n",
      "Downloaded: papers/2021.09.28.462251v1.pdf\n",
      "HTML file saved to html/papers/2021.09.28.462251v1.pdf\n",
      "Downloaded: papers/2021.09.15.460475v2.pdf\n",
      "HTML file saved to html/papers/2021.09.15.460475v2.pdf\n",
      "Downloaded: papers/2021.09.14.460272v1.pdf\n",
      "HTML file saved to html/papers/2021.09.14.460272v1.pdf\n",
      "Downloaded: papers/2021.09.10.459728v1.pdf\n",
      "HTML file saved to html/papers/2021.09.10.459728v1.pdf\n",
      "Downloaded: papers/2021.09.07.459217v1.pdf\n",
      "HTML file saved to html/papers/2021.09.07.459217v1.pdf\n",
      "Downloaded: papers/2021.09.03.458862v1.pdf\n",
      "HTML file saved to html/papers/2021.09.03.458862v1.pdf\n",
      "Downloaded: papers/2021.09.02.458674v1.pdf\n",
      "HTML file saved to html/papers/2021.09.02.458674v1.pdf\n",
      "Downloaded: papers/2021.09.01.458518v1.pdf\n",
      "HTML file saved to html/papers/2021.09.01.458518v1.pdf\n",
      "Downloaded: papers/2021.08.30.458191v1.pdf\n",
      "HTML file saved to html/papers/2021.08.30.458191v1.pdf\n",
      "Downloaded: papers/2021.08.24.457596v1.pdf\n",
      "HTML file saved to html/papers/2021.08.24.457596v1.pdf\n",
      "Downloaded: papers/2021.08.26.457874v1.pdf\n",
      "HTML file saved to html/papers/2021.08.26.457874v1.pdf\n",
      "Downloaded: papers/2021.03.19.436173v2.pdf\n",
      "HTML file saved to html/papers/2021.03.19.436173v2.pdf\n",
      "Downloaded: papers/2021.08.11.456034v1.pdf\n",
      "HTML file saved to html/papers/2021.08.11.456034v1.pdf\n",
      "Downloaded: papers/2021.05.20.444927v2.pdf\n",
      "HTML file saved to html/papers/2021.05.20.444927v2.pdf\n",
      "Downloaded: papers/2021.08.05.455277v1.pdf\n",
      "HTML file saved to html/papers/2021.08.05.455277v1.pdf\n",
      "Downloaded: papers/2021.08.03.454900v1.pdf\n",
      "HTML file saved to html/papers/2021.08.03.454900v1.pdf\n",
      "Downloaded: papers/2021.03.23.436560v3.pdf\n",
      "HTML file saved to html/papers/2021.03.23.436560v3.pdf\n",
      "Downloaded: papers/2021.07.30.454505v2.pdf\n",
      "HTML file saved to html/papers/2021.07.30.454505v2.pdf\n",
      "Downloaded: papers/2021.07.23.453484v2.pdf\n",
      "HTML file saved to html/papers/2021.07.23.453484v2.pdf\n",
      "Downloaded: papers/2020.07.28.224584v3.pdf\n",
      "HTML file saved to html/papers/2020.07.28.224584v3.pdf\n",
      "Downloaded: papers/2021.07.18.452853v2.pdf\n",
      "HTML file saved to html/papers/2021.07.18.452853v2.pdf\n",
      "Downloaded: papers/2021.07.12.452063v1.pdf\n",
      "HTML file saved to html/papers/2021.07.12.452063v1.pdf\n",
      "Downloaded: papers/2021.02.04.429752v3.pdf\n",
      "HTML file saved to html/papers/2021.02.04.429752v3.pdf\n",
      "Downloaded: papers/2021.02.18.431773v3.pdf\n",
      "HTML file saved to html/papers/2021.02.18.431773v3.pdf\n",
      "Downloaded: papers/2021.07.02.450893v1.pdf\n",
      "HTML file saved to html/papers/2021.07.02.450893v1.pdf\n",
      "Downloaded: papers/2021.07.02.450883v1.pdf\n",
      "HTML file saved to html/papers/2021.07.02.450883v1.pdf\n",
      "Downloaded: papers/2021.03.16.435535v2.pdf\n",
      "HTML file saved to html/papers/2021.03.16.435535v2.pdf\n",
      "Downloaded: papers/2021.06.22.449514v1.pdf\n",
      "HTML file saved to html/papers/2021.06.22.449514v1.pdf\n",
      "Downloaded: papers/2021.06.23.449389v1.pdf\n",
      "HTML file saved to html/papers/2021.06.23.449389v1.pdf\n",
      "Downloaded: papers/2021.06.16.448630v1.pdf\n",
      "HTML file saved to html/papers/2021.06.16.448630v1.pdf\n",
      "Downloaded: papers/2021.06.14.448320v1.pdf\n",
      "HTML file saved to html/papers/2021.06.14.448320v1.pdf\n",
      "Downloaded: papers/2021.06.08.446897v1.pdf\n",
      "HTML file saved to html/papers/2021.06.08.446897v1.pdf\n",
      "Downloaded: papers/2021.06.05.447065v5.pdf\n",
      "HTML file saved to html/papers/2021.06.05.447065v5.pdf\n",
      "Downloaded: papers/2021.05.23.445375v1.pdf\n",
      "HTML file saved to html/papers/2021.05.23.445375v1.pdf\n",
      "Downloaded: papers/2021.05.17.444424v2.pdf\n",
      "HTML file saved to html/papers/2021.05.17.444424v2.pdf\n",
      "Downloaded: papers/2021.05.17.444412v2.pdf\n",
      "HTML file saved to html/papers/2021.05.17.444412v2.pdf\n",
      "Downloaded: papers/2021.05.17.444522v1.pdf\n",
      "HTML file saved to html/papers/2021.05.17.444522v1.pdf\n",
      "Downloaded: papers/2021.05.17.444419v1.pdf\n",
      "HTML file saved to html/papers/2021.05.17.444419v1.pdf\n",
      "Downloaded: papers/2021.05.12.443913v1.pdf\n",
      "HTML file saved to html/papers/2021.05.12.443913v1.pdf\n",
      "Downloaded: papers/2021.02.24.429166v2.pdf\n",
      "HTML file saved to html/papers/2021.02.24.429166v2.pdf\n",
      "Downloaded: papers/2021.05.04.442591v1.pdf\n",
      "HTML file saved to html/papers/2021.05.04.442591v1.pdf\n",
      "Downloaded: papers/2021.04.27.441597v2.pdf\n",
      "HTML file saved to html/papers/2021.04.27.441597v2.pdf\n",
      "Downloaded: papers/2021.04.21.440749v2.pdf\n",
      "HTML file saved to html/papers/2021.04.21.440749v2.pdf\n",
      "Downloaded: papers/2021.04.21.440844v1.pdf\n",
      "HTML file saved to html/papers/2021.04.21.440844v1.pdf\n",
      "Downloaded: papers/2021.04.19.440412v1.pdf\n",
      "HTML file saved to html/papers/2021.04.19.440412v1.pdf\n",
      "Downloaded: papers/2021.04.02.438271v3.pdf\n",
      "HTML file saved to html/papers/2021.04.02.438271v3.pdf\n",
      "Downloaded: papers/2020.03.29.014704v4.pdf\n",
      "HTML file saved to html/papers/2020.03.29.014704v4.pdf\n",
      "Downloaded: papers/2020.11.23.394700v3.pdf\n",
      "HTML file saved to html/papers/2020.11.23.394700v3.pdf\n",
      "Downloaded: papers/2020.12.22.423111v3.pdf\n",
      "HTML file saved to html/papers/2020.12.22.423111v3.pdf\n",
      "Downloaded: papers/2021.04.02.438176v1.pdf\n",
      "HTML file saved to html/papers/2021.04.02.438176v1.pdf\n",
      "Downloaded: papers/2020.12.20.423642v4.pdf\n",
      "HTML file saved to html/papers/2020.12.20.423642v4.pdf\n",
      "Downloaded: papers/2021.03.29.437575v1.pdf\n",
      "HTML file saved to html/papers/2021.03.29.437575v1.pdf\n",
      "Downloaded: papers/2021.03.26.437296v1.pdf\n",
      "HTML file saved to html/papers/2021.03.26.437296v1.pdf\n",
      "Downloaded: papers/2021.03.22.436504v1.pdf\n",
      "HTML file saved to html/papers/2021.03.22.436504v1.pdf\n",
      "Downloaded: papers/2020.08.07.241729v2.pdf\n",
      "HTML file saved to html/papers/2020.08.07.241729v2.pdf\n",
      "Downloaded: papers/2021.03.08.434372v1.pdf\n",
      "HTML file saved to html/papers/2021.03.08.434372v1.pdf\n",
      "Downloaded: papers/2020.07.24.219691v3.pdf\n",
      "HTML file saved to html/papers/2020.07.24.219691v3.pdf\n",
      "Downloaded: papers/2021.02.23.432530v2.pdf\n",
      "HTML file saved to html/papers/2021.02.23.432530v2.pdf\n",
      "Downloaded: papers/2021.02.26.433113v1.pdf\n",
      "HTML file saved to html/papers/2021.02.26.433113v1.pdf\n",
      "Downloaded: papers/2021.02.04.429837v2.pdf\n",
      "HTML file saved to html/papers/2021.02.04.429837v2.pdf\n",
      "Downloaded: papers/2021.02.22.432219v1.pdf\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'PDFObjRef' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[105], line 93\u001b[0m\n\u001b[1;32m     90\u001b[0m         htmlpath \u001b[38;5;241m=\u001b[39m pdf_to_html(pdfpath, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhtml/\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m pdfpath)\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m---> 93\u001b[0m     \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[105], line 90\u001b[0m, in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     88\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m url \u001b[38;5;129;01min\u001b[39;00m paper_urls:\n\u001b[1;32m     89\u001b[0m     pdfpath \u001b[38;5;241m=\u001b[39m download_pdf(url)\n\u001b[0;32m---> 90\u001b[0m     htmlpath \u001b[38;5;241m=\u001b[39m \u001b[43mpdf_to_html\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpdfpath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhtml/\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mpdfpath\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[105], line 15\u001b[0m, in \u001b[0;36mpdf_to_html\u001b[0;34m(pdf_path, output_html_path)\u001b[0m\n\u001b[1;32m     12\u001b[0m laparams \u001b[38;5;241m=\u001b[39m LAParams()\n\u001b[1;32m     13\u001b[0m html_content \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m<html><body>\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 15\u001b[0m \u001b[43m\u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mpage_layout\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mextract_pages\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpdf_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlaparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlaparams\u001b[49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhtml_content\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m<div class=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpage\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m>\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m     17\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43melement\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mpage_layout\u001b[49m\u001b[43m:\u001b[49m\n",
      "File \u001b[0;32m~/git/imfine/env/lib/python3.12/site-packages/pdfminer/high_level.py:209\u001b[0m, in \u001b[0;36mextract_pages\u001b[0;34m(pdf_file, password, page_numbers, maxpages, caching, laparams)\u001b[0m\n\u001b[1;32m    207\u001b[0m device \u001b[38;5;241m=\u001b[39m PDFPageAggregator(resource_manager, laparams\u001b[38;5;241m=\u001b[39mlaparams)\n\u001b[1;32m    208\u001b[0m interpreter \u001b[38;5;241m=\u001b[39m PDFPageInterpreter(resource_manager, device)\n\u001b[0;32m--> 209\u001b[0m \u001b[43m\u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mpage\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mPDFPage\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_pages\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    210\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpage_numbers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmaxpages\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxpages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpassword\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpassword\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcaching\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcaching\u001b[49m\n\u001b[1;32m    211\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m    212\u001b[0m \u001b[43m    \u001b[49m\u001b[43minterpreter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprocess_page\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpage\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlayout\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mdevice\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/git/imfine/env/lib/python3.12/site-packages/pdfminer/pdfpage.py:171\u001b[0m, in \u001b[0;36mPDFPage.get_pages\u001b[0;34m(cls, fp, pagenos, maxpages, password, caching, check_extractable)\u001b[0m\n\u001b[1;32m    169\u001b[0m         log\u001b[38;5;241m.\u001b[39mwarning(warning_msg)\n\u001b[1;32m    170\u001b[0m \u001b[38;5;66;03m# Process each page contained in the document.\u001b[39;00m\n\u001b[0;32m--> 171\u001b[0m \u001b[43m\u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43mpageno\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpage\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43menumerate\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_pages\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdoc\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m    172\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mpagenos\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43mpageno\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mpagenos\u001b[49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m    173\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01mcontinue\u001b[39;49;00m\n",
      "File \u001b[0;32m~/git/imfine/env/lib/python3.12/site-packages/pdfminer/pdfpage.py:127\u001b[0m, in \u001b[0;36mPDFPage.create_pages\u001b[0;34m(cls, document)\u001b[0m\n\u001b[1;32m    125\u001b[0m     objects \u001b[38;5;241m=\u001b[39m search(document\u001b[38;5;241m.\u001b[39mcatalog[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPages\u001b[39m\u001b[38;5;124m\"\u001b[39m], document\u001b[38;5;241m.\u001b[39mcatalog)\n\u001b[1;32m    126\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m (objid, tree) \u001b[38;5;129;01min\u001b[39;00m objects:\n\u001b[0;32m--> 127\u001b[0m         \u001b[38;5;28;01myield\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mdocument\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobjid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtree\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mnext\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mpage_labels\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    128\u001b[0m         pages \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m pages:\n\u001b[1;32m    130\u001b[0m     \u001b[38;5;66;03m# fallback when /Pages is missing.\u001b[39;00m\n",
      "File \u001b[0;32m~/git/imfine/env/lib/python3.12/site-packages/pdfminer/pdfpage.py:63\u001b[0m, in \u001b[0;36mPDFPage.__init__\u001b[0;34m(self, doc, pageid, attrs, label)\u001b[0m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlastmod \u001b[38;5;241m=\u001b[39m resolve1(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mattrs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLastModified\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[1;32m     60\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mresources: Dict[\u001b[38;5;28mobject\u001b[39m, \u001b[38;5;28mobject\u001b[39m] \u001b[38;5;241m=\u001b[39m resolve1(\n\u001b[1;32m     61\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mattrs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResources\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mdict\u001b[39m())\n\u001b[1;32m     62\u001b[0m )\n\u001b[0;32m---> 63\u001b[0m mediabox_params: List[Any] \u001b[38;5;241m=\u001b[39m \u001b[43m[\u001b[49m\n\u001b[1;32m     64\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresolve1\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmediabox_param\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mmediabox_param\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mattrs\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mMediaBox\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     65\u001b[0m \u001b[43m\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmediabox: Rect \u001b[38;5;241m=\u001b[39m resolve1(mediabox_params)\n\u001b[1;32m     67\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCropBox\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mattrs:\n",
      "\u001b[0;31mTypeError\u001b[0m: 'PDFObjRef' object is not iterable"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import os\n",
    "from pdfminer.high_level import extract_text\n",
    "from pdfminer.layout import LAParams\n",
    "from pdfminer.high_level import extract_pages\n",
    "from pdfminer.layout import LTTextContainer\n",
    "import os\n",
    "\n",
    "def pdf_to_html(pdf_path, output_html_path):\n",
    "    # Extract text and layout information from PDF\n",
    "    laparams = LAParams()\n",
    "    html_content = \"<html><body>\"\n",
    "\n",
    "    for page_layout in extract_pages(pdf_path, laparams=laparams):\n",
    "        html_content += \"<div class='page'>\"\n",
    "        for element in page_layout:\n",
    "            if isinstance(element, LTTextContainer):\n",
    "                text = element.get_text()\n",
    "                html_content += f\"<p>{text}</p>\"\n",
    "        html_content += \"</div>\"\n",
    "\n",
    "    html_content += \"</body></html>\"\n",
    "\n",
    "    # Save the HTML content to a file\n",
    "    with open(output_html_path + '.html', 'w', encoding='utf-8') as f:\n",
    "        f.write(html_content)\n",
    "\n",
    "    print(f\"HTML file saved to {output_html_path}\")\n",
    "    return output_html_path\n",
    "\n",
    "def search_biorxiv(query, max_results=10, perpage=75, page=0, paper_links=[]):\n",
    "    search_url = f\"https://www.biorxiv.org/search/%2522{query}%2522%20numresults%3A{perpage}%20sort%3Apublication-date%20direction%3Adescending?page={page}\"\n",
    "    print(search_url)\n",
    "    response = requests.get(search_url)\n",
    "    soup = BeautifulSoup(response.text, 'html.parser')\n",
    "    papers = soup.find_all('li', class_='search-result')\n",
    "    total = int(soup.find('h1', {\"id\": 'page-title'}).get_text().split()[0])\n",
    "    for paper in papers:\n",
    "        link = paper.find('a', class_='highwire-cite-linked-title')\n",
    "        \n",
    "        if link:\n",
    "            paper_links.append(f\"https://www.biorxiv.org{link.get('href')}\")\n",
    "    if (perpage * page > total):\n",
    "        print('done finding papers')\n",
    "        print('found ' + str(len(paper_links)))\n",
    "        return paper_links[0:total]\n",
    "    else:\n",
    "        print('getting the next page')\n",
    "        print(page)\n",
    "        return search_biorxiv(query, max_results, perpage, page + 1, paper_links)\n",
    "\n",
    "def download_pdf(paper_url, output_dir='papers'):\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "\n",
    "    response = requests.get(paper_url + '.full.pdf', stream=True)\n",
    "    #https://www.biorxiv.org/content/10.1101/2021.06.23.449389v1.full.pdf\n",
    "    #https://www.biorxiv.org/content/10.1101/2024.07.31.606100v1\n",
    "\n",
    "    paper_id = paper_url.split('/')[-1]\n",
    "    file_path = os.path.join(output_dir, f\"{paper_id}.pdf\")\n",
    "\n",
    "\n",
    "    with open(file_path, 'wb') as fd:\n",
    "        for chunk in response.iter_content(2048):\n",
    "            fd.write(chunk)\n",
    "\n",
    "    print(f\"Downloaded: {file_path}\")\n",
    "    return file_path\n",
    "\n",
    "import os\n",
    "import json\n",
    "\n",
    "def create_jsonl_from_directory(directory_path, output_jsonl_path, max_tokens=2048):\n",
    "    # List all files in the directory\n",
    "    file_list = os.listdir(directory_path)\n",
    "\n",
    "    with open(output_jsonl_path, 'w', encoding='utf-8') as jsonl_file:\n",
    "        for filename in file_list:\n",
    "            file_path = os.path.join(directory_path, filename)\n",
    "            \n",
    "            # Check if the path is a file (and not a directory)\n",
    "            if os.path.isfile(file_path):\n",
    "                # Read the contents of the file\n",
    "                with open(file_path, 'r', encoding='utf-8') as f:\n",
    "                    file_contents = f.read()\n",
    "\n",
    "                # Split the file contents into chunks of max_tokens words\n",
    "                words = file_contents.split()\n",
    "                for i in range(0, len(words), max_tokens):\n",
    "                    chunk = words[i:i + max_tokens]\n",
    "                    chunk_text = \" \".join(chunk)\n",
    "\n",
    "                    # Create a JSON object with the chunk contents\n",
    "                    json_object = {\"text\": chunk_text}\n",
    "\n",
    "                    # Write the JSON object as a JSONL line\n",
    "                    jsonl_file.write(json.dumps(json_object) + '\\n')\n",
    "\n",
    "def main():\n",
    "    query = \"graph genome\"\n",
    "    max_results = 100\n",
    "    \n",
    "    print(\"Searching for papers...\")\n",
    "    paper_urls = search_biorxiv(query, max_results=max_results)\n",
    "    print(paper_urls)\n",
    "    # these ones get stuck parsing\n",
    "    remove = [\n",
    "        \"https://www.biorxiv.org/content/10.1101/2024.02.14.580266v1\",\n",
    "        \"https://www.biorxiv.org/content/10.1101/2023.10.04.560829v1\"\n",
    "    ]\n",
    "    print(\"Downloading papers...\")\n",
    "    for r in remove:\n",
    "        if r in paper_urls:\n",
    "            paper_urls.remove(r)\n",
    "    for url in paper_urls:\n",
    "        pdfpath = download_pdf(url)\n",
    "        htmlpath = pdf_to_html(pdfpath, 'html/' + pdfpath)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "\n",
    "# Example usage:\n",
    "directory_path = \"html/papers\"\n",
    "output_jsonl_path = \"output.jsonl\"\n",
    "create_jsonl_from_directory(directory_path, output_jsonl_path, 2048)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "13a97f74-b619-4edf-84a5-b30cebd68720",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat html/papers/*.html > papers.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "64462408-d55c-4163-9c3c-c639274cf025",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def create_jsonl_from_directory(directory_path, output_jsonl_path):\n",
    "    # List all files in the directory\n",
    "    file_list = os.listdir(directory_path)\n",
    "\n",
    "    with open(output_jsonl_path, 'w', encoding='utf-8') as jsonl_file:\n",
    "        for filename in file_list:\n",
    "            file_path = os.path.join(directory_path, filename)\n",
    "            \n",
    "            # Check if the path is a file (and not a directory)\n",
    "            if os.path.isfile(file_path):\n",
    "                # Read the contents of the file\n",
    "                with open(file_path, 'r', encoding='utf-8') as f:\n",
    "                    file_contents = f.read()\n",
    "\n",
    "                # Create a JSON object with the file contents\n",
    "                json_object = {\"text\": file_contents}\n",
    "\n",
    "                # Write the JSON object as a JSONL line\n",
    "                jsonl_file.write(json.dumps(json_object) + '\\n')\n",
    "\n",
    "# Example usage:\n",
    "directory_path = \"html/papers\"\n",
    "output_jsonl_path = \"output.jsonl\"\n",
    "create_jsonl_from_directory(directory_path, output_jsonl_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "f8799d3f-c55f-423b-81b9-d3404034b274",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "def create_jsonl_from_directory(directory_path, output_jsonl_path, max_tokens=2048):\n",
    "    # List all files in the directory\n",
    "    file_list = os.listdir(directory_path)\n",
    "\n",
    "    with open(output_jsonl_path, 'w', encoding='utf-8') as jsonl_file:\n",
    "        for filename in file_list:\n",
    "            file_path = os.path.join(directory_path, filename)\n",
    "            \n",
    "            # Check if the path is a file (and not a directory)\n",
    "            if os.path.isfile(file_path):\n",
    "                # Read the contents of the file\n",
    "                with open(file_path, 'r', encoding='utf-8') as f:\n",
    "                    file_contents = f.read()\n",
    "\n",
    "                # Split the file contents into chunks of max_tokens words\n",
    "                words = file_contents.split()\n",
    "                for i in range(0, len(words), max_tokens):\n",
    "                    chunk = words[i:i + max_tokens]\n",
    "                    chunk_text = \" \".join(chunk)\n",
    "\n",
    "                    # Create a JSON object with the chunk contents\n",
    "                    json_object = {\"text\": chunk_text}\n",
    "\n",
    "                    # Write the JSON object as a JSONL line\n",
    "                    jsonl_file.write(json.dumps(json_object) + '\\n')\n",
    "\n",
    "# Example usage:\n",
    "directory_path = \"html/papers\"\n",
    "output_jsonl_path = \"output.jsonl\"\n",
    "create_jsonl_from_directory(directory_path, output_jsonl_path, 2048)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "70375420-caf1-401f-90e0-e6eda59be84d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"text\": \"<html><body><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>Compression algorithm for colored de Bruijn graphs </p><p>Amatur Rahman1 </p><p>Yoann Dufresne4,5 </p><p>Paul Medvedev1,2,3 </p><p>1 Department of Computer Science and Engineering, The Pennsylvania State University 2 Department of Biochemistry and Molecular Biology, The Pennsylvania State University 3 Huck Institutes of the Life Sciences, The Pennsylvania State University 4 Institut Pasteur, Universit\\u00b4e Paris Cit\\u00b4e, G5 Sequence Bioinformatics, Paris, France 5 Institut Pasteur, Universit\\u00b4e Paris Cit\\u00b4e, Bioinformatics and Biostatistics Hub, F-75015 Paris, France </p><p>Abstract </p><p>A colored de Bruijn graph (also called a set of k-mer sets), is a set of k-mers with every k-mer assigned a set of colors. Colored de Bruijn graphs are used in a variety of applications, including variant calling, genome assembly, and database search. However, their size has posed a scalability challenge to algorithm developers and users. There have been numerous indexing data structures proposed that allow to store the graph compactly while supporting fast query operations. However, disk compression algorithms, which do not need to support queries on the compressed data and can thus be more space- efficient, have received little attention. The dearth of specialized compression tools has been a detriment In this paper, we develop a new tool that to tool developers, tool users, and reproducibility efforts. compresses colored de Bruijn graphs to disk, building on previous ideas for compression of k-mer sets and indexing colored de Bruijn graphs. We test our tool, called ESS-color, on various datasets, including both sequencing data and whole genomes. ESS-color achieves better compression than all evaluated tools and all datasets, with no other tool able to consistently achieve less than 44% space overhead. The software is available at http://github.com/medvedevgroup/ESSColor. </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>1 Introduction </p><p>Modern methods for analyzing biological sequences often reduce the input dataset to a set of short, fixed length strings called k-mers. When working with a collection of such datasets E = (E0, . . . , E|E|\\u22121), it is fruitful to represent them as one union set of k-mers and, for each k-mer, the indices of the datasets to which it belongs. The set of indices of each k-mer is referred to as its color class, and E is referred to as a colored de Bruijn graph [1]. A colored de Bruijn graph (cdBG) is commonly used to represent a sequence database, such as a collection of sequencing experiments or a collection of assembled genomes. For example, it is used by tools for inferring phylogenies [2], quantification of RNA expression [3], and studying the evolution of antimicrobial resistance [4]. </p><p>As sequence database sizes grow to petabytes [5], the cost of storing or transferring the data (e.g. on Amazon Web Services or in-house compute infrastructure) has underscored the need for efficient disk com- pression algorithms. Such costs are often prohibitive for smaller labs and, even for larger labs, limit the scale of data that can be analyzed. Large file sizes also hamper tool development, which relies on iterative load- ing/copying/modifying data, and reproducibility efforts, which require downloading and storing the data. For example, storing the 31-mers from 450,000 microbial genomes in compressed form takes about 12 Tb [4]. Unfortunately, there has not been a lot of work to develop methods for disk compression of colored de Bruijn graphs. </p><p>In contrast to disk compression, indexing cdBGs have received much attention [6]. A slew of data structures have been developed, optimizing various metrics such as index size, construction time, or query time (see the survey [6] and its follow up [7]). Indexing data structures exploit the structure of cdBGs and use clever tricks to compress the color classes of similar k-mers. But they also carry a space overhead to efficiently support queries; since this is not needed for disk compression, indexing data structures are usually not competitive with custom made cdBG compression methods. </p><p>The simplest option for compressing a cdBG is to compress each color (i.e. dataset) independently, using a compression tool designed for a single set of k-mers (e.g. [8]). This approach can work well when k-mers tend to not be shared among colors. However, most cdBGs have a large overlap between the k-mers of various colors. In such cases, independently compressing each color does not exploit the properties of cdBGs and, as we show in this paper, results in subpar compression ratios. There exist two tools designed specifically for disk compression of cdBGs. The first tool is unfortunately limited to only three k-mer sizes [9]. The second tool, called GGCAT [10] is a space efficient indexing method that, while not originally evaluated in this regard, turns out to also be a good disk compression method when combined with a generic post-compression step. In this paper, we design, implement, and evaluate an algorithm ESS-color for the disk compression of cdBGs. We build upon the idea of spectrum-preserving string sets [11, 12, 13] and the followup compression format for a k-mer set [8], called ESS. By constructing an ESS of the union of k-mers in E, we represent the k-mer sequences themselves compactly. We exploit the fact that consecutive k-mers in an ESS have similar color classes in order to efficiently compress the color vectors of each k-mer. </p><p>We evaluate ESS-color on a variety of datasets, including bacteria, fungi, human, and including whole genome sequencing data, metagenome sequencing data, and whole assembled genomes. ESS-color achieves better compression than all evaluated tools and on all datasets, with all other tools using \\u2265 44% more space on at least one of our datasets. On some datasets the improvement over all other tools is quite large, e.g. for a gut metagenome, all the other tools use at \\u2265 27% more space than ESS-color. Compressing each color independently uses between 1.2x and 6.9x more space than ESS-color. The absolute compression ratio is more than 26x on datasets of assembled genomes and between 1.4x and 8.7x on datasets from sequencing experiments. The software is available at http://github.com/medvedevgroup/ESSColor. </p><p>2 Preliminaries </p><p>In this section we give some important definitions. Please refer to Figure 1 for examples of the introduced concepts. </p><p>Strings: A string of length k is called a k-mer. We assume k-mers are over the DNA alphabet. A string x is canonical if it is the lexicographically smaller of x and its reverse complement. Let K be a set of k-mers. </p><p>1 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>A spectrum-preserving string set (SPSS) of K is a set of strings S such that each string s \\u2208 S is at least k characters long, every k-mer that appears in S appears exactly once, and the set of k-mers that appear in S is K [11, 12, 13]. For example, if K = {ACG, CGT, CGA}, then {ACGT, CGA} would an SPSS of K. Note that K can have multiple spectrum-preserving string sets. There are several efficient tools for computing an SPSS so as to minimize the total number of characters [14, 10]. In this paper, we rely on the implementation in [8]. Each string in the resulting SPSS is referred to as a simplitig. </p><p>Compression of a k-mer set ESS is a disk-compression format to store a set of k-mers K. It was introduced in [8] as the output of a compression tool, which, in this paper, we will refer to as ESS-basic. The technical details of the format and of the tool are irrelevant for this paper and can be viewed as black boxes. An ESS representation cannot be queried efficiently but can be decompressed into an SPSS of K. This output gives an ordering of the k-mers of K, and therefore the ESS compression of K induces an ordering on K. Note that because the decompression algorithm is deterministic, by storing an ESS representation, we are implicitly storing an SPSS representation as well. </p><p>Colored k-mer sets: Let C > 0 be an integer indicating the number of colors. Let E = {E0, . . . , EC\\u22121} be a set of C k-mer sets, also referred to as a colored de Bruijn graph. Let E be the set of all k-mers in E, i.e. E = {x | \\u2203i s. t. x \\u2208 Ei}. The (color) class of a k-mer x \\u2208 E is the set of indices i such that x \\u2208 Ei. The color vector of x is a binary vector of length C where position i is 1 iff x \\u2208 Ei. </p><p>Non-compressed representation of cdBGs: Assume you have an ordering of E, e.g. the one given by an ESS of E. A color matrix of E is a file with row i containing the color vector of the ith k-mer. Storing an ESS of E together with a color matrix of E is a lossless representation of E. </p><p>3 Methods </p><p>In this section, we describe our algorithm ESS-color for the compression of cdBGs. Let E = {E0, . . . , EC\\u22121} be a colored dBG over C colors. Recall that E is the set of all k-mers in E. Let M denote the number of distinct color classes in E. </p><p>ESS-color can accept input in one of two formats. First, it can accept each Ei stored in the KFF file format [15]. Alternatively, we can take as input a collection of FASTA files, each one assigned one of C colors, and an abundance parameter a. Ei is then implicitly defined as the set of all canonical k-mers that appear at least a times in the FASTA files of color i. We obtain Ei by running KMC [16] on the FASTA files of color i. </p><p>3.1 Building a color matrix of E and compressing E </p><p>In this step, we first compress E using ESS-basic [8] and then build the color matrix of E, ordered by the ESS order. Specifically, we first compress the nucleotide sequences themselves, i.e. we run ESS-basic [8] on all the input files jointly. We refer to this as the union ESS. We then decompress this file to obtain an SPSS of E, denoted by S. From S, we build an SSHash dictionary [17] that allows us to map each k-mer in E to its rank in S. We then build on top of the KMC API to read in the binary files representing E0, . . . , EC\\u22121 and output a color matrix ordered by the SSHash dictionary. At the end of this stage, we have the union ESS, which is retained in the final compression output, and we have S and the color matrix, which are used in later stages but not retained in the final compression output. </p><p>3.2 Compression of the color matrix </p><p>Given an SPSS S of E and a color matrix of E over the order induced by S, we now generate a compressed representation of the matrix. Our representation consists of a global class table and, for every simplitig of S, a few bits of metadata, a local class table and one bitvector m. The local class table is optional, as we describe below. Figure 2 gives a schematic representation. We now explain each of these in detail. </p><p>2 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>Figure 1: An example illustrating the various definitions in Section 2 and the first step of our compression method (Section 3.1). The input\"}\n",
      "{\"text\": \"to the compression algorithm is a colored de Bruin Graph. The top panel shows an example with three colors (i.e. C = 3), k = 5, and a colored cdBG of E = {E0, E1, E2}. Here, E0 = {T CAAA, CAAAA, AAAAT }, E1 = {T CAAA, CAAAA, AAAAT, AAAT T, CAAAG, AAAT C, AAT CG}, and E2 = {T CAAA, CAAAA, AAAAT, AAAT T, CAAAG}. The color class is shown next to each k-mer, the color class of T CAAA is {0, 1, 2}. There are three distinct color classes, i.e. M = 3. These e.g. are {0, 1, 2}, {1, 2}, and {1}. The lower left panel shows the union ESS, i.e. the ESS representation of the set E = {T CAAA, CAAAA, AAAAT, AAAT T, CAAAG, AAAT C, AAT CG}. This union ESS can be decomposed into an SPSS of E, shown in the figure. The third column in bottom panel shows the color matrix, with k-mers in the order of the SPSS. To obtain the final compressed representation the color matrix is compressed using the algorithm we describe in Section 3.2. </p><p>3 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>Global class table: For most applications, the number of distinct color vectors M is significantly smaller than 2C. Hence, the color matrix representation, which uses C bits per k-mer, is very inefficient. Instead, we use Huffman coding to assign a global ID to each class, so as to minimize the number of bits that will be used to store these IDs later (this is similar to what was done in [18]). To do this, we scan the color matrix to determine all the distinct classes and the number of k-mers that have each class. We then use Huffman coding to assign a global ID to each distinct class, so that more frequent global IDs tend to use less bits. This table is then stored in two forms: one that is compressed to disk, and the other that is stored in memory to be used during the compression algorithm. </p><p>We store the table on disk using three files: a color encoding \\u2206, a boundary bitvector b, and a text file. First, we sort the color classes in increasing numerical order, interpreting each color vector as a C-bit integer. For \\u2206, we write a concatenation of the M color vectors to disk, with the first color vector being written using C bits and the following colors being encoded as a difference with their predecessor. Specifically, if hi is the Hamming distance between the ith and the (i \\u2212 1)st color vectors, then we use hi\\u2308log C\\u2309 bits to encode the indices where the ith color vector is different from the (i \\u2212 1)st color vector. We also store a boundary bitvector b which is the same length as \\u2206 and contains a \\u20181\\u2019 whenever \\u2206 starts a new color class. Finally, we store the frequencies of the color classes in a text file. These three files are then sufficient to reconstruct the global IDs during decompression. 1 </p><p>Simultaneously, we need to be able to map a color vector to an ID during the compression process. To do this, we create a minimal perfect hash function h (CHM [20]) that maps from each distinct color vector to an integer between 0 and M \\u2212 1. We then maintain an array A of size M , where for each color vector c, A(h(c)) holds the global ID of color vector c. </p><p>After the global class table is created, we process the simplitigs of S one at time. For each simplitig, we dynamically set two parameters: a boolean variable U seLocalID, and an integer 0 \\u2264 maxDif \\u2264 2. We postpone the discussion on how these are set until the end of the section. The values of maxDif and U seLocalID are stored using 3 bits of metadata per simplitig. If U seLocalID is set, we create a local class table: </p><p>Local class table: In the case that the frequencies of color classes are evenly distributed, we need approximately log M bits to represent the global class ID of each k-mer. We observe that sometimes a class is used at multiple locations of a simplitig, in which case using log M bits for each occurrence can be wasteful. Let \\u2113 be the number of distinct classes appearing in a simplitig. To save space on class IDs, we create a separate local class table, which maps from \\u2113 integers, called local class IDs, to their respective global IDs. Then, the encoding of k-mer classes for this simplitig can use local class IDs, which take only log \\u2113 space. The local class table is written to disk, with log M bits encoding \\u2113 followed by \\u2113 consecutive global class IDs together </p><p>The bitvector m is constructed by scanning the simplitig from left to right and, for each k-mer x, deciding how to encode it, and appending that encoding to m. Intuitively, the encodings follow three basic possibilities. The first possibility is to just append m with the k-mer\\u2019s class ID. Second, we observed in practice that simplitigs often contain runs of k-mers with identical classes, in which case we can append m with the length of the run, rather than writing out each class IDs (such runs are similar to the monotigs of [21]). Finally, we often observe that a k-mer has a color vector with a small Hamming distance (i.e. 1 or 2) to that of the previous k-mer. In this case, we append m with the indices in the color vector that are different. Since there are three types of encoding, we will also need to prepend each encoding with two bits denoting the type of encoding. Formally, for each k-mer in a simplitig, we choose one of four options: </p><p>Skip: This option is invoked if x is not the first or last k-mer in its simplitig and has the same class as the preceding and succeeding k-mer. In this option, nothing is appended to m. </p><p>Small class difference: Let h be the Hamming distance between the color vector of x and the color vector of the preceding k-mer in the simplitig. This option is invoked when 0 < h \\u2264 maxDif . First, we </p><p>1For readers familiar with Mantis-MST [19], we also tried their approach for storing our global table. Surprisingly, we found that our approach outperformed their more sophisticated approach, at least in our datasets. Though the Mantis-MST approach resulted in a smaller \\u2206 vector, the overhead of storing the tree parent vector outweighed this gain. </p><p>4 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>Figure 2: Example of how we compress the color matrix (Section 3.2). The top panel shows the compression of a simplitig ACTTTGGA and the bottom panel shows the compression of a simplitig AAAATTAC. Other simplitigs exist but are not shown. In this example, C = 6, k = 4, and runDivisor = 4. The figure is divided into three columns. The first column shows the information that the algorithm has about each k-mer (i.e. its color vector and corresponding global ID). The second column shows the metadata which holds the values of maxDif and useLocalId for the corresponding simplitig. The third column shows the m vector. Within the m vector, the type of encoding is shown in gray colored boxes and other values are shown in white boxes. The colors of the values inside the white boxes correspond to the color used for the corresponding k-mers\\u2019 color vector in the first column of the figure. Note that in the bottom simplitig, a local table is used. In particular, there are two distinct color vectors in this simplitig, with global ID 0011 assigned to local ID 0 and global ID 000 assigned to local ID 1. </p><p>append m with \\u201810\\u2019 to indicate that the following encoding will encode a class difference. If maxDif = 2, then we append m with a \\u20181\\u2019 to indicate that h = 2 or a \\u20180\\u2019 to indicate that h = 1. If maxDif = 1, then we do not append this extra bit, since it is implicit. Then, we append m with h log C bits which list the colors that are different. Note that setting maxDif = 0 effectively disables this type of encoding. </p><p>End of run: This option is invoked if x has the same class as the preceding k-mer and either has a different class than the succeeding k-mer or is the last k-mer in the simplitig. First, we indicate that the following encoding will encode a run length by appending m with \\u201811\\u2019 if maxDif > 0 and \\u20181\\u2019 if maxDif = 0. This difference is due to the fact that if maxDif = 0, then there are only two types of encodings and so we can just use one bit for the type. </p><p>Let runLen be the number of consecutive k-mers that preceded x (not including x) and have the same class. We encode runLen by separating it into a quotient q and remainder r (with respect to a global parameter runDivisor), and then encoding the quotient q in unary and the remainder r in binary. Formally, let q = \\u230a runLen runDivisor \\u230b and r = runLen mod runDivisor. We append to m q \\u20181\\u2019s followed by a \\u20190\\u2019. Then, we append to m the binary encoding of r, using log runDivisor bits. For example, if runDivisor = 16 and runLen = 21, then q = 1 and r = 5, and m is appended with 100101. Observe that a smaller value of runDivisor results in more bits used to encode long runs (i.e. q is larger) while a larger value of runDivisor uses more bits to encode short runs (i.e. log runDivisor is larger). We found that a default value of 16 works best in our experiments. </p><p>Store the class ID: This option is invoked when none of the criteria for the other options are satisfied. In this case, we append m with \\u20180\\u2019 to indicate the type of encoding, followed by the class ID of x. If </p><p>5 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>U seLocalID is set, we use the local class ID, otherwise we use the global class ID. </p><p>After finishing with all simplitigs, we compress the global class table, local class tables, and m using RRR [22] and write them to disk. </p><p>Setting the parameters U seLocalID and maxDif involves trade-offs that are difficult to quantify in advance. For example, the cost of having to store the local class table may exceed the benefits of using less bits to encode class IDs for a simplitig where every present class ID is contained within a single run. Similarly, when d is too large, then writing the positions of the color differences to m can take more space then just writing the class ID. Moreover, there is a benefit of setting d = 0, since it enables to save one bit per run by using \\u20181\\u2019 instead of \\u201811\\u2019 for the \\u2018end of run\\u2019 encoding. All bitvectors are additionally compressed with RRR, making it difficult to determine in advance which parameters result in the least space. We therefore try all possibilities of maxDif \\u2208 {0, 1, 2} and U seLocalID \\u2208 {T rue, F alse}, and, for each combination, compute the encoding. We then use the encoding that takes less space and disregard the rest. Though this step can likely be optimized, we found that the\"}\n",
      "{\"text\": \"time taken to try all possibilities was not a large factor in the overall compression time. </p><p>The decompression algorithm for the m vector is straightforward since our color matrix compression scheme is designed to be unambiguously decompressed. Simultaneously, we decompress E with ESS- If the output is decompress. The result is an SPSS S of E and a color matrix of E in the order of S. to be processed downstream in a streaming manner, our decompression algorithm can trivially stream out k-mer sequence and color vectors, one k-mer at a time. </p><p>4 Results </p><p>4.1 Evaluated tools and datasets </p><p>As far as we are aware, there are two other tools designed for compressing colored de Bruijn graphs: KS [9] and GGCAT [10]. We refer to the first tool as KS after the authors\\u2019 last names [9]. KS is limited to support only three k values (15, 19, and 23), so we compare against it for k = 23 but also evaluate ESS-color on a more practical k value of k = 31. For GGCAT, we additionally compressed its Fasta output file with MFC and its binary color table with gzip to maximize its compression ratio. We also compare ESS-color against the naive approach of compressing each color independently using the algorithm of [8], which we refer to as ESS-basic. </p><p>Table 1 shows the datasets we use for evaluation and their properties. We chose five datasets so as to cover a broad range of input types. Three of the datasets are from assembled genomes, one is from RNA-seq reads, and one is from metagenome reads. We used all k-mers from the three assemblies datasets and all k-mers that appear at least twice from the two read datasets. The datasets cover various species, from Bacteria to Fungi to Human. Concretely, we have 1) 100 arbitrarily selected E.coli strains from GenBank, 2) an arbitrary subset of 10 of those, 3) 20 arbitrarily selected fungi sequences from RefSeq, 4) gut microbiome read sets from nine individuals sequenced in [23], and 5) 19 paired-end, human, bulk RNA- seq short-read experiments previously used in [19]. All accession numbers are listed in https://github. com/medvedevgroup/ESSColor/wiki/Experiments. </p><p>Finally, all experiments were run on a server with an Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz processor with 64 cores and 512 GB of memory. We ran all tools in unrestricted memory mode. We used 8 threads for all tools and their components, whenever they supported multi-threading. </p><p>4.2 Comparison against other disk compression tools </p><p>Table 2 shows the bits per k-mer achieved by ESS-color compared with KS, GGCAT ESS-basic, and a p7zip compression of the original fasta file. ESS-color achieves better compression than all evaluated tools and on all datasets. No other tool was able to consistently achieve less than 44% space overhead compared to ESS-color. On some datasets the improvement over all other tools is quite large, e.g. for Gut (k = 23), all the other tools use at least 27% more space than ESS-color. </p><p>The compression ratio of ESS-color relative to the original Fa.p7zip files varies (Table 2). For the read datasets, it is more than 15x, since high-coverage FASTA files are by their nature very redundant. For </p><p>6 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>Source </p><p>type </p><p>C </p><p>M </p><p>n. k-mers \\u00d7106 (% single color) </p><p>n. simplitigs \\u00d7106 </p><p>k = 23 </p><p>k = 31 </p><p>k = 23 </p><p>k = 31 </p><p>E. coli E. coli Fungi Gut Human </p><p>assemblies assemblies assemblies metagenome reads RNA-seq reads </p><p>100 10 20 9 19 </p><p>542,545 826 13,227 511 9,654 </p><p>27 (30%) 13 (38%) 394 (93%) 2,236 (67%) 120 (71%) </p><p>31 (31%) 14 (41%) 409 (93%) 2,477 (70%) 103 (75%) </p><p>0.5 0.2 1.8 76 7.2 </p><p>0.5 0.2 1.7 95 10 </p><p>Table 1: Dataset characteristics. C is the number of colors and M is the number color classes. </p><p>Dataset </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>k = 23 </p><p>k = 31 </p><p>fa.p7zip ESS-color </p><p>KS GGCAT ESS-basic </p><p>fa.p7zip ESS-color KS GGCAT ESS-basic </p><p>37 8 2.4 60 112 </p><p>5.2 2.9 2.2 3.7 5.6 </p><p>17.3 5.3 2.5 5.7 8.8 </p><p>6.4 3.3 2.3 4.9 7.1 </p><p>35.0 7.3 2.3 4.7 10.4 </p><p>33 7 2.3 54 131 </p><p>4.4 2.6 2.2 3.4 8.9 </p><p>\\u2013 \\u2013 \\u2013 \\u2013 \\u2013 </p><p>5.5 3.0 2.3 4.9 10.5 </p><p>30.3 6.6 2.5 4.4 12.1 </p><p>Table 2: Compression results, in bits per k-mer. ESS-color is our new tool. ESS-basic is the non-integrative approach of compressing every color separately. KS is tool from [9], and a hyphen indicates that it does not support k > 23. fa.p7zip is the p7zip compression of the original data. We show bits per k-mer, which is the total compressed size divided by the number of distinct k-mer in the input (i.e. |E|). Compression ratios can be inferred by comparing to the fa.p7zip column. </p><p>the assemblies datasets, the ratio is between 1x and 7x. We found that a good predictor of compressibility of assemblies is the percentage of k-mers that have exactly one color (Table 1). At one extreme, 93% of Fungi20 k-mers have exactly one color, and ESS-color achieves little improvement over fa.p7zip. At the other extreme, only 30% of Ecoli100 k-mers have exactly one color, and the compression ratio is relatively high at 7.1x (for k = 23). This trend makes intuitive sense since single-color k-mers do not benefit from ESS-color\\u2019s multi-color compression algorithm. </p><p>KS is not as effective as ESS-color on our datasets, using between 1.4x and 3.3x more space than ESS-color (Table 2). We note that even though KS is also designed to exploit the fact that k-mers are shared across colors, it makes a different design trade-off compared to ESS-color. Specifically, it does not allow simplitigs to extend beyond a single color class (resulting in more space needed to store k-mers), but, in exchange, it is more efficient in storing color information. </p><p>GGCAT is generally the closest competitor against ESS-color, using between 14 and 44% more space on the non-Fungi datasets (note that for Fungi all tools did well). Like ESS-color, it builds a kind of global class table, constructs an SPSS of the k-mers, and annotates each run of single-class k-mers with their class ID. Unlike ESS-color, however, it does not use ESS, does not encode small class differences, and does not use local class tables. </p><p>As expected, ESS-basic is not as effective as ESS-color, using up to 6.9x more space than ESS-color (Table 2). These results are not surprising since ESS-basic does not exploit the redundancy created by shared k-mers across samples. For the assemblies datasets, the compression improvement of ESS-color over ESS-basic closely tracks that of ESS-color over the original fa.p7zip. For the sequencing datasets, ESS-basic uses between 1.3 and 1.9 more space than ESS-color. </p><p>Tables 3 and 4 show the run time and memory usage of compression, respectively. Here, ESS-color is outperformed by other tools. In particular, if optimal compression space is not needed, then GGCAT becomes a good alternative to ESS-color. Note that the decompression time (Table 5) is an order of magnitude smaller compared to the compression times. </p><p>7 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>k = 23 </p><p>k = 31 </p><p>Dataset </p><p>ESS-color </p><p>KS </p><p>GGCAT ESS-basic </p><p>ESS-color GGCAT ESS-basic </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>14.8 2.6 59.1 1101* 31.9 </p><p>11.6 7.0 7.0 148.5 10.5 </p><p>0.7 0.3 3.3 37.8** 21.2 </p><p>6.3 0.9 7.3 234 31 </p><p>20.7 2.6 76 611.6 60.1 </p><p>0.7 0.3 3.3 92.2 15.0 </p><p>27 2 25 341 39 </p><p>Table 3: Time (min) used by the various compression algorithms. For the Gut9 ESS-color run with k = 23 (marked with *), we used an unoptimized implementation of the color matrix generation step, since SSHash was not working as expected. For GGCAT on Gut9 with k = 23 (denoted by **), the original run crashed because of exceeding the number of open files allowed by the operating system. We therefore re-ran GGCAT using our simplitigs as a starting point, which allowed the run to complete. However, the time shown here does not include the time we used to construct the simplitigs. </p><p>k = 23 </p><p>k = 31 </p><p>Dataset </p><p>ESS-color </p><p>KS </p><p>GGCAT ESS-basic </p><p>ESS-color GGCAT ESS-basic </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>1.2 0.6 5.4 174.6* 26.8 </p><p>0.9 3.2 3.2 50.6 6.0 </p><p>1.5 0.8 4.8 87.2** 8.8 </p><p>1.2 1.1 5.9 33.2 9.1 </p><p>1.1 0.6 4.3 121 12.1 </p><p>1.4 0.8 3.9 78.2 10.2 </p><p>1 1 6 119 7 </p><p>Table 4: Maximum memory (Gb) used by the various compression algorithms. The (*) and (**) annotations are the same as in Table 3. </p><p>Dataset </p><p>Memory (GB) Time (min) </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>0.5 0.5 0.5 8.5 1.0 </p><p>2 1 13 90 5 </p><p>Table 5: Time and memory for decompression of ESS-color, for k = 31. </p><p>Dataset </p><p>union ESS m global table (\\u2206 and b) </p><p>global table (frequencies) </p><p>local tables </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>51 81 95 77 73 </p><p>23 18 5 22 26 </p><p>23 <0.1 <0.1 <0.1 <0.1 </p><p>2 <0.1 <0.1 <0.1 <0.1 </p><p>1 0.5 0.1 1 0.8 </p><p>Table 6: Breakdown of the space usage (in percentage of total space) of the components of ESS-color, for k = 31. Note that all components except union ESS are shown after compression with RRR. </p><p>Dataset </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>U seLocalID = False </p><p>U seLocalID = True </p><p>maxDif = 0 </p><p>maxDif = 1 </p><p>maxDif = 2 </p><p>maxDif = 0 </p><p>maxDif = 1 </p><p>maxDif = 2 </p><p>75 82 99 80 91 </p><p>20 16 1 19 9 </p><p>5 2 0.2 2 0.1 </p><p><0.1 <0.1 <0.1 0.2 0.1 </p><p><0.1 <0.1 <0.1 <0.1 <0.1 </p><p><0.1 <0.1 <0.1 <0.1 <0.1 </p><p>Table 7: The percentage of simplitigs (k = 31) that fall into the six compression modes, i.e. combinations of U seLocalID and maxDif . </p><p>8 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>Dataset </p><p>ESS-color RowDiff-ESS RowDiff+ Rainbow-MST </p><p>Ecoli100 Ecoli10 Fungi20 Gut9 HumRNA19 </p><p>4.4 2.6 2.2 3.4 8.9 </p><p>14.6 7.8 3.6 11.2 37.9 </p><p>8.0 6.3 4.3 40.6 112.6 </p><p>34.2 19.3 9.5 15.6 19.1 </p><p>Table 8: Compression results in bits per k-mer (k = 31) of indexing approaches, compared to ESS-color. </p><p>4.3 Inside the space usage of ESS-color </p><p>ESS-color\\u2019s compressed representation includes several components, with the major ones being the union ESS, the m vector, the global table, and the local tables. Table 6 shows that the majority of space used by ESS-color is taken by the union ESS. Except for Ecoli100, the rest of the space is taken up almost exclusively by m. For Ecoli100, which has the largest number of colors, the global table takes 23% of the total space. </p><p>Recall that ESS-color chooses one of six different compression modes for each simplitig, i.e. U seLocalId \\u2208 {0, 1} and maxDif \\u2208 {0, 1, 2}. In order to access the relative contribution of the various compression techniques, we count the frequency with which each mode occurs (Table 7). First, we observe that the idea of a local table was rarely helpful on our data. Local tables are only beneficial when a single color class appears in more than one run in a simplitig, which apparently was rare. Second, the majority of simplitigs use maxDif = 0. This mode is optimal when the simplitig has just one color class. There is also a more complicated trade-off since setting maxDif > 0 adds one extra bit for each run encoding, which may outweigh the benefits of encoding some k-mers with a class difference. Third, the Gut dataset demonstrates the benefit of encoding class differences, especially compared to GGCAT. It is the dataset with the highest percentage of simplitigs using maxDif > 0 (21%) and, simultaneously, it is also the dataset where GGCAT uses\"}\n",
      "{\"text\": \"the most space relative to ESS-color (44%). </p><p>4.4 Comparison to indexing data structures </p><p>There exist numerous indexing data structures for cdBGs [6]. Indexing data structures are similar to disk compression but additionally support efficient membership and color queries. We expect this overhead to make them non-competitive with respect to disk compression schemes. To verify this, we compared the space taken by ESS-color against three indexing approaches. We note that since these approaches are designed for indexing, they do not implement decompression and are thus not viable for disk compression in their current state. We also note that GGCAT also supports indexing, but, since it is trivial to decompress, we included it in the main analysis of Section 4.2. </p><p>The first two approaches are ones that are shown in [24, 25] to be the most space efficient. These are RowDiff+, which is the latest version [25] of RowDiff [24], and Rainbow-MST [24], which is a space-improved version of Mantis-MST [19]. As a trivial improvement, we further compress these indices using gzip. The third approach we compare to is a natural hybrid of ESS-color and the RowDiff indexing algorithm for cdBGs [24]. We refer to this as RowDiff-ESS and describe it in detail in the Appendix. We do not compare against other indices such as REINDEER [21], Bifrost [26], Themisto [27], or Mantis-MST [19], because they are less space efficient than RowDiff+, and we do not compare against Sequence Bloom Tree approaches (e.g. [28, 29]) because they are lossy. </p><p>Table 8 shows the results. As expected, the compression ratios of these indexing tools are not competitive against ESS-color. Even the most space efficient indexing approach for each dataset takes 60% more space than ESS-color. We do note that GGCAT, which was shown in Table 2, is an exception, since it implements both efficient indexing and disk compression. </p><p>5 Conclusion </p><p>Colored de Bruijn graphs are a popular way to represent sequence databases. In spite of their ever-growing sizes, there have not been many specialized tools for compressing them to disk. In this paper, we present a </p><p>9 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>novel disk compression algorithm tailormade for cdBGs that achieves superior space compression compared to all other tools on the evaluated datasets. </p><p>Our algorithm is a novel combination of ideas borrowed from previous work on disk compression of k-mer sets and indexing of cdBGs. We use a spectrum-preserving string set (SPSS) as a basis for both compressing the nucleotide sequences and for ordering the rows in the color matrix. By using the SPSS ordering, we can avoid the costly storage of an indexed de Bruijn graph (e.g. BOSS in [24, 25] or a counting quotient filter in [19]). We also exploit the fact that consecutive k-mers in an SPSS likely have the same or similar color class. A major component of our approach is that we select a different compression scheme for each simplitig, depending on what gives the best compression on that simplitig. </p><p>The most important practical direction for future work is to improve the running time of our algorithm. The generation of the union ESS is done by ESS-basic. ESS-basic can be sped up by extending the latest SPSS generation tools [14, 10] to also compute an ESS. We could even build on top of GGCAT, taking advantage of their efficient implementation (unfortunately, GGCAT was only released once our project was near completion). Another bottleneck is the color matrix generation step, which could be parallelized or even avoided by using color lists. </p><p>A theoretical future direction is to derive bounds on the bits used by the compression scheme. Unfortu- nately, we do not see an easy way to do this, since the choice of encoding depends on the order of the k-mers in the SPSS and on the decomposition of the k-mers into simplitigs. It is unclear to us how to capture these properties as a function of the input data. </p><p>Finally, we could further improve the compression algorithm by modifying the SPSS generated by the ESS-basic algorithm. Currently, the choice of how to select from multiple simplitig extensions is made arbitrarily. Instead, the choice could be made to use the extension that has the most similar color class. Such a modification to the SPSS construction algorithm would likely be computationally non-trivial, since it would require accessing the color information. </p><p>10 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>References </p><p>[1] Z. Iqbal, M. Caccamo, I. Turner, P. Flicek, and G. McVean, \\u201cDe novo assembly and genotyping of </p><p>variants using colored de Bruijn graphs,\\u201d Nature genetics, vol. 44, no. 2, pp. 226\\u2013232, 2012. </p><p>[2] R. Wittler, \\u201cAlignment-and reference-free phylogenomics with colored de Bruijn graphs,\\u201d Algorithms </p><p>for Molecular Biology, vol. 15, pp. 1\\u201312, 2020. </p><p>[3] R. Patro, G. Duggal, M. I. Love, R. A. Irizarry, and C. Kingsford, \\u201cSalmon provides fast and bias-aware </p><p>quantification of transcript expression,\\u201d Nature methods, vol. 14, no. 4, pp. 417\\u2013419, 2017. </p><p>[4] P. Bradley, H. C. Den Bakker, E. P. Rocha, G. McVean, and Z. Iqbal, \\u201cUltrafast search of all deposited </p><p>bacterial and viral genomic data,\\u201d Nature biotechnology, vol. 37, no. 2, pp. 152\\u2013159, 2019. </p><p>[5] L. Papageorgiou, P. Eleni, S. Raftopoulou, M. Mantaiou, V. Megalooikonomou, and D. Vlachakis, </p><p>\\u201cGenomic big data hitting the storage bottleneck,\\u201d EMBnet. journal, vol. 24, 2018. </p><p>[6] C. Marchet, C. Boucher, S. J. Puglisi, P. Medvedev, M. Salson, and R. Chikhi, \\u201cData structures based on k-mers for querying large collections of sequencing data sets,\\u201d Genome Research, vol. 31, no. 1, pp. 1\\u201312, 2020. </p><p>[7] C. Marchet, \\u201cData-structures for sets of k-mer sets: what\\u2019s new since 2020,\\u201d Blog post, 2022. [Online]. </p><p>Available: https://kamimrcht.github.io/webpage/sets kmer sets.html </p><p>[8] A. Rahman, R. Chikhi, and P. Medvedev, \\u201cDisk compression of k-mer sets,\\u201d Algorithms for Molecular </p><p>Biology, vol. 16, no. 1, pp. 1\\u201314, 2021. </p><p>[9] K. Kitaya and T. Shibuya, \\u201cCompression of multiple k-mer sets by iterative SPSS decomposition,\\u201d in 21st International Workshop on Algorithms in Bioinformatics (WABI 2021). Schloss Dagstuhl-Leibniz- Zentrum f\\u00a8ur Informatik, 2021. </p><p>[10] A. Cracco and A. I. Tomescu, \\u201cExtremely-fast construction and querying of compacted and colored de </p><p>Bruijn graphs with GGCAT,\\u201d BioRxiv, 2022. </p><p>[11] A. Rahman and P. Medevedev, \\u201cRepresentation of k-mer sets using spectrum-preserving string sets,\\u201d </p><p>Journal of Computational Biology, vol. 28, no. 4, pp. 381\\u2013394, 2021. </p><p>[12] K. B\\u02c7rinda, \\u201cNovel computational techniques for mapping and classifying next-generation sequencing </p><p>data,\\u201d Ph.D. dissertation, Universit\\u00b4e Paris-Est, 2016. </p><p>[13] K. B\\u02c7rinda, M. Baym, and G. Kucherov, \\u201cSimplitigs as an efficient and scalable representation of de </p><p>Bruijn graphs,\\u201d Genome biology, vol. 22, pp. 1\\u201324, 2021. </p><p>[14] J. Khan, M. Kokot, S. Deorowicz, and R. Patro, \\u201cScalable, ultra-fast, and low-memory construction of </p><p>compacted de Bruijn graphs with Cuttlefish 2,\\u201d Genome biology, vol. 23, no. 1, p. 190, 2022. </p><p>[15] Y. Dufresne, T. Lemane, P. Marijon, P. Peterlongo, A. Rahman, M. Kokot, P. Medvedev, S. Deorowicz, and R. Chikhi, \\u201cThe K-mer File Format: a standardized and compact disk representation of sets of k-mers,\\u201d Bioinformatics, vol. 38, no. 18, pp. 4423\\u20134425, 2022. </p><p>[16] M. Kokot, M. D(cid:32)lugosz, and S. Deorowicz, \\u201cKMC 3: counting and manipulating k-mer statistics,\\u201d </p><p>Bioinformatics, vol. 33, no. 17, pp. 2759\\u20132761, 2017. </p><p>[17] G. E. Pibiri, \\u201cSparse and skew hashing of k-mers,\\u201d Bioinformatics, vol. 38, no. Supplement 1, pp. </p><p>i185\\u2013i194, 2022. </p><p>[18] F. Almodaresi, P. Pandey, and R. Patro, \\u201cRainbowfish: A succinct colored de Bruijn graph represen- tation,\\u201d in 17th International Workshop on Algorithms in Bioinformatics (WABI 2017), ser. Leibniz International Proceedings in Informatics (LIPIcs), vol. 88, 2017, pp. 18:1\\u201318:15. </p><p>11 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>[19] F. Almodaresi, P. Pandey, M. Ferdman, R. Johnson, and R. Patro, \\u201cAn Efficient, Scalable, and Exact Representation of High-Dimensional Color Information Enabled Using de Bruijn Graph Search,\\u201d Journal of Computational Biology, vol. 27, no. 4, pp. 485\\u2013499, 2020. </p><p>[20] Z. J. Czech, G. Havas, and B. S. Majewski, \\u201cAn optimal algorithm for generating minimal perfect hash </p><p>functions,\\u201d Information processing letters, vol. 43, no. 5, pp. 257\\u2013264, 1992. </p><p>[21] C. Marchet, Z. Iqbal, D. Gautheret, M. Salson, and R. Chikhi, \\u201cREINDEER: efficient indexing of k- mer presence and abundance in sequencing datasets,\\u201d Bioinformatics, vol. 36, no. Supplement 1, pp. i177\\u2013i185, 2020. </p><p>[22] R. Raman, V. Raman, and S. S. Rao, \\u201cSuccinct dynamic data structures,\\u201d in Algorithms and Data Structures: 7th International Workshop, WADS 2001 Providence, RI, USA, August 8\\u201310, 2001 Pro- ceedings 7. Springer, 2001, pp. 426\\u2013437. </p><p>[23] J. Mas-Lloret, M. Ob\\u00b4on-Santacana, G. Ib\\u00b4a\\u02dcnez-Sanz, E. Guin\\u00b4o, M. L. Pato, F. Rodriguez-Moranta, A. Mata, A. Garc\\u00b4\\u0131a-Rodr\\u00b4\\u0131guez, V. Moreno, and V. N. Pimenoff, \\u201cGut microbiome diversity detected by high-coverage 16S and shotgun sequencing of paired stool and colon sample,\\u201d Scientific data, vol. 7, no. 1, p. 92, 2020. </p><p>[24] D. Danciu, M. Karasikov, H. Mustafa, A. Kahles, and G. R\\u00a8atsch, \\u201cTopology-based sparsification of </p><p>graph annotations,\\u201d Bioinformatics, vol. 37, no. Supplement 1, pp. i169\\u2013i176, 2021. </p><p>[25] M. Karasikov, H. Mustafa, G. R\\u00a8atsch, and A. Kahles, \\u201cLossless indexing with counting de Bruijn </p><p>graphs,\\u201d in Research in Computational Molecular Biology, 2022, pp. 374\\u2013376. </p><p>[26] G. Holley and P. Melsted, \\u201cBifrost: highly parallel construction and indexing of colored and compacted </p><p>de Bruijn graphs,\\u201d Genome biology, vol. 21, no. 1, pp. 1\\u201320, 2020. </p><p>[27] J. N. Alanko, J. Vuohtoniemi, T. M\\u00a8aklin, and S. J. Puglisi, \\u201cThemisto: a scalable colored k-mer index </p><p>for sensitive pseudoalignment against hundreds of thousands of bacterial genomes,\\u201d 2023. </p><p>[28] R. S. Harris and P. Medvedev, \\u201cImproved representation of Sequence Bloom Trees,\\u201d Bioinformatics, </p><p>vol. 36, no. 3, pp. 721\\u2013727, 2020. </p><p>[29] B. Solomon and C. Kingsford, \\u201cFast search of thousands of short-read sequencing experiments,\\u201d Nature </p><p>biotechnology, vol. 34, no. 3, pp. 300\\u2013302, 2016. </p><p>Acknowledgements: </p><p>We would like to thank R. Chikhi for helpful discussions. This material is based upon work supported by the National Science Foundation under grant nos. 2138585 and 1931531. Research reported in this publication was also supported by the National Institutes of Health under Grant NIH R01GM146462 (to P.M.). A.R. was supported by the National Institutes of Health Computation, Bioinformatics, and Statistics (CBIOS) training program. The content is solely the responsibility of the authors and does not necessarily represent the official views of the National Institutes of Health. </p><p>12 </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.05.12.540616 </p><p>; </p><p>this version posted June 30, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made </p><p>available under a </p><p>CC-BY 4.0 International license . </p><p>6 Appendix </p><p>Here we describe RowDiff-ESS, the hybrid of ESS-color and the RowDiff indexing algorithm for cdBGs [24]. Though the approach turned out to not be comptetive against ESS-color, we describe it here for completeness. The RowDiff index is composed of two parts: BOSS, which is an index of E, and a compressed color matrix whose labels are implicitly given by BOSS. Because of its structural similarity to our approach, we can swap out BOSS (which supports queries and is therefore not space efficient for disk compression) with an ESS of E. We then feed the k-mer ordering implied by the ESS to the RowDiff color matrix compression algorithm. The space used by this scheme is the ESS space plus the space of the RowDiff\\u2019s color matrix, compressed with gzip. </p><p>13 </p></div></body></html>\"}\n",
      "{\"text\": \"<html><body><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Ancient dolphin genomes reveal rapid repeated adaptation to coastal waters </p><p>Marie Louis1,2,3,4*, Petra Korlevi\\u01075,6*, Milaja Nyk\\u00e4nen7,8*, Frederick Archer9, Simon Berrow 10,11, Andrew Brownlow12, Eline D. Lorenzen2, Joanne O\\u2019Brien 10,11, Klaas Post13, Fernando Racimo2, Emer Rogan8, Patricia E. Rosel14, Mikkel H. S. Sinding15, Henry van der Es13, Nathan Wales16, Michael C. Fontaine3,17**, Oscar Gaggiotti1**, Andrew D. Foote18,19** *joint first author **joint senior author </p><p>1. Scottish Oceans Institute, University of St Andrews, East Sands, St Andrews KY16 8LB, </p><p>Scotland, UK </p><p>2. Globe Institute, University of Copenhagen, \\u00d8ster Voldgade 5, 1350 Copenhagen, </p><p>Denmark </p><p>3. Groningen Institute for Evolutionary Life Sciences (GELIFES), University of Groningen, PO Box </p><p>11103 CC, Groningen, The Netherlands </p><p>4. Greenland Institute of Natural Resources, Kivioq 2, Nuuk 3900, Greenland 5. Max Planck Institute for Evolutionary Anthropology, Deutscher Platz 6, 04103 Leipzig, </p><p>Germany </p><p>6. Wellcome Sanger Institute, Wellcome Genome Campus, Hinxton, Cambridge, CB10 1SA, UK 7. Department of Environmental and Biological Sciences, PO Box 111, FI-80101 Joensuu, Finland 8. School of Biological, Earth and Environmental Sciences, University College Cork, North Mall, </p><p>Cork, Ireland </p><p>9. Marine Mammal and Turtle Division, Southwest Fisheries Science Center, NOAA, 8901 La </p><p>Jolla Shores Drive, La Jolla, CA 92037, USA </p><p>10. Irish Whale and Dolphin Group, Kilrush, Co Clare, Ireland 11. Marine and Freshwater Research Centre, Department of Natural Sciences, School of Science and Computing, Galway-Mayo Institute of Technology, Dublin Road, H91 T8NW Galway, Ireland. </p><p>12. Scottish Marine Animal Stranding Scheme, Institute of Biodiversity, Animal Health & Comparative MedicineCollege of Medical, Veterinary & Life Sciences, University of Glasgow </p><p>13. Natural History Museum Rotterdam, Westzeedijk 345, 3015 AA Rotterdam, Netherlands 14. Marine Mammal and Turtle Division, Southeast Fisheries Science Center, NOAA, Lafayette, LA </p><p>70506, USA </p><p>15. Department of Biology, University of Copenhagen, Ole Maal\\u00f8es Vej 5, 2200 Copenhagen, </p><p>Denmark. </p><p>16. University of York, BioArCh, Environment Building, Wentworth Way, Heslington, York, YO10 </p><p>5DD </p><p>17. MIVEGEC (Universit\\u00e9 de Montpellier, CNRS 5290, IRD 229) Institut de Recherche pour le </p><p>D\\u00e9veloppement (IRD), F-34394, Montpellier, France </p><p>18. Molecular Ecology Fisheries Genetics Lab, School of Biological Sciences, Bangor University, </p><p>Bangor, UK </p><p>19. Department of Natural History, Norwegian University of Science and Technology (NTNU), NO- </p><p>7491 Trondheim, Norway </p><p>Corresponding authors: Marie Louis (marielouis17@hotmail.com), Andrew D. Foote (andrew.foote@ntnu.no) </p><p>1 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Abstract </p><p>Parallel evolution provides among the strongest evidence of the role of natural selection in </p><p>shaping adaptation to the local environment. Yet, the chronology, mode and tempo of the </p><p>process of parallel evolution remains broadly debated and discussed in the field of </p><p>evolutionary biology. In this study, we harness the temporal resolution of paleogenomics to </p><p>understand the tempo and independence of parallel coastal ecotype adaptation in common </p><p>bottlenose dolphins (Tursiops truncatus). For this, we generated whole genome re- </p><p>sequencing data from subfossil dolphins (8,610-5,626 years BP) originating from around the </p><p>formation time of new coastal habitat and compared them with data from contemporary </p><p>populations. Genomic data revealed a shift in genetic affinity, with the oldest ancient sample </p><p>being closer to the pelagic populations, while the younger samples had intermediate ancestry </p><p>that showed greater affinity with the local contemporary coastal populations. We found </p><p>coastal-associated genotypes in the genome of our highest coverage ancient sample, SP1060, </p><p>providing rare evidence of rapid adaptation from standing genetic variation. Lastly, using </p><p>admixture graph analyses, we found a reticulate evolutionary history between pelagic and </p><p>coastal populations. Ancestral gene flow from coastal populations was the probable source </p><p>of standing genetic variation present in the pelagic populations that enabled rapid adaptation </p><p>to newly emerged coastal habitat. The genetic response to past climatic warming provides an </p><p>understanding of how bottlenose dolphins will respond to ongoing directional climate change </p><p>and shifting coastlines. </p><p>2 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Introduction </p><p>Parallel genetic changes can arise from selection repeatedly acting upon standing genetic </p><p>variation 1,2. However, how the timing and the independence of selection vary is still poorly </p><p>understood for most natural study systems 3,4. For example, selection can act upon standing </p><p>genetic variation in a shared ancestral population, or independently in multiple derived </p><p>populations. Alternatively, adaptive standing genetic variation can be shared among derived </p><p>populations through gene flow. Each of these non-mutually exclusive processes leaves a </p><p>distinct genomic signature in the regions flanking the targets of selection 3,4. </p><p>In this study, we investigate the tempo and independence of parallel linked selection on </p><p>standing genetic variation in coastal-adapted common bottlenose dolphins (Tursiops </p><p>truncatus) derived from pelagic ancestors using contemporary and ancient (8,610-5,626 years </p><p>BP) genomes. Coastal and pelagic ecotypes of bottlenose dolphins have recurrently formed </p><p>in different regions of the world 6\\u20138. Genetic variation has segregated under divergent </p><p>selection between multiple coastal and pelagic ecotype pairs 5. In contrast, the same genetic </p><p>variation is shared among widely distributed coastal populations 5. These coastal-associated </p><p>variants are found mainly in ancient ancestry tracts in the genomes of coastal dolphins, and </p><p>are present at low frequency as standing variation in pelagic populations 5. However, the </p><p>mode and chronology of parallel linked selection on coastal-associated genetic variation in </p><p>bottlenose dolphins remain unresolved. Here, we explore potential scenarios, incorporating </p><p>ancient genomes dating from the estimated time of formation of local coastal populations 8,9. </p><p>Four subfossil samples dredged from the Southern North Sea bed in the eastern North Atlantic </p><p>(ENA) were radiocarbon-dated to 5,979-5,626 calendar years before present (BP) for the </p><p>youngest (SP1060, 95% CI) and 8,610-8,243 years BP for the oldest sample (NMR10326, Table </p><p>S1, Figure 1a-b). These ages fall within the 95% CI range of the estimated split time of pelagic </p><p>and coastal ecotypes in the ENA (47,800-4,300 years BP 8) and the emergence of coastal </p><p>habitat in Northern European waters 9. This coastal habitat emerged when Doggerland was </p><p>submerged by the North Sea around 12,000 to 6,000 years BP following post-glacial sea-level </p><p>rise 10. During this warmer period of the early Holocene, subfossil evidence from the region </p><p>confirms common bottlenose dolphin occurrence in the North Sea 11,12. It is thought this </p><p>3 </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>species entered the North Sea through the English Channel after it opened up following </p><p>deglaciation 11. </p><p>To understand the importance of these four early- to mid-Holocene ancient samples in the </p><p>chronology of coastal adaptation, we first established their relationship to modern dolphins. </p><p>We sequenced an ancient dolphin genome, SP1060, at 3x effective coverage (i.e. post-QC </p><p>filtering, repeat masking, removing duplicates, base quality recalibration), to compare with a </p><p>dataset of 60 modern genomes 5. We sequenced an additional three ancient dolphin genomes </p><p>at ultra-low effective coverage (< 0.05x; Table S1) to verify whether SP1060 was </p><p>representative of the genetic variation within the ancient population, rather than, for </p><p>example, a rare admixed individual. </p><p>Figure 1. Sampling locations and ancestry of ancient and modern common bottlenose dolphin individuals. a) Map of sample locations of the four ancient (green) and 60 modern coastal (shades of red (c)) and pelagic (shades of blue(p)) common bottlenose dolphins in the eastern North Atlantic (ENAc and ENAp), western North Atlantic (WNAc and WNAp) and eastern North Pacific (ENPc and ENPp). b) Mandible of a 8,610-8,243 years BP bottlenose dolphin (sample NMR10326) included in this study. c) Principal component analysis of pseudo-haploid data from 60 modern and four projected ancient samples, mapped to the killer whale (Orcinus orca) reference genome to avoid reference bias and removing transitions, showing first and second principal components (PCs) based on 624,969 SNPs. The proportion of genetic variance captured by each component is indicated in the axes (see also Figures S4-S8). d) Ancestry proportions of sample SP1060 (5,979-5,626 years BP) identified using factorial analysis 13. </p><p>4 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Results and discussion </p><p>Relationship of ancient genomes to modern dolphins </p><p>To explore the relationship between the ancient subfossils and modern individuals, we first </p><p>looked at mitochondrial ancestry. The mitogenome sequence of the oldest ancient sample </p><p>NMR10326 (Figure 1b) clusters with modern North Atlantic pelagic haplotypes, while the </p><p>mitogenome sequences of the three other ancient samples clustered with modern </p><p>Mediterranean and Black Sea samples in a Bayesian phylogenetic analysis (Figure S2). Then, </p><p>on the nuclear data, we used principal component analysis 14,15 (PCA) and a factor analysis 13 </p><p>(FA) method that can incorporate sample age, and therefore allows us to correct for temporal </p><p>genetic drift when comparing populations 13. Given the differences in depth of coverage and </p><p>the potential biases inherent in mapping and comparing ancient and modern genomes 16, we </p><p>ran the multivariate analyses using several approaches: i) comparing pseudo-haploid </p><p>genomes generated by sampling a random single base for all genomic positions and using PC </p><p>projection of the ancient genomes on to the principal components segregating the modern </p><p>genomes; ii) sampling a single read for each site with no projection but including sites covered </p><p>in all modern and at least one ancient sample, and iii) comparing called genotypes for the </p><p>modern individuals and SP1060 at sites with a minimum depth of 3x coverage, no missing </p><p>data in SP1060 and both a projection and factorial analysis approaches. Due to the </p><p>fragmented and damaged nature of ancient DNA, reads with non-reference alleles are less </p><p>likely to be mapped than those with reference alleles, creating reference bias. To reduce </p><p>reference bias, we ran those analyses with the genomic data mapped to the common </p><p>bottlenose dolphin reference genome (GenBank: GCA_001922835.1), which is a coastal WNA </p><p>individual, with relaxed parameters in BWA17,18, and again after re-mapping the data to the </p><p>killer whale reference genome (Genbank GCA_000331955.2) 19. Using relaxed parameters or </p><p>mapping to another species should help to include a better representation of alternate alleles. </p><p>We also ran all analyses both including and excluding transitions, as C-T transitions are in </p><p>excess at molecule ends of the SP1060 sequence reads, reflecting post-mortem deamination </p><p>damage (see details in the Methods in the supplementary material). Additionally, we down- </p><p>sampled one modern individual from the ENA coastal population to 0.03x to evaluate whether </p><p>the lower coverage ancient samples could be pulled towards the middle of the PCAs due to </p><p>large amounts of missing data. This ENA coastal individual clustered with the other ENA </p><p>5 </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>coastal individuals after downsampling in the PC projection (Figure S3) and single read </p><p>sampling approaches (results not shown), indicating differences in coverage are not </p><p>responsible for the observed patterns of variation. </p><p>As per Louis et al. (2021) 5, the major axis of differentiation is between the Pacific and the </p><p>Atlantic populations (PC 1, Figure 1c). Independent genetic drift in each coastal population </p><p>drives this pattern, while pelagic populations\"}\n",
      "{\"text\": \"from both oceanic regions cluster in the centre </p><p>of the PCA. Modern coastal samples from other locations in the ENA (northern France, Ireland </p><p>and West Scotland) form a cline from the pelagic populations to the East Scotland samples. </p><p>This could be consistent with a northwards range expansion of ENA coastal populations 9. The </p><p>second axis of differentiation separated the two Atlantic coastal populations. </p><p>The ancient samples cluster with the pelagic samples along PC 2 (Figure 1c). Along PC 1, we </p><p>observe affinity towards the two North Atlantic coastal populations for SP1060 and </p><p>NMR2273, and to a lesser extent for NMR10151. We observe similar results in all analyses, </p><p>regardless of filtering and mapping strategy (Figure 1c, Figures S4-S8). However, we note </p><p>some reference bias for the ultra-low coverage samples (Figure S6). The oldest genome </p><p>(NMR10326, 8,610-8,243 years BP) clusters with the pelagic populations when mapped to the </p><p>killer whale reference genome (Figure 1c), while it does not do so when mapping to the </p><p>bottlenose dolphin reference genome (Figure S6), likely due to reference bias associated with </p><p>ancient DNA giving NMR10326 closer affinity to the WNA coastal population where the </p><p>reference genome is from. </p><p>Having established that SP1060 is broadly representative of our ancient samples, we focus on </p><p>SP1060 in the rest of our analyses. We estimated shared ancestry of SP1060 with the modern </p><p>individuals using a factor analysis, which takes genetic drift into account 13. SP1060 shares the </p><p>highest ancestry with the ENA coastal population (ENAc, 43%), followed by the WNA coastal </p><p>(WNAc, 32%) and ENA pelagic (ENAp, 25%, Figure 1d) populations. Clearly, the inferred </p><p>ancestry proportions do not represent an admixed ancestry composed of multiple modern </p><p>populations, as SP1060 was alive at approximately the time of their divergence 20. Rather, </p><p>they reflect ancestral genetic variation in SP1060, which later segregated in the different </p><p>populations 13,20. </p><p>6 </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>The PCA and ancestry results were further confirmed by the sharing of derived alleles using </p><p>D-statistic tests 21 of the form D(H1,H2; SP1060,Orca). The ancient sample shares a significant </p><p>excess of derived alleles with both ENA and WNA coastal populations, compared with all other </p><p>in-group populations (Figure S10a). The value of the statistic D(ENAc,WNAc; SP1060,Orca) is </p><p>significantly negative (Z-scores of -6.9 to -8.8), indicating that SP1060 is more closely related </p><p>to the ENA coastal dolphins than to the WNA coastal dolphins. Accordingly, SP1060 shares a </p><p>higher excess of derived alleles with the ENA coastal population than with the WNA coastal </p><p>population, in statistics of the form D(coastal,pelagic; SP1060,Orca). </p><p>Having established the broad relationship of ancient samples to modern populations, in terms </p><p>of shared ancestry, we next sought to reconstruct evolutionary history through time as an </p><p>admixture graph 22. Testing across all possible histories, we find one admixture graph with no </p><p>outlier f-statistics using qpBrute (Figure 2), which explores the space of all possible admixture </p><p>graphs of a given maximum complexity, under a brute-force approach 23,24. Graphs were </p><p>estimated using pseudo-haploid data mapped to the killer whale reference genome and using </p><p>the killer whale as an outgroup. We find similar topologies when using called genotypes for </p><p>the modern populations only and mapping to the bottlenose dolphin reference genome </p><p>(Figure S11). </p><p>The best fitting graph reveals a basal split between the lineage that gave rise to the modern </p><p>North Atlantic coastal populations and SP1060, and the lineage that gave rise to the majority </p><p>of ancestry in modern pelagic populations (Figure 2). The WNA coastal, which has recently </p><p>been described as its own species, Tursiops erebennus 25, and SP1060 are modelled as </p><p>independent lineages. The ENA coastal is modelled as a clade whose ancestry is a mixture of </p><p>the two ancestral groups leading to SP1060 and WNA coastal. Ancient sample SP1060 is </p><p>therefore not a direct ancestor of the modern ENA coastal dolphins. The ancestry of both </p><p>North Atlantic modern pelagic populations is modelled as an admixture of approximately 30% </p><p>of the lineage giving rise to the coastal populations, and approximately 70% from a deeply </p><p>divergent lineage. Thus, the model provides a useful visualisation of how coastal-associated </p><p>alleles may have been reintroduced into pelagic populations. The branches leading to the </p><p>pelagic populations have null or small drift values, consistent with pelagic populations having </p><p>7 </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>large ancestral effective population sizes or showing little genetic structure, and indicating </p><p>minimal drift from a shared ancestral population, consistent with previous demographic </p><p>inference 5,26,27. </p><p>Figure 2. Evolutionary relationships between the ancient individual SP1060 and the North Atlantic modern bottlenose dolphin populations. Solid arrows indicate the relationships between populations/samples and the numbers at their right side the estimated genetic drift represented by the arrow. Populations include eastern North Atlantic coastal (ENAc) and pelagic (ENAp) populations, and western North Atlantic coastal (WNAc) and pelagic (WNAp) populations, and the outgroup is the killer whale (KW). Note that the drift value for SP1060 is inflated due to being a single and lower coverage-sample. Dashed lines are admixture edges and the arrows indicate the inferred direction of admixture, with the numbers reflecting the percentage of ancestry deriving from each lineage. </p><p>8 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Patterns of selection to coastal habitat in the ancient individual </p><p>Coastal-associated genotypes at sites inferred as evolving under parallel linked selection in </p><p>coastal populations 5 are also found in the ancient sample SP1060 (Figures 3a and b, S12). This </p><p>observation informs us about the speed of selection. While we observe drift from the pelagic </p><p>populations in the coastal modern individuals, SP1060 and the other ancient samples are </p><p>genetically less diverged from the pelagic populations (Figure 1c). Yet, the genome of SP1060 </p><p>shares the coastal associated variation inferred to have evolved under parallel linked selection </p><p>in the coastal populations 5 (Figure 3a and b). The ancient genome shows excess </p><p>heterozygosity as found in coastal individuals 5, despite its lower coverage (Figure S12). As </p><p>SP1060 is dated to close or shortly after the colonisation time of coastal waters by bottlenose </p><p>dolphins in the North Sea 8,9, this suggests selection occurred rapidly by acting upon standing </p><p>genetic variation. </p><p>Combining modern and ancient samples, we also provide insight on the independence of </p><p>parallel linked selection in coastal bottlenose dolphins from standing genetic variation </p><p>present in pelagic populations. Selection would be considered independent if it acted upon </p><p>standing genetic variation in each derived population (scenario i) 4, that is along branches </p><p>which are not shared among the two North Atlantic coastal populations, and SP1060 in Figure </p><p>2. It would not be independent if selection acted upon standing genetic variation in a shared </p><p>ancestral population (scenario ii) 4, that is along branches shared among the coastal </p><p>populations, or if adaptive standing genetic variation was shared post-selection among </p><p>coastal populations through gene flow (scenario iii) 4. </p><p>To identify which of these three scenarios best fits our data, we compare patterns of variation </p><p>(Figure S12) and the neighbour-joining tree (Figure 3b) obtained for our populations at the </p><p>sites under parallel linked selection with the predictions under each scenario made by Lee </p><p>and Coop 2019 4. For the SNPs under parallel linked selection, the coastal individuals cluster </p><p>by population, but with inter-individual variation 5. SP1060 clusters most closely with the ENA </p><p>coastal population (Figure 3a), but diverges close to the basal node of the Atlantic coastal </p><p>populations (Figure 3b). Such clustering of individuals by population is the pattern expected </p><p>under scenario i), which describes an independent selection from standing genetic variation </p><p>in each population, as it will generate only partially shared haplotypes. In contrast, scenarios </p><p>9 </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>ii) and iii) would have generated shared haplotypes among populations, and individuals from </p><p>the North Atlantic coastal populations would be mixed in the tree. While the Lee and Coop </p><p>approach is based on local trees, we based our analysis on different regions of the genome, </p><p>thus assuming that the different regions fit the same selection scenarios. We compared the </p><p>generated tree with ten trees generated using the same number of randomly sampled neutral </p><p>SNPs. The trees do not show a clustering by ecotypes nor strong differences in branch length </p><p>between coastal and pelagic individuals (Figure S13). </p><p>We observe long branches in the neighbour-joining tree for all the coastal individuals (Figure </p><p>3b). We previously identified these regions as representing old variants (0.6 to 2.3 million </p><p>years); based upon the excess of mutations that had accumulated within them, compared </p><p>with the genome-wide average 5. Given the findings of coastal-associated variants in the </p><p>genome of SP1060, despite low drift from the pelagic population, we hypothesize that there </p><p>was an abundance of coastal-associated standing genetic variation in the ancestral source </p><p>population; for example, through a large influx of standing genetic variation from coastal </p><p>populations in refugia, such as the Mediterranean Sea. </p><p>Conclusion </p><p>The goal of this study was to disentangle the mode of parallel selection through direct </p><p>observations of the chronology and genetic changes associated with the formation of coastal </p><p>ecotype populations of bottlenose dolphins. Using subfossil samples pre-dating much of the </p><p>drift experienced by coastal populations, we have disentangled neutral and selective </p><p>processes. Overall, our results suggest that coastal variants represent balanced </p><p>polymorphisms that were rapidly and repeatedly sieved from standing variation by ecological </p><p>selection. We thereby provide rare direct evidence of rapid adaptation to newly emerged </p><p>habitat from standing genetic variation. </p><p>10 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Figure 3. Patterns of genetic variation of the SNPs under repeated selection to coastal habitat in modern common bottlenose dolphin individuals and ancient individual SP1060. These include 2,122 SNPs with no missing data in SP1060 out of the 7,165 SNPs identified in Louis et al. (2021) 5, mapped to the bottlenose dolphin reference genome. Populations include coastal and pelagic ecotypes from the eastern North Atlantic (ENAc and ENAp), western North Atlantic (WNAc and WNAp) and eastern North Pacific (ENPc and ENPp). a) Principal component analysis and (b) Neighbour-joining distance tree showing the genetic structure of the common bottlenose dolphin samples for this particular SNP set. </p><p>Methods </p><p>Detailed methods are provided in the supplementary material. </p><p>Author contributions </p><p>ML, MN and ADF conceived the study; FA, SB, AB, ER, JOB, KP, PER, HvdE collected or </p><p>curated samples/specimens, ML, PK, MN, MHSS, NW and ADF ran the laboratory work; </p><p>MHSS and NW provided guidance in the ancient lab; ML and MN analysed the data; all </p><p>authors interpreted the data; MCF, OEG, and ADF supervised the work, ML and ADF wrote </p><p>the manuscript, with input from all co-authors. All authors read and approved the final </p><p>manuscript. </p><p>11 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is\"}\n",
      "{\"text\": \"made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>Acknowledgements </p><p>We especially thank Matthias Meyer for providing the SP1060 library, which allowed the </p><p>generation of the ancient genome that was central to this study. We thank Ramon Fallon, </p><p>Joseph Ward and Peter Thorpe from the St Andrews Bioinformatics Unit for help with </p><p>bioinformatics. Bioinformatics and computational analyses were supported by the University </p><p>of St Andrews Bioinformatics Unit which is funded by a Wellcome Trust ISSF award [grant </p><p>105621/Z/14/Z] and ran on cluster marvin and Crop Diversity HPC. We thank Kelly Roberston </p><p>for the laboratory work for the modern samples from the USA and arranging shipment. We </p><p>thank M. Thomas P. Gilbert for our use of the ancient DNA facilities in Copenhagen. We thank </p><p>everyone involved in sample collection including Conor Ryan for sampling some of the </p><p>stranded individuals in Ireland, the stranding networks in Ireland and Scotland and the Groupe </p><p>d\\u2019Etude des C\\u00e9tac\\u00e9s du Cotentin, and the NOAA Fisheries field crews and Simon Ingram for </p><p>biopsy sample collection. </p><p>Funding </p><p>Funding for ancient DNA labwork and sequencing was provided by the Total Foundation </p><p>awarded to ML. Funding for sample collection in Ireland was provided by Science Foundation </p><p>Ireland. Funding for labwork and visits to the University of Copenhagen was provided by the </p><p>Systematics Research Fund, a Godfrey Hewitt mobility award from the European Society for </p><p>Evolutionary Biology (ESEB), Lerner-Gray Grants for Marine Research. ML was supported by a </p><p>Fyssen Fellowship, Total Foundation, the University of St Andrews and the Greenland </p><p>Research Council. Modern sample DNA extractions were supported by People\\u2019s Trust for </p><p>Endangered Species and the sequencing costs were supported by the European Union\\u2019s </p><p>Horizon 2020 research and innovation program under the Marie Sk\\u0142odowska-Curie grant </p><p>agreement No. 663830 awarded to A.D.F., by the Total Foundation awarded to M.L., the </p><p>University of Groningen awarded to M.C.F., and the Marine Alliance for Science and </p><p>Technology for Scotland and The Russell Trust awarded to O.E.G. M.N. was funded by MASTS </p><p>and the Crawford Hayes fund. A.D.F. was funded by Marie Sk\\u0142odowska-Curie grant agreement </p><p>No. 663830 and the European Research Council grant agreement No. ERC-COG-101045346. </p><p>12 </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>References </p><p>1. Barrett, R.D.H., and Schluter, D. (2008). Adaptation from standing genetic variation. </p><p>Trends Ecol. Evol. 23, 38\\u201344. </p><p>2. Waters, J.M., and McCulloch, G.A. (2021). Reinventing the wheel? Reassessing the roles of gene flow, sorting and convergence in repeated evolution. Mol. Ecol. 30, 4162\\u20134172. </p><p>3. Roesti, M., Gavrilets, S., Hendry, A.P., Salzburger, W., and Berner, D. (2014). The </p><p>genomic signature of parallel adaptation from shared genetic variation. Mol. Ecol. 23, 3944\\u20133956. </p><p>4. </p><p>5. </p><p>Lee, K.M., and Coop, G. (2019). Population genomics perspectives on convergent adaptation. Philos. Trans. R. Soc. Lond. B Biol. Sci. 374, 20180236. </p><p>Louis, M., Galimberti, M., Archer, F., Berrow, S., Brownlow, A., Fallon, R., Nyk\\u00e4nen, M., O\\u2019Brien, J., Roberston, K.M., Rosel, P.E., et al. (2021). Selection on ancestral genetic variation fuels repeated ecotype formation in bottlenose dolphins. Sci Adv 7, eabg1245. </p><p>6. Rosel, P.E., Hansen, L., and Hohn, A.A. (2009). Restricted dispersal in a continuously </p><p>distributed marine species: common bottlenose dolphins Tursiops truncatus in coastal waters of the western North Atlantic. Mol. Ecol. 18, 5030\\u20135045. </p><p>7. </p><p>8. </p><p>Lowther-Thieleking, J.L., Archer, F.I., Lang, A.R., and Weller, D.W. (2015). Genetic differentiation among coastal and offshore common bottlenose dolphins, Tursiops truncatus, in the eastern North Pacific Ocean. Mar. Mamm. Sci. 31, 1\\u201320. </p><p>Louis, M., Fontaine, M.C., Spitz, J., Schlund, E., Dabin, W., Deaville, R., Caurant, F., Cherel, Y., Guinet, C., and Simon-Bouhet, B. (2014). Ecological opportunities and specializations shaped genetic divergence in a highly mobile marine top predator. Proc. Biol. Sci. 281. 10.1098/rspb.2014.1558. </p><p>9. Nyk\\u00e4nen, M., Kaschner, K., Dabin, W., Brownlow, A., Davison, N.J., Deaville, R., Garilao, </p><p>C., Kesner-Reyes, K., Gilbert, M.T.P., Penrose, R., et al. (2019). Postglacial colonization of northern coastal habitat by bottlenose dolphins: a marine leading-edge expansion? J. Hered. 110, 662\\u2013674. </p><p>10. Shennan, I., Lambeck, K., Flather, R., Horton, B., McArthur, J., Innes, J., Lloyd, J., </p><p>Rutherford, M., and Wingfield, R. (2000). Modelling western North Sea palaeogeographies and tidal changes during the Holocene. Geological Society, London, Special Publications 166, 299\\u2013319. </p><p>11. Post, K. (2005). A Weichselian marine mammal assemblage from the southern North </p><p>Sea. Deinsea 11, 21\\u201328. </p><p>12. Aaris-S\\u00f8rensen, K. (2010). Diversity and dynamics of the mammalian fauna in Denmark </p><p>throughout the last glacial-interglacial cycle, 115-0 kyr BP (John Wiley & Sons). </p><p>13 </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>13. Fran\\u00e7ois, O., and Jay, F. (2020). Factor analysis of ancient population genomic samples. </p><p>Nat. Commun. 11, 4661. </p><p>14. Patterson, N., Price, A.L., and Reich, D. (2006). Population structure and eigenanalysis. </p><p>PLoS Genet. 2, e190. </p><p>15. Korneliussen, T.S., Albrechtsen, A., and Nielsen, R. (2014). ANGSD: Analysis of Next </p><p>Generation Sequencing Data. BMC Bioinformatics 15, 356. </p><p>16. G\\u00fcnther, T., and Nettelblad, C. (2019). The presence and impact of reference bias on population genomic studies of prehistoric human populations. PLoS Genet. 15, e1008302. </p><p>17. Martiniano, R., Garrison, E., Jones, E.R., Manica, A., and Durbin, R. (2020). Removing </p><p>reference bias and improving indel calling in ancient DNA data analysis by mapping to a sequence variation graph. Genome Biol. 21, 250. </p><p>18. Schubert, M., Ginolhac, A., Lindgreen, S., Thompson, J.F., Al-Rasheid, K.A.S., Willerslev, E., Krogh, A., and Orlando, L. (2012). Improving ancient DNA read mapping against modern reference genomes. BMC Genomics 13, 178. </p><p>19. Foote, A.D., Liu, Y., Thomas, G.W.C., Vina\\u0159, T., Alf\\u00f6ldi, J., Deng, J., Dugan, S., van Elk, </p><p>C.E., Hunter, M.E., Joshi, V., et al. (2015). Convergent evolution of the genomes of marine mammals. Nat. Genet. 47, 272\\u2013275. </p><p>20. Lawson, D.J., van Dorp, L., and Falush, D. (2018). A tutorial on how not to over-interpret </p><p>STRUCTURE and ADMIXTURE bar plots. Nat. Commun. 9, 3258. </p><p>21. Durand, E.Y., Patterson, N., Reich, D., and Slatkin, M. (2011). Testing for ancient admixture between closely related populations. Mol. Biol. Evol. 28, 2239\\u20132252. </p><p>22. Patterson, N., Moorjani, P., Luo, Y., Mallick, S., Rohland, N., Zhan, Y., Genschoreck, T., Webster, T., and Reich, D. (2012). Ancient admixture in human history. Genetics 192, 1065\\u20131093. </p><p>23. N\\u00ed Leathlobhair, M., Perri, A.R., Irving-Pease, E.K., Witt, K.E., Linderholm, A., Haile, J., Lebrasseur, O., Ameen, C., Blick, J., Boyko, A.R., et al. (2018). The evolutionary history of dogs in the Americas. Science 361, 81\\u201385. </p><p>24. Liu, L., Bosse, M., Megens, H.-J., Frantz, L.A.F., Lee, Y.-L., Irving-Pease, E.K., Narayan, G., Groenen, M.A.M., and Madsen, O. (2019). Genomic analysis on pygmy hog reveals extensive interbreeding during wild boar expansion. Nat. Commun. 10, 1992. </p><p>25. Costa, A.P.B., Mcfee, W., and Wilcox, L.A. The common bottlenose dolphin (Tursiops truncatus) ecotypes of the western North Atlantic revisited: an integrative taxonomic investigation supports the presence of distinct species Zool. J. Linn. Soc., zlac025, 10.1093/zoolinnean/zlac025/6585199. </p><p>26. Peter, B.M. (2022). A geometric relationship of F 2 , F 3 and F 4 -statistics with principal </p><p>14 </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2022.11.03.515020 </p><p>; </p><p>this version posted November 4, 2022. </p><p>The copyright holder for this </p><p>preprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in </p><p>perpetuity. It is made available under a </p><p>CC-BY-NC-ND 4.0 International license . </p><p>component analysis. Philosophical Transactions of the Royal Society B: Biological Sciences 377. 10.1098/rstb.2020.0413. </p><p>27. McVean, G. (2009). A genealogical interpretation of principal components analysis. </p><p>PLoS Genet. 5, e1000686. </p><p>15 </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div></body></html>\"}\n",
      "{\"text\": \"<html><body><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>SafeMut: UMI-aware variant simulator </p><p>incorporating </p><p>allele-fraction overdispersion in read editing </p><p>Xiaofei Zhao1,*, Jingyu Guo1, and Sizhen Wang1 1. Genetron Health (Beijing) Co. Ltd, Beijing 102208, China Tuesday, March 14, 2023 </p><p>Abstract </p><p>Next-generation sequencing (NGS) has been widely used for calling biological variants. The </p><p>gold-standard methodology for accessing the ability of a computational method to call a specific </p><p>variant is to perform NGS wet-lab experiments on samples known to harbor this variant. Nevertheless, </p><p>wet-lab experiments are both labor-intensive and time-consuming, and rare variants may not be present </p><p>in a sample of population. Moreover, these two issues are exacerbated in SafeSeqS which enabled </p><p>liquid biopsy and minimum-residual disease (MRD) detection with cell-free DNA by using unique </p><p>molecular identifier (UMI) to detect and/or correct NGS error. Hence, we developed the first </p><p>UMI-aware NGS small-variant simulator named SafeMut which also considered the overdispersion of </p><p>allele fraction. We used the tumor-normal paired sequencing runs from the SEQC2 somatic reference </p><p>sets and cell-free DNA data sets to assess the performance of BamSurgeon, VarBen, and SafeMut. We </p><p>observed that, unlike BamSurgeon and VarBen, the allele-fraction distribution of the variants simulated </p><p>by SafeMut closely resembles such distribution generated by technical replicates of wet-lab </p><p>experiments. SafeMut is able to provide accurate simulation of small variants in NGS data, thereby </p><p>helping with the assessment of the ability to call these variants in a bioinformatics pipeline. </p><p>Introduction </p><p>Nowadays, next-generation sequencing (NGS) is widely used. NGS is especially used in clinical </p><p>settings to detect cancer mutations. To ensure that we can reliably detect a given mutation, the </p><p>gold-standard method is to perform multiple wet-lab NGS experiments. However, such method is </p><p>costly, labor-intensive and time-consuming. Thus, computational methods have been developed to </p><p>simulate variants in NGS data. The first class of methods, such as ART [1], simulates NGS reads from </p><p>only the human reference genome without using any pre-existing NGS data. Such methods require </p><p>neither biological samples nor wet-lab experiments but neither accurately model biological variation </p><p>nor consider experiment-specific error profiles </p><p>[2, 3]. Systematic reviews of the first class of methods </p><p>have been provided [4, 5]. The second class of methods, such as the one used to evaluate the somatic </p><p>variant caller Mutect [6], mixes in silico part of tumor with part of its corresponding paired normal to </p><p>model tumor with lower variant allele fraction (VAF). Such methods can model experiment-specific </p><p>error profiles and are trivial to implement, but the repertoire of spiked-in mutations is limited to </p><p>examples detected previously, and thus already known to be discoverable [2, 3]. Additionally, </p><p>biological variations, such as copy-number variation [7] and variation in the quantity of originally </p><p>sampled DNA molecules [8], intrinsically exist in tumors. These variations unfortunately hinder the </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>determination of the expected ground-truth allele fractions to simulate. The third class of methods, such </p><p>as BamSurgeon and VarBen [2, 3], spikes-in any user-defined mutations to pre-existing sequencing </p><p>data. Such methods can model both biological variation and experiment-specific error profiles. </p><p>However, biological variation is also affected by some experimental errors which are not modeled. </p><p>Moreover, some new wet-lab methods have been developed to reduce NGS errors. In the Safe </p><p>Sequencing System (SafeSeqS), molecular barcodes, also known as UMIs (unique molecular </p><p>identifiers), are attached to the original cell-free DNA (cfDNA) molecules in the preparation of </p><p>libraries, and the libraries are then sequenced at high depth [9]. Thus, we can observe a mutation in </p><p>cancer cells as a variant supported by UMI families, where each UMI-family is almost entirely made of </p><p>fragments supporting only this variant [9]. A low quantity of the DNA released by cancer cells is </p><p>present in cfDNA in the form of circulating tumor DNA (ctDNA), so SafeSeqS has been widely used in </p><p>liquid biopsies based on ctDNA assays [10]. Additionally, SafeSeqS has been used in the detection of </p><p>minimum residual disease (MRD), which is essential for providing adequate treatment to cancer </p><p>patients [11]. Nowadays, UMI-Gen, which belongs to the first class of methods, is the only UMI-aware </p><p>NGS-variant simulator [12], so the second and third classes of UMI-aware NGS variant simulators are </p><p>still lacking. Additionally, duplex sequencing, which attaches two UMIs to each original DNA </p><p>molecule, can suppress NGS errors even further [13]. Hence, duplex sequencing has numerous </p><p>applications [14-16]. However, there is no duplex-aware NGS variant simulator yet. </p><p>Furthermore, the methodology for evaluating NGS variant simulators has not been well </p><p>established. Methods belonging to the first class directly simulate NGS data. Thus, researchers </p><p>compared real NGS data with simulated NGS data in terms of quality-control (QC) metrics such as </p><p>genome coverage and GC bias [17]. This QC-metrics comparison measures the resemblance of </p><p>simulated data to real data but does not measure the resemblance of simulated variants to real variants </p><p>observed in the data. Some researchers compared simulated data and real data in terms of correlation </p><p>with gold-standard VAFs [3]. Nonetheless, VAFs from simulated data may be better correlated with </p><p>gold-standard VAFs because simulation may introduce less noise than real NGS experiments, and the </p><p>correlation between observed and expected values does not measure the similarity between the </p><p>observed and expected distributions generating these values. </p><p>Thus, we proposed the application of a hash-based pseudorandom number generator (PRNG) on </p><p>UMIs to enable UMI-aware simulation of variants and the use of a log-normal distribution to model the </p><p>overdispersion observed in the fluctuation of VAFs. We also proposed the methodology of using scatter </p><p>plots of the mean and variance and of using Q-Q plots of the z-scores to measure the similarity, </p><p>observed from sequencing data and in terms of VAF, between variants derived from in silico simulation </p><p>and variants derived from wet-lab experiments. </p><p>Importantly, we developed an NGS variant simulator named SafeMut. In short, SafeMut spikes-in </p><p>any user-specified small variant (i.e., SNV and Indel) from a given VCF file into a BAM file which is </p><p>not supposed to contain this variant. Thus, according to the before-mentioned classification, SafeMut </p><p>belongs to the third class of methods. Remarkably, the distribution of VAFs simulated by SafeMut is </p><p>approximately identical to the distribution of VAFs observed from real sequencing data. And SafeMut </p><p>is aware of the UMIs, including duplex UMIs, in its input BAM file so that each simulated variant is </p><p>spiked-in to UMI families instead of NGS fragments. Also, the running time and random-access samtools </p><p>memory usages of SafeMut are similar to that of the fast and memory-efficient command \\u201c view </p><p>\\u201d which is ubiquitously used in NGS to process SAM/BAM files [18], so SafeMut runs fast and is </p><p>efficient in terms of its use of computational resources. </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Approach </p><p>In brief, we discovered two key insights that can enable the editing of aligned reads, with </p><p>UMI-awareness and allele-fraction overdispersion, in only one pass over an input BAM file. Our key </p><p>insights are depicted in (Fig. 1). </p><p>First, if a pseudorandom number generator (PRNG) uses the same seed to generate multiple series </p><p>of pseudorandom numbers, then these series are identical to each other. In a typical NGS dataset, reads </p><p>derived from the same original DNA molecule have the same start position, end position and UMI tag. </p><p>Hence, if we derive randomness by using a combination of start position, end position and UMI tag of </p><p>reads, then reads derived from the same original DNA molecule should be associated with the same </p><p>random number that can be normalized between zero and one. In this way, we can spike-in a mutation </p><p>to all these reads grouped together so that the variant occurs at the level of the original DNA molecules. </p><p>Thereby, PRNG using UMI as its seed can enable UMI-aware simulation of variant. </p><p>Second, allele ratio (the number of fragments A supporting the ALT mutant variant allele, divided </p><p>by the number of fragments B supporting all non-ALT alleles) is known to follow a log-normal </p><p>A/B (cid:4) 1 </p><p>10(cid:0)(cid:2)(cid:3)/(cid:2)(cid:5)(cid:6) </p><p>distribution [19]. Additionally, the event </p><p>A/B (cid:4) 2 </p><p> is known to be </p><p> times more likely than </p><p>the event </p><p> [19]. Hence, we used such lognormal distribution to simulate the overdispersion in </p><p>the fluctuation of variant allele fractions (VAFs). </p><p>The two above-mentioned techniques both require only one pass over the reads found in the input </p><p>BAM file to simulate all given variants. Hence, our algorithm is supposed to be fast and memory </p><p>efficient. </p><p>Algorithm </p><p>Our algorithm consists of the following main steps. </p><p>1. For each read in the sorted input BAM file: </p><p>a) </p><p>Iterate through the input VCF file to discard variants before the start position of this read. </p><p>b) Select all remaining variants that overlap with this read. </p><p>(cid:8) </p><p>c) Compute the hash value H </p><p> made of the UMI sequence, insert start position and insert 2(cid:7)(cid:8) </p><p>2(cid:7)(cid:8) </p><p>end position. Compute </p><p>U </p><p> modulo </p><p>, then divide by </p><p> to get a pseudorandom real </p><p>number </p><p> between zero and one. If this read is not associated with any UMI, then use the </p><p>H </p><p>read query name as the UMI sequence. The hash value </p><p> is subsequently used as the seed </p><p>for generating random variable. </p><p>d) For each variant that overlaps with this read: </p><p>V~logN(cid:17)\\u00b5, (cid:20)(cid:7)(cid:21) F </p><p>i. Multiply the allele ratio R by a lognormal random variable </p><p>R (cid:4) F/(cid:17)1 (cid:25) F(cid:21) </p><p>R\\u2019 </p><p>(cid:8) </p><p>generated by using </p><p> as the seed to obtain </p><p>, where </p><p>, </p><p> is the </p><p>theoretical variant allele fraction (VAF) of this simulated variant specified in the input V (cid:4) 2 (cid:26) exp (cid:17)\\u00b5(cid:21) </p><p>\\u00b5 (cid:4) 0 </p><p>(cid:20) </p><p>VCF file, </p><p>, and where </p><p> is set such that the likelihood that 10(cid:9)(cid:2)(cid:3)/(cid:2)(cid:5) V (cid:4) exp(cid:17)\\u00b5(cid:21) </p><p>F\\u2019 </p><p>, </p><p>R\\u2019/(cid:17)1 (cid:31) R\\u2019(cid:21) divided by the likelihood that </p><p>, is </p><p>. Then, let the new VAF </p><p> be </p><p> ! \\\"\\u2019 </p><p>. </p><p>ii. </p><p>If </p><p>, then spike-in the variant into the read. Otherwise, do not do anything. </p><p>e) Output the read with and/or without the spike-in variant(s) into FASTQ files(s). </p><p>Allele ratio is defined as the number of fragments supporting the ALT allele divided by the number of </p><p>fragments supporting all non-ALT (including REF) alleles. Each above step emphasized with bold font </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>is described in more detail as follows. </p><p>(cid:8) </p><p>Compute the hash value </p><p> made of the UMI sequence, insert start position and insert end </p><p>position </p><p>H </p><p>The hash value </p><p> of the information that uniquely identifies the original molecule is computed as </p><p>follows. </p><p>H (cid:4) hash(cid:17)hashString(cid:17)us(cid:21) xor hash(cid:17)c xor hash(cid:17)b xor hash(cid:17)prngs(cid:21)(cid:21)(cid:21) </p><p>us </p><p>b </p><p>c </p><p>where </p><p> is the UMI sequence in string format, prngs </p><p> is the insert begin position, </p><p> is the insert end </p><p>xor </p><p>position, </p><p> is a pseudorandom number generator (PRNG) seed, and the operation </p><p> is </p><p>exclusive or. </p><p>Different reads with the same UMI, insert start position and insert end position all have the same </p><p>hash value H and thus the same pseudorandom number generated for them. Hence, different reads </p><p>derived from the same original DNA molecule should be either all mutated to the same allele or all not </p><p>mutated at all if such molecule is observed in sequencing data. Therefore, each simulated variant adds </p><p>mutation to the original molecules inferred from the sequenced reads, which ensures that reads derived </p><p>from the same original molecule inferred from sequencing data all represent the same allele at each </p><p>locus. </p><p>. </p><p>Multiply the allele ratio </p><p>F by a lognormal random variable </p><p>The input VCF file specifies the VAF </p><p> to simulate for each variant. In sum, our algorithm </p><p>F\\u2019 </p><p>F </p><p>introduces some multiplicative stochastic error into // (cid:7) </p><p>/ (cid:2) </p><p>log-normal distribution, where the event that / / (cid:7) (cid:2) </p><p>// (cid:7) </p><p>/ (cid:2) </p><p>(cid:4) 2 </p><p> to obtain\"}\n",
      "{\"text\": \"(cid:4) 1 </p><p>10(cid:0)(cid:2)(cid:3)/(cid:2)(cid:5)(cid:6) </p><p> such that the allele ratio follows a </p><p> is </p><p> times as likely as the event that </p><p> given two allele ratios </p><p> and </p><p>. This log-normal distribution has been discovered and </p><p>used in the literature [19, 22]. </p><p>F </p><p>R </p><p>R (cid:4) F/(cid:17)1 (cid:25) F(cid:21) </p><p>More specifically, the VAF </p><p>R </p><p> is used to calculate allele ratio </p><p>V </p><p> as follows: </p><p>. logN(cid:17)\\u00b5 (cid:4) 0, (cid:20)(cid:7)(cid:21) </p><p>Then, the allele ratio </p><p>\\u03c3 </p><p> is multiplied by a random variable </p><p> which is distributed as </p><p>logN(cid:17)\\u00b5 (cid:4) 0, (cid:20)(cid:7)(cid:21) </p><p>. </p><p>1 (cid:2) </p><p>1 (cid:7) </p><p>The parameter </p><p>1 (cid:2) </p><p> is chosen so that if /1 (cid:7) </p><p>(cid:4) 2 </p><p> and </p><p> are distributed as /1 (cid:7) </p><p>(cid:4) 1 </p><p>1 (cid:2) </p><p>10(cid:9)(cid:2)(cid:3)/(cid:2)(cid:5) </p><p>likelihood that </p><p>, divided by the likelihood that </p><p>, is </p><p>, then the </p><p> which is the default </p><p>value of the -q command-line parameter. If the parameter -q is set to be smaller than zero, then the </p><p>V </p><p>random variable </p><p> always generates one, so the log-normal transformation is disabled in this case. </p><p> ~ N(cid:17)0, 1(cid:21) </p><p>X </p><p>2 (cid:2) </p><p>, 2 (cid:7) </p><p>Next, we generate two standard normal distributions </p><p>. Let </p><p> be defined as </p><p>follows. </p><p>X </p><p>X(cid:17)s(cid:21) (cid:4) hash 4p xor hash 5tid xor 7hash(cid:17)s(cid:21)89: mod M , </p><p>p </p><p>tid </p><p>where </p><p> is the random variable of interest, </p><p> is the genomic position of the variant, </p><p>s </p><p> is the ID of </p><p>the reference sequence defined in the VCF file format specification [23], X </p><p>M </p><p>X </p><p> is the seed used to generate </p><p>, and </p><p> is the maximum possible hash value so that M </p><p> is normalized to be a real number between </p><p>zero and one after modulo by </p><p>. The function hash has been previously defined. Then, we generate </p><p>(cid:21) </p><p>(cid:21) </p><p>2 (cid:2) </p><p>(cid:4) X(cid:17)= (cid:2) </p><p>2 (cid:7) </p><p>(cid:4) X(cid:17)= (cid:7) </p><p>= (cid:2) </p><p>= (cid:7) </p><p>uniformly at random two integers: we let </p><p> and </p><p>, where </p><p> and </p><p> are by </p><p>default the sum of the first 50 and 500 hash values of the read names in the input BAM file, </p><p>X </p><p>respectively. The values 50 and 500 have been arbitrarily selected to introduce randomness into </p><p>Z </p><p>Then, we use Box-Muller transform to generate a standard normal random variable 2 (cid:7) </p><p> from </p><p> [24]. </p><p>L~exp(cid:17)\\u03c3 (cid:26) Z(cid:21) </p><p>. </p><p>2 (cid:2) and </p><p>Next, the lognormal random variable </p><p> is used as the multiplier for the allele ratio </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>R\\u2019 (cid:4) L @ R </p><p>R so that </p><p> is the new allele odds ratio, where </p><p>F\\u2019 (cid:4) R\\u2019/(cid:17)1 (cid:31) R\\u2019(cid:21). </p><p>F\\u2019 </p><p>\\u03c3 (cid:26) Z </p><p>N(cid:17)0, \\u03c3(cid:7)(cid:21) </p><p> is equivalent to </p><p>. Next, we </p><p>compute the new VAF </p><p> as follows: </p><p> ! \\\"B </p><p>If </p><p>, then spike-in the variant into the read </p><p>If the variant is a single-nucleotide variant (SNV), then we transform the SNV base-call quality </p><p>(BQ) into its corresponding sequencing-error probability. Next, we generate a pseudorandom number BQN </p><p> between zero and one as follows. </p><p>BQN (cid:4) (cid:10)(cid:11)(cid:12)(cid:10)(cid:13)(cid:10)(cid:11)(cid:12)(cid:10)(cid:14)(cid:10)(cid:11)(cid:12)(cid:10)(cid:15)(cid:16)(cid:17)(cid:18)(cid:7)(cid:19)(cid:11)(cid:17)(cid:20)(cid:16)(cid:17)(cid:18)(cid:0)(cid:2)(cid:21),(cid:8)(cid:6)(cid:23) (cid:25)(cid:26)(cid:19) (cid:10)(cid:11)(cid:12)(cid:10)(cid:27)(cid:18)(cid:19)(cid:16)(cid:17)(cid:28)(cid:0)(cid:17)(cid:6)(cid:29) (cid:25)(cid:26)(cid:19) (cid:0)(cid:30)(cid:6)(cid:31) (cid:26)(cid:20) ! , </p><p>! </p><p>M </p><p>n </p><p>p </p><p>where </p><p>BQN </p><p> is the read name, </p><p> is the read start position, and </p><p>BQN </p><p> is the maximum hash value to ensure </p><p>that </p><p> is between zero and one. If </p><p> is less than the sequencing-error probability determined </p><p>from its corresponding BQ, then we change the original base supporting the SNV into a random base </p><p>among A, C, G and T. Otherwise, we keep the original base. </p><p>U ! \\\"B </p><p>If the variant is an insertion/deletion and </p><p>, then the corresponding bases are always </p><p>inserted/deleted because insertion/deletion does not have any BQ so that the variant is always spiked-in. </p><p>The default BQ of each inserted base is 30. </p><p>Implementation </p><p>The full implementation of SafeMut is at https://github.com/genetronhealth/safesim/. SafeMut is </p><p>implemented with C++ using htslib version 1.11. SafeMut accepts one BAM file and one VCF file as </p><p>input and generates two sets of FASTQ files as output: tumor and normal FASTQ sets containing tumor </p><p>and normal reads, respectively, to enable the user to identify the tumor/normal origin of each read. The </p><p>fastq-sort utility from fastq-tools (https://github.com/dcjones/fastq-tools) is then used to sort each </p><p>output FASTQ file in the two FASTQ sets. Then, the tumor and normal FASTQ sets are merged into </p><p>one set of FASTQ files. Here, tumor/normal denote any biomaterials that are/aren\\u2019t characterized by </p><p>mutations of interest, respectively. For example, tumor/normal can also denote samples with/without </p><p>some de novo germline mutations, respectively. If the -q command-line parameter is set to a negative </p><p>number, then the aforementioned log-normal transformation of allele odds ratio is disabled. </p><p>Results </p><p>Overall evaluation procedure </p><p>We used two datasets published by the Food-and-Drug Administration (FDA)-led Sequencing </p><p>Quality Control phase 2 (SEQC2) consortium. These two datasets were produced by the whole-exome </p><p>sequencing (WES) of the HCC1395 and HCC1395BL cell lines [25] and by the cell-free DNA (cfDNA) </p><p>sequencing of a well characterized sample [26], respectively (Table S1). These two datasets are both </p><p>rigorously produced using detailed protocols and open to the general public allowing for permissive </p><p>usage. More importantly, each of these two datasets is generated by multiple tumor-normal paired </p><p>sequencing runs and both tumor and normal samples are sequenced at approximately the same depth. </p><p>Thus, we let each simulator spike-in mutations found in the tumor sample, as specified in the </p><p>gold-standard tumor variant VCF file, into its matched normal BAM file, to simulate the corresponding </p><p>tumor BAM. Then, we compared the distributions of variant allele fractions (VAFs) in the tumor BAM </p><p>files with the VAFs in the simulated BAM files. More specifically, for each variant locus, we first </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>compared the mean and variance of the tumor VAFs with the simulated VAFs. Then, we compared the </p><p>z-scores, which were normalized by mean and variance, of the tumor VAFs and simulated VAFs. More </p><p>bcftools mpileup </p><p>specifically, we ran the \\u201c </p><p>extract allele depths to compute VAFs. The \\u201c </p><p>\\u201d command and then used the FORMAT/AD tag to </p><p>bcftools mpileup </p><p>\\u201d command already filtered variant </p><p>read support by mapping quality (MQ), base quality (BQ) and base alignment quality (BAQ) in its </p><p>computation of allele depths, so our evaluation also indirectly compared simulated variants with real </p><p>variants in terms of MQ, BQ and BAQ. </p><p>We used samtools version 1.11 [27], bcftools version 1.11 [18], BWA MEM version 0.7.17 [28], </p><p>Picard version 2.26.10 (http://broadinstitute.github.io/picard/), Velvet version 1.2.10 [29], Exonerate </p><p>version 2.4.0 [30], Fgbio version 1.5.0 (https://github.com/fulcrum-genomics/fgbio), BamSurgeon </p><p>version 1.3 [2], VarBen commit 0f66e35 [3], and SafeMut version 0.1.5.0b4d0f8 in our evaluation. </p><p>We ran all software packages with their default parameters except that we initialized the seeds of </p><p>the pseudo-random number generators (PRNGs) to zero for BamSurgeon and VarBen. By default, </p><p>BamSurgeon and VarBen seed their pseudo-random number generators (PRNGs) with the current </p><p>system time, which leads to the loss of exact reproducibility of variant-simulation results. Hence, we let </p><p>BamSurgeon and VarBen seed their PRNGs with zero to ensure exact reproducibility without </p><p>compromising our evaluation. </p><p>Dataset-specific evaluation procedure </p><p>For the WES dataset, we used BWA MEM to align all reads to the GRCh38 human reference </p><p>genome because variant coordinates in the gold-standard VCF are based on GRCh38 [25]. Then, we </p><p>ran BamSurgeon, VarBen, and SafeMut to spike-in mutations into the normal BAM files which have </p><p>approximately the same depth as the corresponding tumor BAM files, generating simulated tumor-like </p><p>BAM files. Next, we compared the VAF distribution of the real tumor BAM files with respect to the </p><p>VAF distribution of the simulated tumor-like BAM files. A few variants have exactly zero reads </p><p>supporting the variant allele. Thus, every variant having zero read support for the variant allele in any </p><p>sample is discarded in our evaluation to prevent the issue of division by zero. </p><p>For the cfDNA dataset, we only used the data submitted by Illumina (ILM) and IDT, so we did not </p><p>use the data submitted by ROC, BRP and TFS because ROC only provided reads processed by </p><p>UMI-consensus, BRP did not provide UMI information, and TFS data generated runtime errors for </p><p>both BamSurgeon and VarBen. For the cfDNA dataset, we used BWA MEM to align all reads to the </p><p>HS37D5 human reference genome because variant coordinates in the gold-standard VCF are based on </p><p>HG19/HS37D5 [31]. Then, we assumed that all simulators are UMI-aware and thus used the </p><p>simulation-then-consensus evaluation strategy. In this strategy, we first ran BamSurgeon, VarBen and </p><p>SafeMut to spike-in mutations into the normal BAM files which have approximately the same depth as </p><p>the tumor BAM files, generating simulated tumor-like BAM files. Then, we used Fgbio to collapse the </p><p>sequenced reads into duplex consensus sequences (DCSs) representing the double-stranded original </p><p>DNA molecules. Next, we compared the DCS VAF distribution of the real tumor BAM files with </p><p>respect to the DCS VAF distribution of the simulated tumor-like BAM files. </p><p>The aforementioned simulation-then-consensus strategy provides comprehensive evaluation for </p><p>UMI-aware variant simulators. However, BamSurgeon and VarBen are not UMI-aware, so we also used </p><p>the consensus-then-simulation strategy as a work-around to let UMI-unaware simulators incorporate </p><p>the effect of the labeling of reads by UMIs in their simulations. In the consensus-then-simulation </p><p>strategy, we first used Fgbio to collapse sequenced reads into duplex consensus sequences (DCSs) </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>representing the double-stranded original DNA molecules, so both tumor and normal BAM files </p><p>contain DCSs instead of raw reads. Then, we ran BamSurgeon, VarBen and SafeMut to spike-in </p><p>mutations into the normal BAM files which have approximately the same depth as the tumor BAM </p><p>files, generating simulated tumor-like BAM files. Next, we compared the DCS VAF distribution of the </p><p>real tumor BAM files with respect to the DCS VAF distribution of the simulated tumor-like BAM files. </p><p>Results </p><p>We compared the performance of SafeMut with BamSurgeon and VarBen by measuring the </p><p>similarity between the observed and expected VAF means, VAF variances, and z-score distributions </p><p>using mean-square errors (MSEs) between the observed and expected means, variances, and </p><p>distribution quantiles. The observed data were obtained by simulating variants in the normal BAM files </p><p>and the expected data are directly in the corresponding tumor BAM. Unlike VAF means and variances, </p><p>z-scores were first sorted, and only then the sorted observed z-scores were compared with the sorted </p><p>expected z-scores, resulting in the generation z-score Q-Q plots. SafeMut overall performs the best </p><p>(Figs. 2 and 3). BamSurgeon was affected by a runtime error when simulating variants in the cell-free </p><p>DNA dataset so that the corresponding results are not available. </p><p>Overall, SafeMut is characterized by lower MSEs for all combinations of variant simulators </p><p>(BamSurgeon, VarBen and SafeMut), assay types (whole-exome sequencing (WES), IDT cfDNA assay </p><p>and </p><p>Illumina </p><p>(ILM) </p><p>cfDNA </p><p>assay, with both </p><p>cfDNA </p><p>assays </p><p>evaluated with </p><p>the </p><p>simulation-then-consensus strategy) and statistical measures (mean, variance, low-AF-variant z-score, </p><p>medium-AF-variant z-score and high-AF-variant z-score) except that BamSurgeon has lower MSE than </p><p>SafeMut in the scenario corresponding to the combination of WES and low-AF-variant z-score (Figs. 2 </p><p>and 3). Moreover, SafeMut is characterized by much smaller MSEs for fitting the linear relationships </p><p>between observed and expected variances because SafeMut is aware of the AF overdispersion in NGS </p><p>data. Furthermore, SafeMut strongly outperforms BamSurgeon and VarBen, which are not UMI-aware, </p><p>on the data generated by UMI-enabled targeted sequencing because SafeMut is UMI-aware (Figs. 2 </p><p>and 3). </p><p>We also evaluated using the cell-free DNA dataset with the consensus-then-simulation strategy </p><p>(Fig. 4). This strategy enables UMI-unaware simulators to incorporate the effect of the labeling of </p><p>reads by UMIs in their simulations. However, this strategy is still only a workaround because it </p><p>discards UMI information such as UMI-family size and UMI-family consensus percent identity. Thus, </p><p>this strategy disregards UMI-awareness, the advantage that SafeMut has compared with UMI-unaware </p><p>simulators such as BamSurgeon and VarBen. However, if the consensus-then-simulation strategy is </p><p>used, then SafeMut\"}\n",
      "{\"text\": \"still performs the best overall: SafeMut is characterized by the lowest MSEs for all </p><p>combinations of variant simulators (BamSurgeon, VarBen and SafeMut), assay types (IDT and Illumina </p><p>cfDNA assays) and statistical measures (mean, variance, low-AF-variant z-score, medium-AF-variant </p><p>z-score and high-AF-variant z-score) except that VarBen has lower MSEs than SafeMut in the </p><p>scenarios corresponding to the combinations of both IDT and ILM with mean and that BamSurgeon </p><p>has lower MSEs than SafeMut in the scenarios corresponding to the combinations of IDT with </p><p>low-AF-variant z-score and ILM with medium-AF-variant z-score (Fig. 4). </p><p>The reference truth AFs, real tumor-data AFs, and simulated tumor-data AFs generated by </p><p>different bioinformatics tools for each locus are presented at https://doi.org/10.5281/zenodo.7139579 </p><p>for each sequencing project as detailed results. Importantly, the detailed results show that SNVs and </p><p>InDels approximately follow the same AF distribution except that simulation by VarBen resulted in </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>over-dispersed InDel AFs but under-dispersed SNV AFs for cfDNA data. Hence, we evaluated SNVs </p><p>and InDels together instead of evaluating each variant type separately. </p><p>The running time (runtime) and random-access memory (RAM) consumed by SafeMut are samtools </p><p>approximately the same as the respective runtime and RAM consumed by the command \\u201c view -bh </p><p>\\u201d. And this command is known for its fast runtime and high memory-efficiency in NGS data </p><p>processing [18]. Therefore, SafeMut is fast and efficient in terms of its usage of computational </p><p>resources. </p><p>Discussion </p><p>The detection of small variants, namely SNVs and InDels, caused by clinically significant mutations </p><p>are crucial to the diagnosis, prognosis and treatment monitoring of cancer and heritable diseases. </p><p>Nevertheless, it is labor-intensive and time-consuming to perform wet-lab experiments to obtain the </p><p>real sequencing data needed for checking the ability of a pipeline to detect such variants. Furthermore, </p><p>some variants are extremely rare, so a large number of samples have to be collected and sequenced to </p><p>find the samples characterized by such rare variants, and only then we can check how such variants </p><p>look like in the sequencing data. Hence, variant simulators become especially important to tackle such </p><p>problems. Furthermore, the sequencing of cell-free DNA (cfDNA) with molecular barcodes is routinely </p><p>used in liquid biopsy and minimum residual disease (MRD) detection for cancer patients. As a result, </p><p>UMI-aware variant simulators are urgently needed, but such simulators have not been developed yet. </p><p> We designed and developed SafeMut which simulates small variants (i.e., SNVs and InDels) by </p><p>editing reads and then in silico spike-in the edited reads. SafeMut is applicable to data generated by </p><p>SafeSeqS and duplex sequencing. Additionally, SafeMut models allele-fraction overdispersion with a </p><p>specific log-normal distribution. We showed that SafeMut outperforms the state-of-the-art simulators </p><p>by assessing the similarity between real and simulated variants using two open-access datasets </p><p>published by the SEQC2 consortium. The variants simulated by SafeMut are extremely similar to real </p><p>variants in terms of their distributions of allele fractions. Thus, SafeMut is able to provide realistic in </p><p>silico variants which can be used to benchmark the performance for calling these variants. </p><p>Consequently, SafeMut has numerous applications in bioinformatics. For example, SafeMut can help </p><p>the user to establish the limit of blank, limit of detection, and limit of quantitation of a rare variant of </p><p>interest. </p><p>SafeMut edits reads in such a way that all variants are in the same haplotype, but sometimes </p><p>different variants are located in different haplotypes. Hence, the ability to simulate haplotypes may be </p><p>implemented in a future version of SafeMut. Additionally, SafeMut should be able to work with </p><p>third-generation sequencing (TGS) data without any modification whatsoever, but further evaluation is </p><p>needed to assess the performance of SafeMut on TGS. </p><p>Data and code availability </p><p>The datasets used to perform the evaluation presented in this work are all accessible to the general </p><p>public, with accessions listed in (Table S1). The code used to implement SafeMut is publicly available </p><p>(and free for non-commercial third-party usage) at https://github.com/genetronhealth/safesim with </p><p>commit 0b4d0f8. The code used to perform the evaluation presented in this work is publicly available </p><p>at https://github.com/genetronhealth/safesim-eval/ with commit 08c5e8c. </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Conflict of interest </p><p>The algorithms presented in this manuscript were applied for a patent in China. X.Z. and J.G. were </p><p>employed by Genetron Health. S.W. is one of the founders of Genetron Health. The authors declare no </p><p>other competing interests. </p><p>Funding </p><p>None declared. </p><p>Acknowledgments </p><p>The authors would like to thank Xiaoyue Wang for helping with the coordination between the authors. </p><p>Key Points </p><p>(cid:0) </p><p> By using UMI-aware generation of variant and by modeling allele-fraction overdispersion with a </p><p>specific log-normal distribution, we developed SafeMut, an algorithm for spiking-in small </p><p>variants into existing sequencing data. </p><p>(cid:0) </p><p>(cid:0) </p><p> The distribution of the allele fractions (AFs) of the variants that SafeMut spiked-in to the </p><p>tumor-matched normal samples closely resemble the distribution of the AFs of the real variants </p><p>directly observed in the corresponding tumor samples. </p><p> Traditionally, sample collections and wet-lab experiments are performed to generate sequencing </p><p>data of samples characterized by some pre-defined variants. SafeMut alternatively offers a more </p><p>cost-efficient, less labor-intensive and less time-consuming way to generate such sequencing data </p><p>by simulating these pre-defined variants in silico. </p><p> </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Figures </p><p>Figure 1: Pseudo-random generation of mutations by SafeMut using unique molecular identifier </p><p>(UMI)-aware in silico modification of reads and incorporating the overdispersion of allele fractions. </p><p> </p><p> </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Figure 2: Observed versus expected means (top), variances (mid), and z-scores (bottom) generated by </p><p>SafeMut (left), BamSurgeon (mid), and VarBen (right) for the HCC1395 whole-exome sequencing </p><p>y (cid:4) x (WES) reference standard dataset (accession SRP162370). SafeMut is characterized by the smallest </p><p>mean-square error (MSE) with respect to the </p><p> line in all plots except that BamSurgeon is </p><p>characterized by smaller MSE for the low-AF range z-score Q-Q plot. Low, medium and high AF </p><p>ranges are delimited by thresholds of 0.3 and 0.5. </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Figure 3: Observed versus expected means (top), variances (mid), and z-scores (bottom) generated by </p><p>SafeMut (left) and VarBen (right) for the cell-free DNA reference standard dataset (accession </p><p>SRP296025) evaluated with the simulate-then-consensus strategy. BamSurgeon was affected by </p><p>runtime error and thus did not generate any result. SafeMut is characterized by the smallest </p><p>y (cid:4) x </p><p>mean-square error (MSE) with respect to the </p><p> line in all plots. Low, medium and high AF ranges </p><p>are delimited by thresholds of 0.008 and 0.016. </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Figure 4: Observed versus expected means (top), variances (mid), and z-scores (bottom) generated by </p><p>SafeMut (left), BamSurgeon (middle) and VarBen (right) for the cell-free DNA reference standard </p><p>dataset (accession SRP296025) evaluated with the consensus-then-simulate strategy. Low, medium and </p><p>high AF ranges are delimited by thresholds of 0.008 and 0.016. </p><p> </p><p> </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>Reference </p><p>1. </p><p>Huang, W., et al., ART: a next-generation sequencing read simulator. Bioinformatics, </p><p>2012. 28(4): p. 593-594. </p><p>2. </p><p>Ewing, A.D., et al., Combining tumor genome simulation with crowdsourcing to </p><p>benchmark somatic single-nucleotide-variant detection. Nature methods, 2015. 12(7): </p><p>p. 623-630. </p><p>3. </p><p>Li, Z., et al., VarBen: Generating in silico reference data sets for clinical </p><p>next-generation sequencing bioinformatics pipeline evaluation. The Journal of </p><p>Molecular Diagnostics, 2021. 23(3): p. 285-299. </p><p>4. </p><p>Alosaimi, S., et al., A broad survey of DNA sequence data simulation tools. Briefings in </p><p>functional genomics, 2020. 19(1): p. 49-59. </p><p>5. </p><p>Zhao, M., D. Liu, and H. Qu, Systematic review of next-generation sequencing </p><p>simulators: computational tools, features and perspectives. Briefings in Functional </p><p>Genomics, 2016. 16(3): p. 121-128. </p><p>6. </p><p>Cibulskis, K., et al., Sensitive detection of somatic point mutations in impure and </p><p>heterogeneous cancer samples. Nature biotechnology, 2013. 31(3): p. 213-219. </p><p>7. </p><p>Zack, T.I., et al., Pan-cancer patterns of somatic copy number alteration. Nature </p><p>genetics, 2013. 45(10): p. 1134-1140. </p><p>8. </p><p>Huang, R.S., et al., Circulating cell-free DNA yield and circulating-tumor DNA quantity </p><p>from liquid biopsies of 12 139 cancer patients. Clinical Chemistry, 2021. 67(11): p. </p><p>1554-1566. </p><p>9. </p><p>Kinde, I., et al., Detection and quantification of rare mutations with massively parallel </p><p> </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>sequencing. Proceedings of the National Academy of Sciences, 2011. 108(23): p. </p><p>9530-9535. </p><p>10. Wan, J.C., et al., Liquid biopsies come of age: towards implementation of circulating </p><p>tumour DNA. Nature Reviews Cancer, 2017. 17(4): p. 223-238. </p><p>11. </p><p>Chin, R.-I., et al., Detection of solid tumor molecular residual disease (MRD) using </p><p>circulating tumor DNA (ctDNA). Molecular diagnosis & therapy, 2019. 23(3): p. </p><p>311-331. </p><p>12. </p><p>Sater, V., et al., UMI-Gen: A UMI-based read simulator for variant calling evaluation in </p><p>paired-end sequencing NGS libraries. Computational and structural biotechnology </p><p>journal, 2020. 18: p. 2270-2280. </p><p>13. </p><p>Schmitt, M.W., et al., Detection of ultra-rare mutations by next-generation sequencing. </p><p>Proceedings of the National Academy of Sciences, 2012. 109(36): p. 14508-14513. </p><p>14. </p><p>Valentine, C.C., et al., Direct quantification of in vivo mutagenesis and carcinogenesis </p><p>using duplex sequencing. Proceedings of the National Academy of Sciences, 2020. </p><p>117(52): p. 33414-33425. </p><p>15. </p><p>Short, N.J., et al., Ultra-accurate Duplex Sequencing for the assessment of </p><p>pretreatment ABL1 kinase domain mutations in Ph+ ALL. Blood cancer journal, 2020. </p><p>10(5): p. 1-9. </p><p>16. </p><p>Ahn, E.H., et al., Detection of ultra-rare mitochondrial mutations in breast stem cells by </p><p>duplex sequencing. PloS one, 2015. 10(8): p. e0136216. </p><p>17. </p><p>Kim, S., K. Jeong, and V. Bafna, Wessim: a whole-exome sequencing simulator based </p><p>on in silico exome capture. Bioinformatics, 2013. 29(8): p. 1076-1077. </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is the author/funder. All rights reserved. No reuse allowed without permission. </p><p>18. </p><p>Danecek, P., et al., Twelve years of SAMtools and BCFtools. Gigascience, 2021. </p><p>10(2): p. giab008. </p><p>19. </p><p>Zhao, X., et al., Calling small variants using universality with Bayes-factor-adjusted </p><p>odds ratios. Briefings in Bioinformatics, 2021. </p><p>20. </p><p>Bonfield, J.K., et al., HTSlib: C library for reading/writing high-throughput sequencing </p><p>data. Gigascience, 2021. 10(2): p. giab007. </p><p>21. </p><p>Jones, L., WG14 N1539 Committee Draft ISO/IEC 9899: 201x. 2010, International </p><p>Standards Organization. </p><p>22. </p><p>Niu, D., et al., Evaluation of next generation sequencing for detecting HER2 copy </p><p>number in breast and gastric cancers. Pathology & Oncology Research, 2020. 26(4): p. </p><p>2577-2585. </p><p>23. </p><p>Danecek, P., et al., The variant call format and VCFtools. Bioinformatics, 2011. 27(15): </p><p>p. 2156-2158. </p><p>24. </p><p>Box, G.E., A note on the generation of random normal deviates. Ann. Math. Statist., </p><p>1958. 29: p. 610-611. </p><p>25. </p><p>Deveson, I.W., et al., Evaluating the analytical validity of circulating tumor DNA </p><p>sequencing assays for precision oncology. Nature biotechnology, 2021. 39(9): p. </p><p>1115-1128. </p><p>26. </p><p>Fang, L.T., et al., Establishing community reference samples, data and call sets for </p><p>benchmarking cancer mutation detection using whole-genome sequencing. Nature </p><p>biotechnology, 2021. 39(9): p. 1151-1160. </p><p>27. </p><p>Li, H., et al., The Sequence Alignment/Map format and SAMtools. Bioinformatics, </p></div><div class='page'><p>bioRxiv preprint </p><p>doi: </p><p>https://doi.org/10.1101/2023.03.14.532524 </p><p>; </p><p>this version posted March 15, 2023. </p><p>The copyright holder for this preprint </p><p>(which was not certified by peer review) is\"}\n"
     ]
    }
   ],
   "source": [
    "!head output.jsonl\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "db71218b-4295-4f02-bcd6-b8e43656dd4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "def create_jsonl_from_directory(directory_path, output_jsonl_path, max_tokens=2048):\n",
    "    # List all files in the directory\n",
    "    file_list = os.listdir(directory_path)\n",
    "\n",
    "    with open(output_jsonl_path, 'w', encoding='utf-8') as jsonl_file:\n",
    "        for filename in file_list:\n",
    "            file_path = os.path.join(directory_path, filename)\n",
    "            \n",
    "            # Check if the path is a file (and not a directory)\n",
    "            if os.path.isfile(file_path):\n",
    "                # Read the contents of the file\n",
    "                with open(file_path, 'r', encoding='utf-8') as f:\n",
    "                    file_contents = f.read()\n",
    "\n",
    "                # Split the file contents into chunks of max_tokens characters\n",
    "                for i in range(0, len(file_contents), max_tokens):\n",
    "                    chunk = file_contents[i:i + max_tokens]\n",
    "\n",
    "                    # Create a JSON object with the chunk contents\n",
    "                    json_object = {\"text\": chunk}\n",
    "\n",
    "                    # Write the JSON object as a JSONL line\n",
    "                    jsonl_file.write(json.dumps(json_object) + '\\n')\n",
    "\n",
    "# Example usage:\n",
    "directory_path = \"html/papers\"\n",
    "output_jsonl_path = \"output.jsonl\"\n",
    "create_jsonl_from_directory(directory_path, output_jsonl_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "1d0cc210-2bc1-4c49-867d-90f47ab9fa47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   19217 output.jsonl\n"
     ]
    }
   ],
   "source": [
    "!wc -l output.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "561916b4-1614-49a5-b520-6f2cbc342c1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!head -n 19000 output.jsonl > data/train.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "a1e1ef40-2050-4c0f-a950-9fe2e5352ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "!tail -n +19000 output.jsonl > data/valid.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "9498d18a-51b5-492d-9558-3ee9ff19cf6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!tail -n +50 data/valid.jsonl > data/test.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "68cc315c-03d6-41eb-a47f-b00dba4f4458",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   19000 data/train.jsonl\n"
     ]
    }
   ],
   "source": [
    "!wc -l data/train.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "21c52555-1a81-417c-a0b9-eb5c4469dbb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     218 data/valid.jsonl\n"
     ]
    }
   ],
   "source": [
    "!wc -l data/valid.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "112ed2a9-07be-4638-84a7-7649270af4c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     169 data/test.jsonl\n"
     ]
    }
   ],
   "source": [
    "!wc -l data/test.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "f7aca99b-680f-48cb-8d04-06329772cf07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"text\": \"<html><body><div class='page'><p>bioRxiv preprint \\n</p><p>doi: \\n</p><p>https://doi.org/10.1101/2023.05.12.540616\\n</p><p>; \\n</p><p>this version posted June 30, 2023. \\n</p><p>The copyright holder for this preprint\\n</p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made \\n</p><p>available under a\\n</p><p>CC-BY 4.0 International license\\n.\\n</p><p>Compression algorithm for colored de Bruijn graphs\\n</p><p>Amatur Rahman1\\n</p><p>Yoann Dufresne4,5\\n</p><p>Paul Medvedev1,2,3\\n</p><p>1 Department of Computer Science and Engineering, The Pennsylvania State University\\n2 Department of Biochemistry and Molecular Biology, The Pennsylvania State University\\n3 Huck Institutes of the Life Sciences, The Pennsylvania State University\\n4 Institut Pasteur, Universit\\u00b4e Paris Cit\\u00b4e, G5 Sequence Bioinformatics, Paris, France\\n5 Institut Pasteur, Universit\\u00b4e Paris Cit\\u00b4e, Bioinformatics and Biostatistics Hub, F-75015 Paris, France\\n</p><p>Abstract\\n</p><p>A colored de Bruijn graph (also called a set of k-mer sets), is a set of k-mers with every k-mer\\nassigned a set of colors. Colored de Bruijn graphs are used in a variety of applications, including variant\\ncalling, genome assembly, and database search. However, their size has posed a scalability challenge\\nto algorithm developers and users. There have been numerous indexing data structures proposed that\\nallow to store the graph compactly while supporting fast query operations. However, disk compression\\nalgorithms, which do not need to support queries on the compressed data and can thus be more space-\\nefficient, have received little attention. The dearth of specialized compression tools has been a detriment\\nIn this paper, we develop a new tool that\\nto tool developers, tool users, and reproducibility efforts.\\ncompresses colored de Bruijn graphs to disk, building on previous ideas for compression of k-mer sets\\nand indexing colored de Bruijn graphs. We test our tool, called ESS-color, on various datasets, inc\"}\n",
      "{\"text\": \"luding\\nboth sequencing data and whole genomes. ESS-color achieves better compression than all evaluated\\ntools and all datasets, with no other tool able to consistently achieve less than 44% space overhead. The\\nsoftware is available at http://github.com/medvedevgroup/ESSColor.\\n</p></div><div class='page'><p>bioRxiv preprint \\n</p><p>doi: \\n</p><p>https://doi.org/10.1101/2023.05.12.540616\\n</p><p>; \\n</p><p>this version posted June 30, 2023. \\n</p><p>The copyright holder for this preprint\\n</p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made \\n</p><p>available under a\\n</p><p>CC-BY 4.0 International license\\n.\\n</p><p>1 Introduction\\n</p><p>Modern methods for analyzing biological sequences often reduce the input dataset to a set of short, fixed\\nlength strings called k-mers. When working with a collection of such datasets E = (E0, . . . , E|E|\\u22121), it is\\nfruitful to represent them as one union set of k-mers and, for each k-mer, the indices of the datasets to which\\nit belongs. The set of indices of each k-mer is referred to as its color class, and E is referred to as a colored\\nde Bruijn graph [1]. A colored de Bruijn graph (cdBG) is commonly used to represent a sequence database,\\nsuch as a collection of sequencing experiments or a collection of assembled genomes. For example, it is used\\nby tools for inferring phylogenies [2], quantification of RNA expression [3], and studying the evolution of\\nantimicrobial resistance [4].\\n</p><p>As sequence database sizes grow to petabytes [5], the cost of storing or transferring the data (e.g. on\\nAmazon Web Services or in-house compute infrastructure) has underscored the need for efficient disk com-\\npression algorithms. Such costs are often prohibitive for smaller labs and, even for larger labs, limit the scale\\nof data that can be analyzed. Large file sizes also hamper tool development, which relies on iterative load-\\ning/copying/modifying data, and reproducibility efforts, which require downloa\"}\n",
      "{\"text\": \"ding and storing the data.\\nFor example, storing the 31-mers from 450,000 microbial genomes in compressed form takes about 12 Tb [4].\\nUnfortunately, there has not been a lot of work to develop methods for disk compression of colored de Bruijn\\ngraphs.\\n</p><p>In contrast to disk compression, indexing cdBGs have received much attention [6]. A slew of data\\nstructures have been developed, optimizing various metrics such as index size, construction time, or query\\ntime (see the survey [6] and its follow up [7]). Indexing data structures exploit the structure of cdBGs and\\nuse clever tricks to compress the color classes of similar k-mers. But they also carry a space overhead to\\nefficiently support queries; since this is not needed for disk compression, indexing data structures are usually\\nnot competitive with custom made cdBG compression methods.\\n</p><p>The simplest option for compressing a cdBG is to compress each color (i.e. dataset) independently, using\\na compression tool designed for a single set of k-mers (e.g. [8]). This approach can work well when k-mers\\ntend to not be shared among colors. However, most cdBGs have a large overlap between the k-mers of various\\ncolors. In such cases, independently compressing each color does not exploit the properties of cdBGs and, as\\nwe show in this paper, results in subpar compression ratios. There exist two tools designed specifically for disk\\ncompression of cdBGs. The first tool is unfortunately limited to only three k-mer sizes [9]. The second tool,\\ncalled GGCAT [10] is a space efficient indexing method that, while not originally evaluated in this regard,\\nturns out to also be a good disk compression method when combined with a generic post-compression step.\\nIn this paper, we design, implement, and evaluate an algorithm ESS-color for the disk compression of\\ncdBGs. We build upon the idea of spectrum-preserving string sets [11, 12, 13] and the followup compression\\nformat for a k-mer set [8], called ESS. By constructing an ESS of the union of k-mers in E, we represent the\\nk-mer seque\"}\n",
      "{\"text\": \"nces themselves compactly. We exploit the fact that consecutive k-mers in an ESS have similar\\ncolor classes in order to efficiently compress the color vectors of each k-mer.\\n</p><p>We evaluate ESS-color on a variety of datasets, including bacteria, fungi, human, and including whole\\ngenome sequencing data, metagenome sequencing data, and whole assembled genomes. ESS-color achieves\\nbetter compression than all evaluated tools and on all datasets, with all other tools using \\u2265 44% more space\\non at least one of our datasets. On some datasets the improvement over all other tools is quite large, e.g.\\nfor a gut metagenome, all the other tools use at \\u2265 27% more space than ESS-color. Compressing each color\\nindependently uses between 1.2x and 6.9x more space than ESS-color. The absolute compression ratio is\\nmore than 26x on datasets of assembled genomes and between 1.4x and 8.7x on datasets from sequencing\\nexperiments. The software is available at http://github.com/medvedevgroup/ESSColor.\\n</p><p>2 Preliminaries\\n</p><p>In this section we give some important definitions. Please refer to Figure 1 for examples of the introduced\\nconcepts.\\n</p><p>Strings: A string of length k is called a k-mer. We assume k-mers are over the DNA alphabet. A string x\\nis canonical if it is the lexicographically smaller of x and its reverse complement. Let K be a set of k-mers.\\n</p><p>1\\n</p></div><div class='page'><p>bioRxiv preprint \\n</p><p>doi: \\n</p><p>https://doi.org/10.1101/2023.05.12.540616\\n</p><p>; \\n</p><p>this version posted June 30, 2023. \\n</p><p>The copyright holder for this preprint\\n</p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made \\n</p><p>available under a\\n</p><p>CC-BY 4.0 International license\\n.\\n</p><p>A spectrum-preserving string set (SPSS) of K is a set of strings S such that each string s \\u2208 S is at least k\\ncharacters long, every k-mer that appears in S appears exactly once, and the set of k-mers that appear in S\\nis K [11, 12, 13]. For\"}\n",
      "{\"text\": \" example, if K = {ACG, CGT, CGA}, then {ACGT, CGA} would an SPSS of K. Note\\nthat K can have multiple spectrum-preserving string sets. There are several efficient tools for computing an\\nSPSS so as to minimize the total number of characters [14, 10]. In this paper, we rely on the implementation\\nin [8]. Each string in the resulting SPSS is referred to as a simplitig.\\n</p><p>Compression of a k-mer set ESS is a disk-compression format to store a set of k-mers K.\\nIt was\\nintroduced in [8] as the output of a compression tool, which, in this paper, we will refer to as ESS-basic. The\\ntechnical details of the format and of the tool are irrelevant for this paper and can be viewed as black boxes.\\nAn ESS representation cannot be queried efficiently but can be decompressed into an SPSS of K. This\\noutput gives an ordering of the k-mers of K, and therefore the ESS compression of K induces an ordering\\non K. Note that because the decompression algorithm is deterministic, by storing an ESS representation,\\nwe are implicitly storing an SPSS representation as well.\\n</p><p>Colored k-mer sets: Let C > 0 be an integer indicating the number of colors. Let E = {E0, . . . , EC\\u22121}\\nbe a set of C k-mer sets, also referred to as a colored de Bruijn graph. Let E be the set of all k-mers in E,\\ni.e. E = {x | \\u2203i s. t. x \\u2208 Ei}. The (color) class of a k-mer x \\u2208 E is the set of indices i such that x \\u2208 Ei.\\nThe color vector of x is a binary vector of length C where position i is 1 iff x \\u2208 Ei.\\n</p><p>Non-compressed representation of cdBGs: Assume you have an ordering of E, e.g. the one given by\\nan ESS of E. A color matrix of E is a file with row i containing the color vector of the ith k-mer. Storing\\nan ESS of E together with a color matrix of E is a lossless representation of E.\\n</p><p>3 Methods\\n</p><p>In this section, we describe our algorithm ESS-color for the compression of cdBGs. Let E = {E0, . . . , EC\\u22121}\\nbe a colored dBG over C colors. Recall that E is the set of all k-mers in E. Let M denote the number of\\ndistinct color classes in E.\\n</p><p>ESS-co\"}\n",
      "{\"text\": \"lor can accept input in one of two formats. First, it can accept each Ei stored in the KFF file\\nformat [15]. Alternatively, we can take as input a collection of FASTA files, each one assigned one of C\\ncolors, and an abundance parameter a. Ei is then implicitly defined as the set of all canonical k-mers that\\nappear at least a times in the FASTA files of color i. We obtain Ei by running KMC [16] on the FASTA\\nfiles of color i.\\n</p><p>3.1 Building a color matrix of E and compressing E\\n</p><p>In this step, we first compress E using ESS-basic [8] and then build the color matrix of E, ordered by the\\nESS order. Specifically, we first compress the nucleotide sequences themselves, i.e. we run ESS-basic [8] on\\nall the input files jointly. We refer to this as the union ESS. We then decompress this file to obtain an SPSS\\nof E, denoted by S. From S, we build an SSHash dictionary [17] that allows us to map each k-mer in E to\\nits rank in S. We then build on top of the KMC API to read in the binary files representing E0, . . . , EC\\u22121\\nand output a color matrix ordered by the SSHash dictionary. At the end of this stage, we have the union\\nESS, which is retained in the final compression output, and we have S and the color matrix, which are used\\nin later stages but not retained in the final compression output.\\n</p><p>3.2 Compression of the color matrix\\n</p><p>Given an SPSS S of E and a color matrix of E over the order induced by S, we now generate a compressed\\nrepresentation of the matrix. Our representation consists of a global class table and, for every simplitig of\\nS, a few bits of metadata, a local class table and one bitvector m. The local class table is optional, as we\\ndescribe below. Figure 2 gives a schematic representation. We now explain each of these in detail.\\n</p><p>2\\n</p></div><div class='page'><p>bioRxiv preprint \\n</p><p>doi: \\n</p><p>https://doi.org/10.1101/2023.05.12.540616\\n</p><p>; \\n</p><p>this version posted June 30, 2023. \\n</p><p>The copyright holder for this preprint\\n</p><p>(which was not certified by peer review) \"}\n",
      "{\"text\": \"is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made \\n</p><p>available under a\\n</p><p>CC-BY 4.0 International license\\n.\\n</p><p>Figure 1: An example illustrating the various definitions in Section 2 and the first step of our compression\\nmethod (Section 3.1). The input to the compression algorithm is a colored de Bruin Graph. The top panel\\nshows an example with three colors (i.e. C = 3), k = 5, and a colored cdBG of E = {E0, E1, E2}. Here, E0 =\\n{T CAAA, CAAAA, AAAAT }, E1 = {T CAAA, CAAAA, AAAAT, AAAT T, CAAAG, AAAT C, AAT CG},\\nand E2 = {T CAAA, CAAAA, AAAAT, AAAT T, CAAAG}. The color class is shown next to each k-mer,\\nthe color class of T CAAA is {0, 1, 2}. There are three distinct color classes, i.e. M = 3. These\\ne.g.\\nare {0, 1, 2}, {1, 2}, and {1}. The lower left panel shows the union ESS, i.e.\\nthe ESS representation of\\nthe set E = {T CAAA, CAAAA, AAAAT, AAAT T, CAAAG, AAAT C, AAT CG}. This union ESS can be\\ndecomposed into an SPSS of E, shown in the figure. The third column in bottom panel shows the color\\nmatrix, with k-mers in the order of the SPSS. To obtain the final compressed representation the color matrix\\nis compressed using the algorithm we describe in Section 3.2.\\n</p><p>3\\n</p></div><div class='page'><p>bioRxiv preprint \\n</p><p>doi: \\n</p><p>https://doi.org/10.1101/2023.05.12.540616\\n</p><p>; \\n</p><p>this version posted June 30, 2023. \\n</p><p>The copyright holder for this preprint\\n</p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made \\n</p><p>available under a\\n</p><p>CC-BY 4.0 International license\\n.\\n</p><p>Global class table: For most applications, the number of distinct color vectors M is significantly smaller\\nthan 2C. Hence, the color matrix representation, which uses C bits per k-mer, is very inefficient. Instead,\\nwe use Huffman coding to assign a global ID to each class, so as to minimize the number of bits that\\nwill be used to store these IDs later (\"}\n",
      "{\"text\": \"this is similar to what was done in [18]). To do this, we scan the\\ncolor matrix to determine all the distinct classes and the number of k-mers that have each class. We\\nthen use Huffman coding to assign a global ID to each distinct class, so that more frequent global IDs\\ntend to use less bits. This table is then stored in two forms: one that is compressed to disk, and the\\nother that is stored in memory to be used during the compression algorithm.\\n</p><p>We store the table on disk using three files: a color encoding \\u2206, a boundary bitvector b, and a text\\nfile. First, we sort the color classes in increasing numerical order, interpreting each color vector as a\\nC-bit integer. For \\u2206, we write a concatenation of the M color vectors to disk, with the first color vector\\nbeing written using C bits and the following colors being encoded as a difference with their predecessor.\\nSpecifically, if hi is the Hamming distance between the ith and the (i \\u2212 1)st color vectors, then we use\\nhi\\u2308log C\\u2309 bits to encode the indices where the ith color vector is different from the (i \\u2212 1)st color vector.\\nWe also store a boundary bitvector b which is the same length as \\u2206 and contains a \\u20181\\u2019 whenever \\u2206\\nstarts a new color class. Finally, we store the frequencies of the color classes in a text file. These three\\nfiles are then sufficient to reconstruct the global IDs during decompression. 1\\n</p><p>Simultaneously, we need to be able to map a color vector to an ID during the compression process. To\\ndo this, we create a minimal perfect hash function h (CHM [20]) that maps from each distinct color\\nvector to an integer between 0 and M \\u2212 1. We then maintain an array A of size M , where for each color\\nvector c, A(h(c)) holds the global ID of color vector c.\\n</p><p>After the global class table is created, we process the simplitigs of S one at time. For each simplitig, we\\ndynamically set two parameters: a boolean variable U seLocalID, and an integer 0 \\u2264 maxDif \\u2264 2. We\\npostpone the discussion on how these are set until the end of the section. The values of m\"}\n",
      "{\"text\": \"axDif and\\nU seLocalID are stored using 3 bits of metadata per simplitig. If U seLocalID is set, we create a local class\\ntable:\\n</p><p>Local class table: In the case that the frequencies of color classes are evenly distributed, we need\\napproximately log M bits to represent the global class ID of each k-mer. We observe that sometimes a\\nclass is used at multiple locations of a simplitig, in which case using log M bits for each occurrence can\\nbe wasteful. Let \\u2113 be the number of distinct classes appearing in a simplitig. To save space on class\\nIDs, we create a separate local class table, which maps from \\u2113 integers, called local class IDs, to their\\nrespective global IDs. Then, the encoding of k-mer classes for this simplitig can use local class IDs,\\nwhich take only log \\u2113 space. The local class table is written to disk, with log M bits encoding \\u2113 followed\\nby \\u2113 consecutive global class IDs together\\n</p><p>The bitvector m is constructed by scanning the simplitig from left to right and, for each k-mer x, deciding how\\nto encode it, and appending that encoding to m. Intuitively, the encodings follow three basic possibilities.\\nThe first possibility is to just append m with the k-mer\\u2019s class ID. Second, we observed in practice that\\nsimplitigs often contain runs of k-mers with identical classes, in which case we can append m with the length\\nof the run, rather than writing out each class IDs (such runs are similar to the monotigs of [21]). Finally,\\nwe often observe that a k-mer has a color vector with a small Hamming distance (i.e. 1 or 2) to that of the\\nprevious k-mer. In this case, we append m with the indices in the color vector that are different. Since there\\nare three types of encoding, we will also need to prepend each encoding with two bits denoting the type of\\nencoding. Formally, for each k-mer in a simplitig, we choose one of four options:\\n</p><p>Skip: This option is invoked if x is not the first or last k-mer in its simplitig and has the same class as\\nthe preceding and succeeding k-mer. In this option, nothing is appen\"}\n",
      "{\"text\": \"ded to m.\\n</p><p>Small class difference: Let h be the Hamming distance between the color vector of x and the color\\nvector of the preceding k-mer in the simplitig. This option is invoked when 0 < h \\u2264 maxDif . First, we\\n</p><p>1For readers familiar with Mantis-MST [19], we also tried their approach for storing our global table. Surprisingly, we found\\nthat our approach outperformed their more sophisticated approach, at least in our datasets. Though the Mantis-MST approach\\nresulted in a smaller \\u2206 vector, the overhead of storing the tree parent vector outweighed this gain.\\n</p><p>4\\n</p></div><div class='page'><p>bioRxiv preprint \\n</p><p>doi: \\n</p><p>https://doi.org/10.1101/2023.05.12.540616\\n</p><p>; \\n</p><p>this version posted June 30, 2023. \\n</p><p>The copyright holder for this preprint\\n</p><p>(which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in perpetuity. It is made \\n</p><p>available under a\\n</p><p>CC-BY 4.0 International license\\n.\\n</p><p>Figure 2: Example of how we compress the color matrix (Section 3.2). The top panel shows the compression\\nof a simplitig ACTTTGGA and the bottom panel shows the compression of a simplitig AAAATTAC. Other\\nsimplitigs exist but are not shown. In this example, C = 6, k = 4, and runDivisor = 4. The figure is divided\\ninto three columns. The first column shows the information that the algorithm has about each k-mer (i.e. its\\ncolor vector and corresponding global ID). The second column shows the metadata which holds the values\\nof maxDif and useLocalId for the corresponding simplitig. The third column shows the m vector. Within\\nthe m vector, the type of encoding is shown in gray colored boxes and other values are shown in white boxes.\\nThe colors of the values inside the white boxes correspond to the color used for the corresponding k-mers\\u2019\\ncolor vector in the first column of the figure. Note that in the bottom simplitig, a local table is used. In\\nparticular, there are two distinct color vectors in this simplitig, with gl\"}\n"
     ]
    }
   ],
   "source": [
    "!head data/train.jsonl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddd95bd-2acd-4452-911b-bb2e4077c20a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
